<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Microbial Genomics: evolution, phylogenetics, and molecular epidemiology: Reads, QC and trimming</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav data"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Data Carpentry" src="assets/images/data-logo.svg"></div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/quality-control.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="assets/images/data-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 50%" class="percentage">
    50%
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: 50%" aria-valuenow="50" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/quality-control.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="ncbi.html">1. NCBI</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="blast.html">2. BLAST</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="msa.html">3. Multiple sequence alignment</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="phylogenetics.html">4. Phylogenetics</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        5. Reads, QC and trimming
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#introduction">Introduction</a></li>
<li><a href="#background-of-the-data">Background of the data</a></li>
<li><a href="#getting-your-project-started">Getting your project started</a></li>
<li><a href="#selection-of-a-reference-genome">Selection of a reference genome</a></li>
<li><a href="#loops">Loops</a></li>
<li><a href="#download-the-sequenced-genomes-from-the-european-nucleotide-archive-ena">Download the sequenced genomes from the European Nucleotide Archive
(ENA)</a></li>
<li><a href="#bioinformatics-workflows">Bioinformatics workflows</a></li>
<li><a href="#quality-control">Quality Control</a></li>
<li><a href="#cleaning-reads">Cleaning Reads</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="sequence-assembly.html">6. Sequence assembly</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="annotation.html">7. Annotation</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="snps-phylogeny.html">8. Mapping</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="data-visualization.html">9. Molecular epidemiology</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="phylogenetics.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="sequence-assembly.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="phylogenetics.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Phylogenetics
        </a>
        <a class="chapter-link float-end" href="sequence-assembly.html" rel="next">
          Next: Sequence assembly... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Reads, QC and trimming</h1>
        <p>Last updated on 2024-03-15 |
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/quality-control.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<p>This episode starts the a series of labs where you will trace an
outbreak of <em>Mycobacterium tuberculosis</em>, from downloading the
reads to making assumptions about the transmissions between
patients.</p>
<p>The first part, here, describes how to retrieve all necessary data
from public repositories - the raw sequenced data of our isolates and a
reference genome. It also introduces for loops which we will use for the
rest of our analysis.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Trace an outbreak of Mycobacterium, from reads to phylogenetic
tree.</li>
<li>How can I organize my file system for a new bioinformatics
project?</li>
<li>How and where can data be downloaded?</li>
<li>How can I describe the quality of my data?</li>
<li>How can I get rid of sequence data that doesn’t meet my quality
standards?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>How do you use genomics and phylogenetics to trace a microbial
outbreak?</li>
<li>Create a file system for a bioinformatics project.</li>
<li>Download files necessary for further analysis.</li>
<li>Use ‘for’ loops to automate operations on multiple files</li>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Clean FastQC reads for further analysis.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width"><p>This part of the course is largely inspired from <a href="https://github.com/aschuerch/MolecularEpidemiology_AnalysisWGS" class="external-link">another
tutorial</a> developed by <a href="https://aschuerch.github.io/" class="external-link">Anita
Schürch</a>. We will use (part of) the data described in <a href="https://bmcinfectdis.biomedcentral.com/articles/10.1186/1471-2334-13-110" class="external-link">Bryant
et al., 2013</a>.</p>
</section><section id="background-of-the-data"><h2 class="section-heading">Background of the data<a class="anchor" aria-label="anchor" href="#background-of-the-data"></a>
</h2>
<hr class="half-width"><p>Epidemiological contact tracing through interviewing of patients can
identify potential chains of patients that transmitted an infectious
disease to each other. Contact tracing was performed following the <a href="https://www.ncbi.nlm.nih.gov/pubmed/1643300" class="external-link">stone-in-the-pond
principle</a>, which interviews and tests potential contacts in
concentric circles around a potential source case.</p>
<p>Tuberculosis (TB) is an infectious disease caused by
<em>Mycobacterium tuberculosis</em>. It mostly affects the lungs. An
infection with <em>M. tuberculosis</em> is often asymptomatic (latent
infection). Only in about 10% of the cases the latent infection
progresses to an active infection during a patients lifetime, which, if
untreated, leads to death in about half of the cases. The symptoms of an
active TB infection include cough, fever, night sweats, weight loss etc.
An active TB infection can spread. Once exposed, people often have
latent TB. To identify people with latent TB, a <a href="https://www.cdc.gov/tb/publications/factsheets/testing/skintesting.htm" class="external-link">skin
test</a> can be applied.</p>
<p>Here we have 7 tuberculosis patients with active TB, that form three
separate clusters of potential transmission as determined by
epidemiological interviews. Patients were asked if they have been in
direct contact with each other, or if they visited the same localities.
From all patients, a bacterial isolate was grown, DNA isolated, and
whole-genome sequenced on an Illumina sequencer.</p>
<p>The three clusters consist of:</p>
<ul><li>Cluster 1
<ul><li>Patient A1 (2011) - sample ERR029207</li>
<li>Patient A2 (2011) - sample ERR029206</li>
<li>Patient A3 (2008) - sample ERR026478</li>
</ul></li>
<li>Cluster 2
<ul><li>Patient B1 (2001) - sample ERR026474</li>
<li>Patient B2 (2012) - sample ERR026473</li>
</ul></li>
<li>Cluster 3
<ul><li>Patient C1 (2011) - sample ERR026481</li>
<li>Patient C2 (2016) - sample ERR026482</li>
</ul></li>
</ul><p>The second goal of this lab is to affirm or dispute the inferred
transmissions by comparing the bacterial genomes with each other. We
will do that by identifying SNPs between the genomes.</p>
</section><section id="getting-your-project-started"><h2 class="section-heading">Getting your project started<a class="anchor" aria-label="anchor" href="#getting-your-project-started"></a>
</h2>
<hr class="half-width"><p>Project organization is one of the most important parts of a
sequencing project, and yet is often overlooked amidst the excitement of
getting a first look at new data. Of course, while it’s best to get
yourself organized before you even begin your analyses,it’s never too
late to start, either.</p>
<p>Genomics projects can quickly accumulate hundreds of files across
tens of folders. Every computational analysis you perform over the
course of your project is going to create many files, which can
especially become a problem when you’ll inevitably want to run some of
those analyses again. For instance, you might have made significant
headway into your project, but then have to remember the PCR conditions
you used to create your sequencing library months prior.</p>
<p>Other questions might arise along the way:</p>
<ul><li>What were your best alignment results?</li>
<li>Which folder were they in: Analysis1, AnalysisRedone, or
AnalysisRedone2?</li>
<li>Which quality cutoff did you use?</li>
<li>What version of a given program did you implement your analysis
in?</li>
</ul><p>In this exercise we will setup a file system for the project we will
be working on during this workshop.</p>
<p>We will start by creating a directory that we can use for the rest of
the workshop. First navigate to your home directory. Then confirm that
you are in the correct directory using the <code>pwd</code> command.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">$</span> pwd</span></code></pre>
</div>
<p>You should see the output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>/home/chewbe</code></pre>
</div>
<div id="tip" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip" class="callout-inner">
<h3 class="callout-title">Tip<a class="anchor" aria-label="anchor" href="#tip"></a>
</h3>
<div class="callout-content">
<p>If you aren’t in your home directory, the easiest way to get there is
to enter the command <code>cd</code>, which always returns you to
home.</p>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>Use the <code>mkdir</code> command to make the following
directories:</p>
<pre><code><span><span class="va">molepi</span></span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">docs</span>  </span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">data</span>  </span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">results</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> mkdir molepi</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">$</span> mkdir molepi/docs</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="ex">$</span> mkdir molepi/data</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="ex">$</span> mkdir molepi/results</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Use <code>ls -R</code> to verify that you have created these
directories. The <code>-R</code> option for <code>ls</code> stands for
recursive. This option causes <code>ls</code> to return the contents of
each subdirectory within the directory iteratively.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-R</span> molepi</span></code></pre>
</div>
<p>You should see the following output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>molepi/:
data  docs  results

molepi/data:

molepi/docs:

molepi/results: </code></pre>
</div>
</section><section id="selection-of-a-reference-genome"><h2 class="section-heading">Selection of a reference genome<a class="anchor" aria-label="anchor" href="#selection-of-a-reference-genome"></a>
</h2>
<hr class="half-width"><p>Reference sequences (including many pathogen genomes) are available
at <a href="https://www.ncbi.nlm.nih.gov/refseq/" class="external-link">NCBI’s refseq
database</a></p>
<p>A reference genome is a genome that was previously sequenced and is
closely related to the isolates we would like to analyse. The selection
of a closely related reference genome is not trivial and will warrant an
analysis in itself. However, for simplicity, here we will work with the
<em>M. tuberculosis</em> reference genome H37Rv.</p>
<div class="section level3">
<h3 id="download-reference-genomes-from-ncbi">Download reference genomes from NCBI<a class="anchor" aria-label="anchor" href="#download-reference-genomes-from-ncbi"></a></h3>
<p>Download the <em>M.tuberculosis</em> reference genome from the NCBI
ftp site.</p>
<p>First, we switch to the data folder to store all our data</p>
<pre class="source"><code>$ cd molepi/data </code></pre>
<p>The reference genome will be downloaded programmatically from NCBI
with <code>wget</code>. <code>wget</code> is a computer program that
retrieves content from web servers. Its name derives from World Wide Web
and get.</p>
<pre class="source"><code>$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/195/955/GCF_000195955.2_ASM19595v2/GCF_000195955.2_ASM19595v2_genomic.fna.gz</code></pre>
<p>This file is compressed as indicated by the extension of
<code>.gz</code>. It means that this file has been compressed using the
<code>gzip</code> command.</p>
<p>Extract the file by typing</p>
<pre class="source"><code>$ gunzip GCF_000195955.2_ASM19595v2_genomic.fna.gz</code></pre>
<p>Make sure that is was extracted</p>
<pre class="source"><code>$ ls</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GCF_000195955.2_ASM19595v2_genomic.fna</span></span></code></pre>
</div>
<div id="challenge-what-is-the-size-of-the-genome" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-what-is-the-size-of-the-genome" class="callout-inner">
<h3 class="callout-title">Challenge: What is the size of the
genome?<a class="anchor" aria-label="anchor" href="#challenge-what-is-the-size-of-the-genome"></a>
</h3>
<div class="callout-content">
<p>Find out how many basepairs the genome has. Hints:</p>
<pre><code><span><span class="va">assembly</span><span class="op">-</span><span class="va">stats</span></span></code></pre>
<p>Get assembly statistics from FASTA and FASTQ files.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>The genome has 4’411’532 bp.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> assembly-stats GCF_000195955.2_ASM19595v2_genomic.fna</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
stats for GCF_000195955.2_ASM19595v2_genomic.fna
sum = 4411532, n = 1, ave = 4411532.00, largest = 4411532
N50 = 4411532, n = 1
N60 = 4411532, n = 1
N70 = 4411532, n = 1
N80 = 4411532, n = 1
N90 = 4411532, n = 1
N100 = 4411532, n = 1
N_count = 0
Gaps = 0</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="loops"><h2 class="section-heading">Loops<a class="anchor" aria-label="anchor" href="#loops"></a>
</h2>
<hr class="half-width"><p><em>Loops</em> are key to productivity improvements through
automation as they allow us to execute commands repeatedly. Similar to
wildcards and tab completion, using loops also reduces the amount of
typing (and typing mistakes). Our next task is to download our <a href="molecular-epidemiology.html">data</a> from the short read archive
<a href="https://www.ebi.ac.uk/ena" class="external-link">(SRA) at the European Nucleotide
Archive (ENA)</a>. There are many repositories for public data. Some
model organisms or fields have specific databases, and there are ones
for particular types of data. Two of the most comprehensive are the
National Center for Biotechnology Information (NCBI) and European
Nucleotide Archive (EMBL-EBI). In this lesson we’re working with the
ENA, but the general process is the same for any database.</p>
<p>We can do this one by one but given that each download takes about
one to two hours, this could keep us up all night. Instead of
downloading one by one we can apply a loop. Let’s see what that looks
like and then we’ll discuss what we’re doing with each line of our
loop.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">$</span> for filename in ERR01 ERR02 ERR03</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="op">&gt;</span> echo <span class="ex">ftp://ftp.sra.ebi.ac.uk/</span><span class="st">"</span><span class="va">${filename}</span><span class="st">"</span><span class="ex">.fastq.gz</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>When the shell sees the keyword <code>for</code>, it knows to repeat
a command (or group of commands) once for each item in a list. Each time
the loop runs (called an iteration), an item in the list is assigned in
sequence to the <strong>variable</strong>, and the commands inside the
loop are executed, before moving on to the next item in the list.</p>
<p>Inside the loop, we call for the variable’s value by putting
<code>$</code> in front of it. The <code>$</code> tells the shell
interpreter to treat the <strong>variable</strong> as a variable name
and substitute its value in its place, rather than treat it as text or
an external command.</p>
<p>In this example, the list is seven filenames/ Each time the loop
iterates, it will assign a file name to the variable
<code>filename</code> and run the <code>wget</code> command. The first
time through the loop, <code>$filename</code> is
<code>ERR01.fastq.gz</code>. The interpreter runs the command
<code>wget</code> on <code>ERR01.fastq.gz</code> at the server <a href="ftp://ftp.sra.ebi.ac.uk/" class="uri">ftp://ftp.sra.ebi.ac.uk/</a>
For the second iteration, <code>$filename</code> becomes
<code>ERR02.fastq.gz</code>. This time, the shell runs <code>wget</code>
on <code>ERR02.fastq.gz</code>.</p>
<p>Use {} to wrap the variable so that .fastq.gz will not be interpreted
as part of the variable name. In addition, quoting the shell variables
is a good practice AND necessary if your variables have spaces in
them.</p>
<p>For more, check <a href="https://mywiki.wooledge.org/BashPitfalls" class="external-link">Bash Pitfalls</a></p>
<div id="follow-the-prompt" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="follow-the-prompt" class="callout-inner">
<h3 class="callout-title">Follow the Prompt<a class="anchor" aria-label="anchor" href="#follow-the-prompt"></a>
</h3>
<div class="callout-content">
<p>The shell prompt changes from <code>$</code> to <code>&gt;</code> and
back again as we were typing in our loop. The second prompt,
<code>&gt;</code>, is different to remind us that we haven’t finished
typing a complete command yet. A semicolon, <code>;</code>, can be used
to separate two commands written on a single line.</p>
</div>
</div>
</div>
<div id="same-symbols-different-meanings" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="same-symbols-different-meanings" class="callout-inner">
<h3 class="callout-title">Same Symbols, Different Meanings<a class="anchor" aria-label="anchor" href="#same-symbols-different-meanings"></a>
</h3>
<div class="callout-content">
<p>Here we see <code>&gt;</code> being used a shell prompt, whereas
<code>&gt;</code> is also used to redirect output. Similarly,
<code>$</code> is used as a shell prompt, but, as we saw earlier, it is
also used to ask the shell to get the value of a variable.</p>
<p>If the <em>shell</em> prints <code>&gt;</code> or <code>$</code> then
it expects you to type something, and the symbol is a prompt.</p>
<p>If <em>you</em> type <code>&gt;</code> or <code>$</code> yourself, it
is an instruction from you that the shell to redirect output or get the
value of a variable.</p>
</div>
</div>
</div>
<p>We have called the variable in this loop <code>filename</code> in
order to make its purpose clearer to human readers. The shell itself
doesn’t care what the variable is called; if we wrote this loop as:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> for x in ERR01 ERR02 ERR03</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${x}</span><span class="st">"</span>.fastq.gz</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>or:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">$</span> for temperature in ERR01 ERR02 ERR03</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${temperature}</span><span class="st">"</span>.fastq.gz</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>it would work exactly the same way. <em>Don’t do this.</em> Programs
are only useful if people can understand them, so meaningless names
(like <code>x</code>) or misleading names (like
<code>temperature</code>) increase the odds that the program won’t do
what its readers think it does.</p>
<div id="multipart-commands" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="multipart-commands" class="callout-inner">
<h3 class="callout-title">Multipart commands<a class="anchor" aria-label="anchor" href="#multipart-commands"></a>
</h3>
<div class="callout-content">
<p>The <code>for</code> loop is interpreted as a multipart command. If
you press the up arrow on your keyboard to recall the command, it may be
shown like so (depends on your bash version):</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> for filename in ERR01 ERR02 ERR03<span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${filename}</span><span class="st">"</span>.fastq.gz <span class="kw">;</span> <span class="cf">done</span></span></code></pre>
</div>
<p>When you check your history later, it will help your remember what
you did!</p>
</div>
</div>
</div>
</section><section id="download-the-sequenced-genomes-from-the-european-nucleotide-archive-ena"><h2 class="section-heading">Download the sequenced genomes from the European Nucleotide Archive
(ENA)<a class="anchor" aria-label="anchor" href="#download-the-sequenced-genomes-from-the-european-nucleotide-archive-ena"></a>
</h2>
<hr class="half-width"><p>To ensure that we complete this tutorial in time, we will not perform
all operations on all genomes, only on one or two. The principles are -
however - the same for two or seven (or, to some extent, 189)
genomes.</p>
<p>Let’s download data for the <em>M. tuberculosis</em> genomes with a
<code>for</code> loop. We would use the following loop to download all
samples:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> for files in ERR029/ERR029207 ERR029/ERR029206 ERR026/ERR026478 ERR026/ERR026474 ERR026/ERR026473 ERR026/ERR026481 ERR026/ERR026482</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="co"># wget "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/"${files}"/*fastq.gz"</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>Note the <code>#</code> that prevents the <code>wget</code> command
from running: that would take too long, so we’re downloading only one
(that’s a short loop!):</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> cd molepi/data</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="ex">$</span> for files in ERR029/ERR029206 </span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="fu">wget</span> <span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/"</span><span class="va">${files}</span><span class="st">"/*fastq.gz"</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>This will run for a little while (hopefully no more than 5-10
minutes). Let’s check if the downloads finished. Let’s have a look at
the files.</p>
<pre class="source"><code>$ ls</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">ERR029206_1.fastq.gz</span></span>
<span><span class="va">ERR029206_2.fastq.gz</span> </span></code></pre>
</div>
<p>All new files start with ERR and have a <code>fastq.gz</code>
extension. It means they are in fastq format and compressed. Let’s make
a copy of one file and uncompress this file to inspect it further</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">cp</span> ERR029206_1.fastq.gz ERR029206_1.copy.fastq.gz</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="fu">gunzip</span> ERR029206_1.copy.fastq.gz</span></code></pre>
</div>
</section><section id="bioinformatics-workflows"><h2 class="section-heading">Bioinformatics workflows<a class="anchor" aria-label="anchor" href="#bioinformatics-workflows"></a>
</h2>
<hr class="half-width"><p>When working with high-throughput sequencing data, the raw reads you
get off of the sequencer will need to pass through a number of different
tools in order to generate your final desired output. The execution of
this set of tools in a specified order is commonly referred to as a
<em>workflow</em> or a <em>pipeline</em>.</p>
<p>An example of the workflow we will be using is provided below with a
brief description of each step.</p>
<figure><img src="fig/Workflow.png" alt="Schematic workflow of the bioinformatics workflow" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal"><li>Quality control - Assessing quality using FastQC</li>
<li>Quality control - Trimming and/or filtering reads (if
necessary)</li>
<li>Align reads to reference genome - Snippy</li>
<li>Variant calling - Snippy</li>
<li>Clustering variants - iTOL</li>
<li>Assembly - SKESA</li>
<li>Annotation (extra material)</li>
<li>Pangenome analysis (extra material)</li>
<li>Clustering presence and absence of genes (extra material)</li>
<li>Comparison of clustering methods (extra material)</li>
<li>Data visualization - Microreact</li>
</ol><p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community. The
tools that are used to analyze data at different stages of the workflow
are therefore built under the assumption that the data will be provided
in a specific format.</p>
</section><section id="quality-control"><h2 class="section-heading">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<hr class="half-width"><p>The first step in the variant calling workflow is to take the FASTQ
files received from the sequencing facility and assess the quality of
the sequence reads.</p>
<figure><img src="fig/var_calling_workflow_qc.png" alt="Quality control workflow" class="figure mx-auto d-block"></figure><p>You’ve already seen that part, but repetition doesn’t hurt. If you
feel that it’s anyway unnecessary, move on to the next part.</p>
<div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a></h3>
<p>Although it looks complicated (and it is), it’s easy to understand
the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">fastq</a>
format with a little decoding. Some rules about the format include…</p>
<table class="table"><colgroup><col width="16%"><col width="83%"></colgroup><thead><tr class="header"><th>Line</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td>1</td>
<td>Always begins with ‘@’ and then information about the read</td>
</tr><tr class="even"><td>2</td>
<td>The actual DNA sequence</td>
</tr><tr class="odd"><td>3</td>
<td>Always begins with a ‘+’ and sometimes the same info in line 1</td>
</tr><tr class="even"><td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr></tbody></table><p>We can view the first complete read in one of the files our dataset
by using <code>head</code> to look at the first four lines.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n4</span> ERR029206_1.copy.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@ERR029206.1 IL31_5505:8:1:6233:1087#4/1
TCNAGTCAGCACACACATGCGAAAGAATCCACCGACTAGGGTCAGCGGGGTTTGCAGTTGGTCGCGGACGTAACCG
+
&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>One of the nucleotides in this read is unknown (<code>N</code>).</p>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g. 1 in 10)
or, equivalently, the base call accuracy (eg 90%). To make it possible
to line up each individual nucleotide with its quality score, the
numerical score is converted into a code where each individual character
represents the numerical quality score for an individual nucleotide. For
example, in the line above, the quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>The <code>&lt;</code> character and each of the <code>=</code>
characters represent the encoded quality for an individual nucleotide.
The numerical value assigned to each of these characters depends on the
sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using by Illumina version 1.8 onwards. Each character is
assigned a quality score between 0 and 40 as shown in the chart
below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI
                  |         |         |         |         |
Quality score:    0........10........20........30........40                                </code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. This quality score is logarithmically
based, so a quality score of 10 reflects a base call accuracy of 90%,
but a quality score of 20 reflects a base call accuracy of 99%. These
probability values are the results from the base calling algorithm and
dependent on how much signal was captured for the base
incorporation.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@ERR029206.1 IL31_5505:8:1:6233:1087#4/1
TCNAGTCAGCACACACATGCGAAAGAATCCACCGACTAGGGTCAGCGGGGTTTGCAGTTGGTCGCGGACGTAACCG
+
&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>We can now see that the quality of an <code>N</code> (<code>'</code>)
is 6, which corresponds to a over 25% error probability ($10^{-} $).</p>
<blockquote>
<h3 id="quality-encodings-vary">Quality Encodings Vary</h3>
<p>Although we’ve used a particular quality encoding system to
demonstrate interpretation of read quality, different sequencing
machines use different encoding systems. This means that, depending on
which sequencer you use to generate your data, a <code>#</code> may not
be an indicator of a poor quality base call.</p>
</blockquote>
<div id="callout5" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout5"></a>
</h3>
<div class="callout-content">
<p>This mainly relates to older Solexa/Illumina data, but it’s essential
that you know which sequencing platform was used to generate your data,
so that you can tell your quality control program which encoding to use.
If you choose the wrong encoding, you run the risk of throwing away good
reads or (even worse) not throwing away bad reads!</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="assessing-quality-using-fastqc">Assessing Quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a></h3>
<p>In real life, you won’t be assessing the quality of your reads by
visually inspecting your FASTQ files. Rather, you’ll be using a software
program to assess read quality and filter out poor quality reads. We’ll
first use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads.Later in our workflow, we’ll use
another program to filter out poor quality reads.</p>
<p>FastQC has a number of features which can give you a quick impression
of any problems your data may have, so you can take these issues into
consideration before moving forward with your analyses. Rather than
looking at quality scores for each individual read, FastQC looks at
quality collectively across all reads within a sample. The image below
shows a FastQC-generated plot that indicates a very high quality
sample:</p>
<figure><img src="fig/good_quality1.8.png" alt="A FastQC output showing reads with good quality" class="figure mx-auto d-block"></figure><p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. For each position, there is a box-and-whisker plot
showing the distribution of quality scores for all reads at that
position. The horizontal red line indicates the median quality score and
the yellow box shows the 2nd to 3rd quartile range. This means that 50%
of reads have a quality score that falls within the range of the yellow
box at that position. The whiskers show the range to the 1st and 4th
quartile.</p>
<p>For each position in this sample, the quality values do not drop much
lower than 32. This is a high quality score. The plot background is also
color-coded to identify good (green), acceptable (yellow), and bad (red)
quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the
spectrum.</p>
<figure><img src="fig/bad_quality1.8.png" alt="A FastQC output showing reads with bad quality" class="figure mx-auto d-block"></figure><p>Here, we see positions within the read in which the boxes span a much
wider range. Also, quality scores drop quite low into the “bad” range,
particularly on the tail end of the reads. The FastQC tool produces
several other diagnostic plots to assess sample quality, in addition to
the one plotted above.</p>
</div>
<div class="section level3">
<h3 id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a></h3>
<p>Navigate to your FASTQ dataset:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/data/</span></code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-1"></a>
</h3>
<div class="callout-content">
<p>We downloaded the samples. How many FASTQ files are in this dataset?
Why is this? What is the difference between the two files that start
with the same name? Discuss with your neighbour.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>There are 2 FASTQ files. Each sample has <a href="https://thesequencingcenter.com/knowledge-base/what-are-paired-end-reads/" class="external-link">paired-end
reads</a>.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, including files that
are compressed (<code>.gz</code>) so we can use the
<code>*.fastq.gz</code> wildcard to run FastQC on all of the compressed
FASTQ files in this directory.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="pp">*</span>.fastq.gz</span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of ERR029206_1.fastq.gz
Approx 5% complete for ERR029206_1.fastq.gz
Approx 10% complete for ERR029206_1.fastq.gz
Approx 15% complete for ERR029206_1.fastq.gz
Approx 20% complete for ERR029206_1.fastq.gz
Approx 25% complete for ERR029206_1.fastq.gz
Approx 30% complete for ERR029206_1.fastq.gz</code></pre>
</div>
<p>In total, it should take about 15-20 minutes for FastQC to run on all
four of our FASTQ files. When the analysis completes, your prompt will
return. So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 85% complete for ERR029206_2.fastq.gz
Approx 90% complete for ERR029206_2.fastq.gz
Approx 95% complete for ERR029206_2.fastq.gz
Analysis complete for ERR029206_2.fastq.gz
data $</code></pre>
</div>
<p>The FastQC program has created several new files within our
<code>/data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.copy.fastq</span></span>
<span><span class="va">ERR029206_1.fastq.gz</span></span>
<span><span class="va">ERR029206_1_fastqc.html</span></span>
<span><span class="va">ERR029206_1_fastqc.zip</span></span>
<span><span class="va">ERR029206_2.fastq.gz</span></span>
<span><span class="va">ERR029206_2_fastqc.html</span></span>
<span><span class="va">ERR029206_2_fastqc.zip</span></span></code></pre>
</div>
<p>For each input gzipped FASTQ file, FastQC has created a
<code>.zip</code> file and a <code>.html</code> file. The
<code>.zip</code> file extension indicates that this is actually a
compressed set of multiple output files. We’ll be working with these
output files soon. The <code>.html</code> file is a stable webpage
displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="ex">$</span> mkdir ../results/fastqc_untrimmed_reads</span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.zip ../results/fastqc_untrimmed_reads/</span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.html ../results/fastqc_untrimmed_reads/</span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="ex">$</span> cd ../results/fastqc_untrimmed_reads/</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="viewing-html-files">Viewing HTML files<a class="anchor" aria-label="anchor" href="#viewing-html-files"></a></h3>
<p>If we are working on our local computers, we may display each of
these HTML files as a webpage:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="ex">$</span> firefox ERR029206_1_fastqc.html</span></code></pre>
</div>
<p>You may need to replace the command <code>firefox</code> with the
name of another web navigator if you have one.</p>
<div id="exercise-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-2"></a>
</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Do the sample looks good in
terms of per base sequence quality?</p>
</div>
</div>
</div>
<div id="exercise-3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-3"></a>
</h3>
<div class="callout-content">
<p>What are the read lengths of the different samples?</p>
</div>
</div>
</div>
</div>
</section><section id="cleaning-reads"><h2 class="section-heading">Cleaning Reads<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a>
</h2>
<hr class="half-width"><p>It’s very common to have some reads within a sample, or some
positions (near the beginning or end of reads) across all reads that are
low quality and should be discarded. We will use a program called <a href="https://github.com/lh3/seqtk" class="external-link">seqtk</a> to filter poor quality
reads and trim poor quality bases from our samples.</p>
<div class="section level3">
<h3 id="seqtk-options">Seqtk Options<a class="anchor" aria-label="anchor" href="#seqtk-options"></a></h3>
<p>Seqtk is a program written C and aims to be a Swiss army knife for
sequencing reads. You don’t need to learn C to use Seqtk, but the fact
that it’s a C program helps explain the syntax that is used to run
Seqtk. Seqtk takes as input files either FASTQ files or gzipped FASTQ
files and outputs FASTQ or FASTA files. The basic command to run Seqtk
starts like this:</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="ex">$</span> seqtk</span></code></pre>
</div>
<p>That’s just the basic command, however. Seqtk has a variety of
options and parameters. We will need to specify what options we want to
use for our analysis. Here are some of the options:</p>
<table class="table"><thead><tr class="header"><th>option</th>
<th>meaning</th>
</tr></thead><tbody><tr class="odd"><td><code>seq</code></td>
<td>common transformation of FASTA/Q</td>
</tr><tr class="even"><td><code>comp</code></td>
<td>get the nucleotide composition of FASTA/Q</td>
</tr><tr class="odd"><td><code>trimfq</code></td>
<td>trim FASTQ using the Phred algorithm</td>
</tr></tbody></table><p>In addition to these options, there are a number if trimming options
available:</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="ex">$</span> seqtk trimfq</span></code></pre>
</div>
<table class="table"><thead><tr class="header"><th>step</th>
<th>meaning</th>
</tr></thead><tbody><tr class="odd"><td><code>-q</code></td>
<td>error rate threshold (disabled by -b/-e) [0.05]</td>
</tr><tr class="even"><td><code>-l</code></td>
<td>maximally trim down to INT bp (disabled by -b/-e) [30]</td>
</tr><tr class="odd"><td><code>-b</code></td>
<td>trim INT bp from left (non-zero to disable -q/-l) [0]</td>
</tr><tr class="even"><td><code>-e</code></td>
<td>trim INT bp from right (non-zero to disable -q/-l) [0].</td>
</tr></tbody></table><p>We will use only a few of these options in our analysis. It is
important to understand the steps you are using to clean your data.</p>
<p>A complete command for trimming with seqtk will look something like
this:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="ex">$</span> seqtk trimfq <span class="at">-q</span> 0.01 ERR01_1.fastq.gz <span class="op">&gt;</span> ERR01_1_trim.fastq</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="trimming">Trimming<a class="anchor" aria-label="anchor" href="#trimming"></a></h3>
<p>Now we will run seqtk trimfq on our data. To begin, navigate to your
<code>data</code> directory:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/data/</span></code></pre>
</div>
<p>We are going to run seqtk on one sample giving it an error rate
threshold of 0.01 which indicates the base call accuracy. We request
that, after trimming, the chances that a base is called incorrectly are
only 1 in 10000.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="ex">$</span> seqtk trimfq <span class="at">-q</span> 0.01 ERR029206_1.fastq.gz <span class="op">&gt;</span> ERR029206_1_trim.fastq</span></code></pre>
</div>
<p>Notice that we needed to redirect the output to a file. If we don’t
do that, the trimmed fastq data will be displayed in the console.</p>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-4"></a>
</h3>
<div class="callout-content">
<p>Use seqtk fqchk to compare the untrimmed and trimmed reads of
ERR029206_1 in terms of number of sequenced bases, percentage of A,G,C,T
and N and average quality. What do you notice? Discuss with your
neighbor.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Show me the solution</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1.copy.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 76; max_len: 76; avg_len: 76.00; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 315791552   18.0    32.1    32.2    17.7    0.0 30.6    21.4    5.4 94.6</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 69.12; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 287211920   18.2    32.0    31.9    17.8    0.0 31.7    28.8    1.3 98.7</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We’ve just successfully trimmed one of our FASTQ files! However,
there is some bad news. <code>seqtk</code> can only operate on one
sample at a time and we have more than one sample. The good news is that
we can use a <code>for</code> loop to iterate through our sample files
quickly!</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="ex">$</span> for infile in <span class="pp">*</span>.fastq.gz</span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a><span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a><span class="ex">seqtk</span> trimfq <span class="at">-q</span> 0.01 <span class="st">"</span><span class="va">${infile}</span><span class="st">"</span> <span class="op">&gt;</span>  <span class="st">"</span><span class="va">${outfile}</span><span class="st">"</span> </span>
<span id="cb47-5"><a href="#cb47-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>The new part in our <code>for</code> loop is the line:</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span></code></pre>
</div>
<p><code>infile</code> is the first variable in our loop and takes the
value of each of the FASTQ files in our directory. <code>outfile</code>
is the second variable in our loop and is defined by adding
<code>_trim.fastq</code> to the end of our input file name. Use
<code>{}</code> to wrap the variable so that <code>_trim.fastq</code>
will not be interpreted as part of the variable name. In addition,
quoting the shell variables is a good practice AND necessary if your
variables have spaces in them.</p>
<p>For more, check <a href="https://mywiki.wooledge.org/BashPitfalls" class="external-link">Bash Pitfalls</a>. There
are no spaces before or after the <code>=</code>.</p>
<p>Go ahead and run the for loop. It should take a few minutes for seqtk
to run for each of our fourteen input files. Once it’s done running,
take a look at your directory contents.</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.copy.fastq</span> </span>
<span><span class="va">ERR029206_1.fastq.gz</span></span>
<span><span class="va">ERR029206_1.fastq.gz_trim.fastq</span> </span>
<span><span class="va">ERR029206_1_trim.fastq</span> </span>
<span><span class="va">ERR029206_2.fastq.gz</span></span>
<span><span class="va">ERR029206_2.fastq.gz_trim.fastq</span></span>
<span><span class="va">GCF_000195955.2_ASM19595v2_genomic.fna</span></span></code></pre>
</div>
<p>We’ve now completed the trimming and filtering steps of our quality
control process! Before we move on, let’s move our trimmed FASTQ files
to a new subdirectory within our <code>data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="ex">$</span> mkdir trimmed_fastq</span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>fastq.gz_trim<span class="pp">*</span> trimmed_fastq</span>
<span id="cb51-3"><a href="#cb51-3" tabindex="-1"></a><span class="ex">$</span> cd trimmed_fastq</span>
<span id="cb51-4"><a href="#cb51-4" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.fastq.gz_trim.fastq</span> </span>
<span><span class="va">ERR029206_2.fastq.gz_trim.fastq</span> </span></code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<p>Again, use seqtk fqchk to compare the untrimmed and trimmed reads of
both samples. Note the number of bases ‘#bases’ of the trimmed and
untrimmed reads. Calculate the theoretical coverage of the genomes
before and after trimming, assuming that all our genomes do have the
same size as our reference genome (4411532 bases).</p>
<p>Hint: Sum up forward and reverse reads!</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">Show me the solution</h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1.fastq.gz_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 69.12; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 287211920   18.2    32.0    31.9    17.8    0.0 31.7    28.8    1.3 98.7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_2.fastq.gz_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 63.79; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 265048852   18.1    32.1    32.1    17.7    0.0 31.2    25.7    2.5 97.5</code></pre>
</div>
<p>Coverage = #bases (forward + reverse) / genome size In this case:
125.18 = (287211920 + 265048852) / 4411532</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>
<code>wget</code> is a computer program to get data from the
internet</li>
<li>
<code>for</code> loops let you perform the same set of operations on
multiple files with a single command</li>
<li>Sequencing data is large</li>
<li>In bioinformatic workflows the output of one tool is the input of
the other.</li>
<li>FastQC is used to judge the quality of sequencing reads.</li>
<li>Data cleaning is an essential step in a genomics pipeline.</li>
<li>We will work towards confirming or disputing transmission in TB
cases</li>
<li>After this practical training you will have some familiarity with
working on the command line</li>
</ul></div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="phylogenetics.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="sequence-assembly.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="phylogenetics.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Phylogenetics
        </a>
        <a class="chapter-link float-end" href="sequence-assembly.html" rel="next">
          Next: Sequence assembly... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/quality-control.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:lionel.guy@imbim.uu.se">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper" class="external-link">sandpaper (0.16.4.9000)</a>, <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.7.4.9000)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.1" class="external-link">varnish (1.0.1)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries.github.io/workbench-template-rmd/quality-control.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "evolution, phylogenetics, molecular epidemiology",
  "name": "Reads, QC and trimming",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/quality-control.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/quality-control.html",
  "dateCreated": "2023-05-08",
  "dateModified": "2024-03-15",
  "datePublished": "2024-04-02"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

