<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Microbial Genomics: evolution, phylogenetics, and molecular epidemiology: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="manifest" href="../site.webmanifest">
<link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav data"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Data Carpentry" src="../assets/images/data-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="../assets/images/data-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="genetic-drift.html">2. Genetic drift</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="msa.html">3. Multiple sequence alignment</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="phylogenetics.html">4. Phylogenetics - UPGMA</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="phylogenetics-advanced.html">5. Phylogenetics - Neighbor-joining and Maximum likelihood</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="molecular-epidemiology.html">6. Molecular Epidemiology</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="downloading-data.html">7. Downloading data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="quality-control.html">8. Assessing Read Quality and Trimming</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="sequence-assembly.html">9. Sequence assembly</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="snps-phylogeny.html">10. SNP phylogeny</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="data-visualization.html">11. Data Visualization</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush13">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading13">
        <a href="conclusions.html">12. Transmission?</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="../instructor/aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Introduction</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 30 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How does evolution work (in microbes)?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Test your prior knowledge about evolutionary biology.</li>
<li>Discuss a few “difficult” questions about evolution, with a specific
focus on how microbes evolve.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="prior-knowledge"><h2 class="section-heading">Prior knowledge<a class="anchor" aria-label="anchor" href="#prior-knowledge"></a>
</h2>
<hr class="half-width">
<p>This is an ice-breaking exercise. Don’t panic.</p>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Instructor Note
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<p>Let the students discuss two and two for 4-5 minutes. Encourage them
to write keywords on the board.</p>
</div>
</div>
</div>
</div>
<div id="what-does-evolution-need-to-work" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="what-does-evolution-need-to-work" class="callout-inner">
<h3 class="callout-title">What does evolution need to work<a class="anchor" aria-label="anchor" href="#what-does-evolution-need-to-work"></a>
</h3>
<div class="callout-content">
<ul>
<li>Discuss with your neighbor for 5 minutes</li>
<li>Write up to three keywords on the board</li>
</ul>
</div>
</div>
</div>
<div id="quiz-about-evolution" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="quiz-about-evolution" class="callout-inner">
<h3 class="callout-title">Quiz about evolution<a class="anchor" aria-label="anchor" href="#quiz-about-evolution"></a>
</h3>
<div class="callout-content">
<p>Answer the quiz as you follow along, <a href="https://pollev.com/lionelguy323/" class="external-link">using this link</a> or by using
the following QR code <img src="../fig/qr_code_quiz.png" alt="QR code linking to the interactive quiz" class="figure"></p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>The answers will be discussed in the presentation later.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Learn the very basic principle of evolution.</li>
<li>Evolution can be counter-intuitive.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-genetic-drift"><p>Content from <a href="genetic-drift.html">Genetic drift</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/genetic-drift.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How does random genetic drift influence fixation of alleles?</li>
<li>What is the influence of population size?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how population size influences the probability of
fixation of an allele.</li>
<li>Understand how slightly deleterious mutations may get fixed in small
populations.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>This exercise is to illustrate the concepts of selection, population
size and genetic drift, using simulations. We will use mostly <a href="https://cartwrig.ht/apps/redlynx/" class="external-link">Red Lynx</a> by <a href="https://github.com/reedacartwright" class="external-link">Reed A. Cartwright</a>.</p>
<p>Another option is to use a <a href="https://phytools.shinyapps.io/drift-selection/" class="external-link">web interface</a>
to the <a href="https://www.r-project.org/" class="external-link">R</a> <a href="https://github.com/liamrevell/learnPopGen" class="external-link">learnPopGen</a>
package, but the last one is mostly for biallelic genes (and thus not
that relevant for bacteria).</p>
<p>Open now the <a href="https://cartwrig.ht/apps/redlynx/" class="external-link">Red Lynx</a>
website and get familiar with the different options.</p>
<p>You won’t need the right panel (but feel free to explore). The
dominance option in the left panel won’t be used either.</p>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>In the first part, you will only play with the number of generations,
the initial frequency and the population size.</p>
<ul>
<li>Lower the number of generations to 200.</li>
<li>Adjust the population size to 1000.</li>
<li>Make sure the intial frequency is set to 50%.</li>
<li>Run a few simulations (ca 20).</li>
</ul>
<p>Did any allele got fixed? What is the range of frequencies after 200
generations?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>In my simulations, no allele got fixed, the final allele frequencies
range 20-80%</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>
(continued)</em><a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>Now increase the population to 100’000, clear the graph, and repeat
the simulations. What’s the range of final frequencies now?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>In my simulations, no allele got fixed, and the final allele
frequencies range 45-55%</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>
(continued)</em><a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>Now decrease the population to 10 individuals, clear the graph and
repeat these simulations. What’s the range of final frequencies now?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>In my simulations, one allele got fixed quickly, the latest one was
at generation 100.</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>
(continued)</em><a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>What do you conclude here?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>It is clear that stochastic (random) variation in allele frequencies
strongly affects the probability of fixation of alleles in small
populations, not so much in large ones.</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-with-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-with-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift with selection<a class="anchor" aria-label="anchor" href="#genetic-drift-with-selection"></a>
</h3>
<div class="callout-content">
<p>So far we’ve only looked at how allele frequencies vary in the
absence of selection, that is when the two alleles provide an equal
fitness. What’s the influence of random genetic drift when alleles are
not neutral?</p>
<p>The selection strength is equivalent to the selection coefficient,
i.e. how much higher relative fitness the new allele provides. A
selection coefficient of 0.01 means that the organism harboring the new
allele has a 1% increased fitness.</p>
<ul>
<li>Increase the number of generations to 1000.</li>
<li>Set the selection strength to 0.01 (1%).</li>
<li>Set the population size first to 100’000, then to 1000, then to
100.</li>
</ul>
<p>How long does it take for the allele to get fixed - in average - with
the three population sizes?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>About the same time, but the trajectories are much smoother with
larger populations. In the small population, it happens that the
beneficial allele disappears from the population (although not
often).</p>
</div>
</div>
</div>
</div>
<div id="fixation-of-slightly-deleterious-alleles-in-very-small-populations" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="fixation-of-slightly-deleterious-alleles-in-very-small-populations" class="callout-inner">
<h3 class="callout-title">Fixation of slightly deleterious alleles in
very small populations<a class="anchor" aria-label="anchor" href="#fixation-of-slightly-deleterious-alleles-in-very-small-populations"></a>
</h3>
<div class="callout-content">
<p>We’re now simulating what would happen in a very small population (or
a population that undergoes very narrow bottlenecks), when a gene
mutates. We’ll have a very small population (10 individuals), a
selection value of -0.01 (the mutated allele provides a 1% lower
fitness), and a 10% initial frequency, which corresponds to one
individual getting a mutation:</p>
<ul>
<li>Set the population to 10 individuals.</li>
<li>Set generations to 200.</li>
<li>Set initial frequency to 10%.</li>
<li>Set the selection strength to -0.01.</li>
</ul>
<p>Run many simulations. What happens?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<p>Most new alleles go extinct quickly. In my simulations, I usually get
one of the slightly deleterious mutations fixed after 20
simulations.</p>
</div>
</div>
</div>
</div>
<p>That’s it for that part, but feel free to continue playing with the
different settings later.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Random genetic drift has a large influence on the probability of
fixation of alleles in small populations, even for non-neutral
alleles.</li>
<li>Random genetic drift has very little influence of the probability of
fixation of alleles in large populations.</li>
<li>Slightly deleterious mutations can get fixed into the population
through random genetic drift, if the population is small enough and the
selective value is not too large.</li>
</ul>
</div>
</div>
</div>
<!-- References -->
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-msa"><p>Content from <a href="msa.html">Multiple sequence alignment</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/msa.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 40 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you align multiple sequences?</li>
<li>Why is it important to properly align sequences?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use <code>mafft</code> to align multiple sequences.</li>
<li>Test different algorithms.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In this short episode, we are exploring multiple sequence alignment
(MSA). We are going to use <a href="https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html" class="external-link"><code>mafft</code></a>
to align homologs of <a href="https://en.wikipedia.org/wiki/RpoB" class="external-link">RpoB</a>, the β subunit of the
bacterial RNA polymerase. It is a long, multi-domain protein, suitable
for showing issues related to MSA.</p>
<p>First, go to your home folder and create a phylogenetics folder</p>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>Go to your home, use the <code>mkdir</code> command to make a
<code>phylogenetics</code> subfolder, and move into it:</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir phylogenetics</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Second, download a <a href="../files/rpoB.fasta">fasta file</a>
containing 19 sequences of RpoB, gathered by aligning RpoB from <em>E.
coli</em> to a very small database present at NCBI, the <a href="https://blast.ncbi.nlm.nih.gov/smartblast/smartBlast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs" class="external-link">landmark</a>
database. Move that file to your newly created
<code>~/phylogenetics/</code> folder.</p>
</section><section id="renaming-sequences"><h2 class="section-heading">Renaming sequences<a class="anchor" aria-label="anchor" href="#renaming-sequences"></a>
</h2>
<hr class="half-width">
<p>Look at the accession ids of the fasta sequences: they are not very
informative:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> grep <span class="st">'&gt;'</span> rpoB.fasta <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;WP_000263098.1 MULTISPECIES: DNA-directed RNA polymerase subunit beta [Enterobacteriaceae]
&gt;WP_011070610.1 DNA-directed RNA polymerase subunit beta [Shewanella oneidensis]
&gt;WP_003114335.1 DNA-directed RNA polymerase subunit beta [Pseudomonas aeruginosa]
&gt;WP_002228870.1 DNA-directed RNA polymerase subunit beta [Neisseria meningitidis]
&gt;WP_003436174.1 MULTISPECIES: DNA-directed RNA polymerase subunit beta [Eubacteriales]</code></pre>
</div>
<p>Keep the taxonomic name following the accession id, separated by
<code>_</code>, using a bit of <code>sed</code> magic. Save the
resulting file for later use, and show the headers again.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> sed <span class="st">"/^&gt;/ s/&gt;\([^ ]*\) \([^\[]*\)\[\(.*\)\]/&gt;</span><span class="dt">\\</span><span class="st">3_</span><span class="dt">\\</span><span class="st">1/"</span>  rpoB.fasta <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/ /_/g"</span> <span class="op">&gt;</span> rpoB_renamed.fasta</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> grep <span class="st">'&gt;'</span> rpoB_renamed.fasta <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;Enterobacteriaceae_WP_000263098.1
&gt;Shewanella_oneidensis_WP_011070610.1
&gt;Pseudomonas_aeruginosa_WP_003114335.1
&gt;Neisseria_meningitidis_WP_002228870.1
&gt;Eubacteriales_WP_003436174.1</code></pre>
</div>
<p>It looks better.</p>
</section><section id="alignment-with-progressive-algorithm"><h2 class="section-heading">Alignment with progressive algorithm<a class="anchor" aria-label="anchor" href="#alignment-with-progressive-algorithm"></a>
</h2>
<hr class="half-width">
<p>Use <code>mafft</code> with a progressive algorithm to align the
sequences.</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>Use the FFT-NS-2 algorithm from mafft to align the renamed sequences.
Also, record the time it takes for <code>mafft</code> to complete the
task.</p>
<p>Hint: type <code>mafft</code> and try tab-complete to see all
versions of <code>mafft</code>.</p>
<p>Hint: try the command <code>time</code></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> time mafft-fftns rpoB_renamed.fasta <span class="op">&gt;</span> rpoB.fftns.aln</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Strategy:
 FFT-NS-2 (Fast but rough)
 Progressive method (guide trees were built 2 times.)

If unsure which option to use, try 'mafft --auto input &gt; output'.
For more information, see 'mafft --help', 'mafft --man' and the mafft page.

The default gap scoring scheme has been changed in version 7.110 (2013 Oct).
It tends to insert more gaps into gap-rich regions than previous versions.
To disable this change, add the --leavegappyregion option.

mafft-fftns rpoB_renamed.fasta &gt; rpoB.fftns.aln  1.02s user 0.52s system 104% cpu 1.471 total</code></pre>
</div>
<p>The last line is the output of the <code>time</code> command. It took
1.47 seconds to complete this time.</p>
</div>
</div>
</div>
</div>
</section><section id="alignment-with-iterative-algorithm"><h2 class="section-heading">Alignment with iterative algorithm<a class="anchor" aria-label="anchor" href="#alignment-with-iterative-algorithm"></a>
</h2>
<hr class="half-width">
<p>Now use one of the supposedly better iterative algorithm of
<code>mafft</code> to align the same sequences. Choose the <a href="https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html" class="external-link">E-INS-i
algorithm</a> which is suited for proteins that have highly conserved
motifs interspersed with less conserved ones.</p>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge3"></a>
</h3>
<div class="callout-content">
<p>Use the E-INS-i algorithm from mafft to align the renamed sequences.
Also, record the time it takes for <code>mafft</code> to complete the
task.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> time mafft-einsi rpoB_renamed.fasta <span class="op">&gt;</span> rpoB.einsi.aln</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Strategy:
 E-INS-i (Suitable for sequences with long unalignable regions, very slow)
 Iterative refinement method (&lt;16) with LOCAL pairwise alignment with generalized affine gap costs (Altschul 1998)

If unsure which option to use, try 'mafft --auto input &gt; output'.
For more information, see 'mafft --help', 'mafft --man' and the mafft page.

The default gap scoring scheme has been changed in version 7.110 (2013 Oct).
It tends to insert more gaps into gap-rich regions than previous versions.
To disable this change, add the --leavegappyregion option.

Parameters for the E-INS-i option have been changed in version 7.243 (2015 Jun).
To switch to the old parameters, use --oldgenafpair, instead of --genafpair.

mafft-einsi rpoB_renamed.fasta &gt; rpoB.einsi.aln  6.84s user 0.65s system 99% cpu 7.542 total</code></pre>
</div>
<p>It now took 7.54 seconds to complete this time, i.e. about 5 times
slower than with the progressive algorithm. It doesn’t make a big
difference now, but with hundreds of sequences it will make one.</p>
</div>
</div>
</div>
</div>
</section><section id="alignment-visualization"><h2 class="section-heading">Alignment visualization<a class="anchor" aria-label="anchor" href="#alignment-visualization"></a>
</h2>
<hr class="half-width">
<p>We want to compare the two alignments. Use <code>seaview</code>:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> seaview rpoB.fftns.aln <span class="kw">&amp;</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> seaview rpoB.einsi.aln <span class="kw">&amp;</span></span></code></pre>
</div>
<p>Arrange the two windows on top of each other. Change the fontsize
(Props -&gt; Fontsize) to 8 to see a larger portion of the
alignment.</p>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge4"></a>
</h3>
<div class="callout-content">
<p>Can you spot differences? Which alignment is longer?</p>
<p>Hint: try to scroll to position 800-900.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<figure><img src="../fig/seaview.png" alt="Alignments shown in seaview" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<p>If you have time over, spend it exploring the different options of
Seaview.</p>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-phylogenetics"><p>Content from <a href="phylogenetics.html">Phylogenetics - UPGMA</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/phylogenetics.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 30 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you build a phylogenetic tree?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn the basic of phylogenetics tree building, taking the simplest
of the examples with a UPGMA tree.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="setup"><h2 class="section-heading">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<hr class="half-width">
<p>The exercise is done for a large part with pen and paper, and then a
demonstration in R. We’ll also use the R package <a href="https://emmanuelparadis.github.io/" class="external-link"><code>ape</code></a> , which
you should install if it’s not present on your setup:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">'ape'</span><span class="op">)</span></span></code></pre>
</div>
<p>And to load it:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="upgma-based-tree"><h2 class="section-heading">UPGMA-based tree<a class="anchor" aria-label="anchor" href="#upgma-based-tree"></a>
</h2>
<hr class="half-width">
<p>Load the tree in fasta format, reading from a <code>temp</code>
file</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FASTAfile</span> <span class="op">&lt;-</span> <span class="fu">tempfile</span><span class="op">(</span><span class="st">"aln"</span>, fileext <span class="op">=</span> <span class="st">".fasta"</span><span class="op">)</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"&gt;A"</span>, <span class="st">"----ATCCGCTGATCGGCTG----"</span>,</span>
<span>    <span class="st">"&gt;B"</span>, <span class="st">"GCTGATCCGTTGATCGG-------"</span>,</span>
<span>    <span class="st">"&gt;C"</span>, <span class="st">"----ATCTGCTCATCGGCT-----"</span>,</span>
<span>    <span class="st">"&gt;D"</span>, <span class="st">"----ATTCGCTGAACTGCTGGCTG"</span>,</span>
<span>    file <span class="op">=</span> <span class="va">FASTAfile</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu">read.dna</span><span class="op">(</span><span class="va">FASTAfile</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre>
</div>
<p>Now look at the alignment. Notice there are gaps, which we don’t want
in this example. We also remove the non-informative (identical)
columns.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">alview</span><span class="op">(</span><span class="va">aln</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  000000000111111111122222
  123456789012345678901234
A ----ATCCGCTGATCGGCTG----
B GCTG.....T.......---....
C .......T...C.......-....
D ......T......A.T....GCTG</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aln_filtered</span> <span class="op">&lt;-</span> <span class="va">aln</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">7</span>,<span class="fl">8</span>,<span class="fl">10</span>,<span class="fl">12</span>,<span class="fl">14</span>,<span class="fl">16</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">alview</span><span class="op">(</span><span class="va">aln_filtered</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  123456
A CCCGTG
B ..T...
C .T.C..
D T...AT</code></pre>
</div>
<p>Now we have a simple alignment as in the lecture. Dots
(<code>.</code>) mean that the sequence is identical to the top row,
which makes it easier to calculate distances.</p>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<a class="anchor" aria-label="anchor" href="#calculating-distance-by-hand"></a>
</h3>
<div class="callout-content">
<p>Let’s use a matrix to calculate distances between sequences. We’ll
just count the number of differences between the sequences. We’ll then
group the two closest sequences. Which are they?</p>
<table class="table">
<caption>Distances between the sequences.</caption>
<thead><tr class="header">
<th></th>
<th align="right">A</th>
<th align="right">B</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>B</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>C</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Here is the solution:</p>
<table class="table">
<caption>Distances between the sequences.</caption>
<thead><tr class="header">
<th></th>
<th align="right">A</th>
<th align="right">B</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>B</td>
<td align="right">1</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>C</td>
<td align="right">2</td>
<td align="right">3</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right"></td>
</tr>
</tbody>
</table>
<p>The two closest sequences are A and B.</p>
</div>
</div>
</div>
</div>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<em>
(continued)</em><a class="anchor" aria-label="anchor" href="#calculating-distance-by-hand"></a>
</h3>
<div class="callout-content">
<p>Let’s now cluster together A and B, and calculate the average
distance from AB to the other sequences, weighted by the size of the
clusters.</p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">AB</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AB</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>C</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>D</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>The average distance for AB to C is calculated as follow:</p>
<p><span class="math inline">\(d(AB,C) = \dfrac{d(A,C) + d(B,C)}{2} =
\dfrac{2 + 3}{2} = 2.5\)</span></p>
<p>And so on for the other distances:</p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">AB</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AB</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>C</td>
<td align="right">2.5</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>D</td>
<td align="right">3.5</td>
<td align="right">5</td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<em>
(continued)</em><a class="anchor" aria-label="anchor" href="#calculating-distance-by-hand"></a>
</h3>
<div class="callout-content">
<p>Now the shortest distance is AB to C. Let’s recalculate the distance
to D again.</p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">ABC</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ABC</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p><span class="math inline">\(d(ABC,D) = \dfrac{d(AB,D) * 2 + d(C,D) *
1}{2 + 1} = \dfrac{3.5 * 2 + 5 * 1}{3} = 4\)</span></p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">ABC</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ABC</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right">4</td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<p>The tree is reconstructed by dividing the distances equally between
the two leaves. - A-B: each 0.5. - AB-C: each side gets 2.5/2 = 1.25.
The branch to AB is 1.25 - 0.75 = 0.75 - ABC-D: each side gets 4/2 = 2.
The branch to ABC is 2 - 0.75 - 0.5 = 0.75</p>
<figure><img src="../fig/upgma_manual.png" alt="Manually built UPGMA tree" class="figure mx-auto d-block"><figcaption>UPGMA tree</figcaption></figure><p>Let’s know do the same using bioinformatics tools.</p>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge4"></a>
</h3>
<div class="callout-content">
<p>We’ll use <code>dist.dna</code> to calculate the distances. We’ll use
a “N” model, that just counts the differences and doesn’t correct or
normalizes. We’ll use the function <code>hclust</code> to perform the
UPGMA method calculation. The tree is then plotted, and the branch
lengths plotted with <code>edgelabels</code>:</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist_matrix</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">aln_filtered</span>, model<span class="op">=</span><span class="st">"N"</span><span class="op">)</span></span>
<span><span class="va">dist_matrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  A B C
B 1    
C 2 3  
D 3 4 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">as.phylo</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="va">dist_matrix</span>, <span class="st">"average"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="fu">edgelabels</span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">edge.length</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/phylogenetics-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>The simplest of the trees are distance-based.</li>
<li>UPGMA works by clustering the two closest leaves and recalculating
the distance matrix.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-phylogenetics-advanced"><p>Content from <a href="phylogenetics-advanced.html">Phylogenetics - Neighbor-joining and Maximum likelihood</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/phylogenetics-advanced.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you build a phylogenetic tree with more advanced
methods?</li>
<li>How do you ascertain statistical support for phylogenetic
trees?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn how to use phylogenetics algorithms, neighbor-joining and
maximum-likelihood.</li>
<li>Learn how to perform and show bootstraps.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In the previous episode, we inferred a simple phylogenetic tree using
UPGMA, without correcting the distance matrix for multiple
substitutions. UPGMA has many shortcomings, and gets worse as distance
increases. Here we’ll test Neighbor-Joining, which is also a
distance-based, and a maximum-likelihood based approach with <a href="http://www.iqtree.org/" class="external-link">IQ-TREE</a>.</p>
</section><section id="neighbor-joining"><h2 class="section-heading">Neighbor-joining<a class="anchor" aria-label="anchor" href="#neighbor-joining"></a>
</h2>
<hr class="half-width">
<p>The principle of <a href="https://en.wikipedia.org/wiki/Neighbor_joining" class="external-link">Neighbor-joining
method</a> (NJ) is to start from a star-like tree, find the two branches
that, if joined, minimize the total tree length, join then, and repeat
with the joined branches and the rest of the branches.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>To perform NJ on our sequences, we’ll use the function inbuilt in
Seaview. First (re-)open the alignment we obtained from
<code>mafft</code> with the E-INS-i method.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seaview</span> rpoB.einsi.aln <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<em> (continued)</em><a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>In the Seaview window, select Trees -&gt; Distance methods. Keep the
BioNJ option ticked. BioNJ is an updated version of NJ, more accurate
for longer distances. Tick the “Bootstrap” option and leave it at 100.
We’ll discuss these later. Then click on “Go”.</p>
</div>
</div>
</div>
<p>What do you see? Is the tree rooted? Is the pattern species coherent
with what you know of the tree of life? Any weird results?</p>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>Redo the BioNJ tree for the other alignment we inferred.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seaview</span> rpoB.fftns.aln <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Do you see any differences between the two trees? What can you make
out of it?</p>
</section><section id="maximum-likelihood"><h2 class="section-heading">Maximum likelihood<a class="anchor" aria-label="anchor" href="#maximum-likelihood"></a>
</h2>
<hr class="half-width">
<p>We will now use IQ-TREE to infer a maximum-likelihood tree of the
RpoB dataset we aligned with <code>mafft</code> previously.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> iqtree <span class="at">-s</span> rpoB.einsi.aln <span class="at">-m</span> TEST <span class="at">-B</span> 1000</span></code></pre>
</div>
<p>Here, we tell IQ-TREE to use the alignment
<code>rpoB.einsi.aln</code>, and to test among the standard substitution
models which one fits best (<code>-m TEST</code>). We also tell IQ-TREE
to perform 1000 ultra-fast bootstraps (<code>-B 1000</code>). We’ll
discuss these later.</p>
<p>IQ-TREE is a very complete program that can do a large variety of
phylogenetic analysis. To get a flavor of what it’s capable of, look at
its <a href="http://www.iqtree.org/doc/" class="external-link">extensive
documentation</a>.</p>
<p>Have a look at the output files:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> rpoB.einsi.aln.<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>rpoB.einsi.aln.bionj      rpoB.einsi.aln.iqtree     rpoB.einsi.aln.model.gz
rpoB.einsi.aln.ckp.gz     rpoB.einsi.aln.log        rpoB.einsi.aln.splits.nex
rpoB.einsi.aln.contree    rpoB.einsi.aln.mldist     rpoB.einsi.aln.treefile</code></pre>
</div>
<p>There are two important files:<br>
* <code>*.iqtree</code> file provides a text summary of the analysis. *
<code>*.treefile</code> is the resulting tree in <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick
format</a>.</p>
<p>Now display the resulting tree in FigTree.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> figtree rpoB.einsi.aln.treefile <span class="kw">&amp;</span></span></code></pre>
</div>
<p>A window will pop-up and ask to provide a name for branch/node
labels. Use e.g. ‘bootstraps’.</p>
<p>The tree appears as unrooted. It is good practice to start by
ordering the nodes and root it. Since we don’t really know otherwise
where the root is, we’ll do a mid-point rooting to start with.</p>
<p>In the left menu, develop the ‘Trees’ menu. Tick the ‘Root tree’ and
select ‘Midpoint’ from the drop-down menu. Tick the ‘Order nodes’ box
and keep the ‘increasing’ option selected.</p>
<p>Scrutinize the tree. Is it different from the BioNJ tree generated in
Seaview? How?</p>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge3"></a>
</h3>
<div class="callout-content">
<p>Redo a ML tree from the other alignment (FFT-NS) we inferred with
<code>mafft</code> and display the resulting tree in FigTree</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> iqtree <span class="at">-s</span> rpoB.fftns.aln <span class="at">-m</span> TEST <span class="at">-B</span> 1000</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> figtree rpoB.fftns.aln.treefile <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="bootstraps"><h2 class="section-heading">Bootstraps<a class="anchor" aria-label="anchor" href="#bootstraps"></a>
</h2>
<hr class="half-width">
<p>We have inferred four trees: * Two based on the alignment generated
from the E-INS-i algorithm, two from the FFT-NS. * Two inferred with the
BioNJ algorithm and two with the ML algorithm (IQ-Tree)</p>
<p>Along the way, we’ve generated bootstraps for all our trees. Now show
them on all four trees. * For the Seaview trees, tick the ‘Br support’
box * For the trees shown in FigTree, develop the ‘Branch labels’ menu,
tick the corresponding box and select ‘bootstraps’ from the ‘Display’
drop-down menu.</p>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge4"></a>
</h3>
<div class="callout-content">
<p>Compare all four trees. Do you find any significant differences?</p>
<p>Hint: what are <em>Glycine</em> and <em>Arabidopsis</em>? What about
<em>Synechocystis</em> and <em>Microcystis</em>?</p>
<p>Hint: search for the accession number of the <em>Glycine</em> RpoB
sequence on <a href="https://www.ncbi.nlm.nih.gov/search/" class="external-link">NCBI</a>. Any
hint there?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>The two plant RpoB sequences are from chloroplastic genomes, and thus
branch together with the cyanobacterial sequences, in some of the
phylogenies at least.</p>
</div>
</div>
</div>
</div>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<em> (continued)</em><a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>What about the support values for grouping these two groups? How high
are they?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Neighbor-joining is distance-based and fast, but not necessarily
very accurate</li>
<li>Maximum-likelihood is slower, but more accurate</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-molecular-epidemiology"><p>Content from <a href="molecular-epidemiology.html">Molecular Epidemiology</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/molecular-epidemiology.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 12 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you use genomics and phylogenetics to trace a microbial
outbreak?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Trace an outbreak of Mycobacterium, from reads to phylogenetic
tree.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>This part of the course is largely inspired from <a href="https://github.com/aschuerch/MolecularEpidemiology_AnalysisWGS" class="external-link">another
tutorial</a> developed by <a href="https://aschuerch.github.io/" class="external-link">Anita
Schürch</a>. We will use (part of) the data described in <a href="https://bmcinfectdis.biomedcentral.com/articles/10.1186/1471-2334-13-110" class="external-link">Bryant
et al., 2013</a>.</p>
</section><section id="background-of-the-data"><h2 class="section-heading">Background of the data<a class="anchor" aria-label="anchor" href="#background-of-the-data"></a>
</h2>
<hr class="half-width">
<p>Epidemiological contact tracing through interviewing of patients can
identify potential chains of patients that transmitted an infectious
disease to each other. Contact tracing was performed following the <a href="https://www.ncbi.nlm.nih.gov/pubmed/1643300" class="external-link">stone-in-the-pond
principle</a>, which interviews and tests potential contacts in
concentric circles around a potential source case.</p>
<p>Tuberculosis (TB) is an infectious disease caused by
<em>Mycobacterium tuberculosis</em>. It mostly affects the lungs. An
infection with <em>M. tuberculosis</em> is often asymptomatic (latent
infection). Only in about 10% of the cases the latent infection
progresses to an active infection during a patients lifetime, which, if
untreated, leads to death in about half of the cases. The symptoms of an
active TB infection include cough, fever, night sweats, weight loss etc.
An active TB infection can spread. Once exposed, people often have
latent TB. To identify people with latent TB, a <a href="https://www.cdc.gov/tb/publications/factsheets/testing/skintesting.htm" class="external-link">skin
test</a> can be applied.</p>
<p>Here we have 7 tuberculosis patients with active TB, that form three
separate clusters of potential transmission as determined by
epidemiological interviews. Patients were asked if they have been in
direct contact with each other, or if they visited the same localities.
From all patients, a bacterial isolate was grown, DNA isolated, and
whole-genome sequenced on an Illumina sequencer.</p>
<p>The three clusters consist of:</p>
<ul>
<li>Cluster 1
<ul>
<li>Patient A1 (2011) - sample ERR029207</li>
<li>Patient A2 (2011) - sample ERR029206</li>
<li>Patient A3 (2008) - sample ERR026478</li>
</ul>
</li>
<li>Cluster 2
<ul>
<li>Patient B1 (2001) - sample ERR026474</li>
<li>Patient B2 (2012) - sample ERR026473</li>
</ul>
</li>
<li>Cluster 3
<ul>
<li>Patient C1 (2011) - sample ERR026481</li>
<li>Patient C2 (2016) - sample ERR026482</li>
</ul>
</li>
</ul>
<p>The second goal of this lab is to affirm or dispute the inferred
transmissions by comparing the bacterial genomes with each other. We
will do that by identifying SNPs between the genomes.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>We will work towards confirming or disputing transmission in TB
cases</li>
<li>After this practical training you will have some familiarity with
working on the command line</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-downloading-data"><p>Content from <a href="downloading-data.html">Downloading data</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/downloading-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 60 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>This episode describes how to retrieve all necessary data from public
repositories - the raw sequenced data of our isolates and a reference
genome. It also introduces for loops which we will use for the rest of
our analysis.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I organize my file system for a new bioinformatics
project?</li>
<li>How and where can data be downloaded?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create a file system for a bioinformatics project.</li>
<li>Download files necessary for further analysis.</li>
<li>Use ‘for’ loops to automate operations on multiple files</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="getting-your-project-started"><h2 class="section-heading">Getting your project started<a class="anchor" aria-label="anchor" href="#getting-your-project-started"></a>
</h2>
<hr class="half-width">
<p>Project organization is one of the most important parts of a
sequencing project, and yet is often overlooked amidst the excitement of
getting a first look at new data. Of course, while it’s best to get
yourself organized before you even begin your analyses,it’s never too
late to start, either.</p>
<p>Genomics projects can quickly accumulate hundreds of files across
tens of folders. Every computational analysis you perform over the
course of your project is going to create many files, which can
especially become a problem when you’ll inevitably want to run some of
those analyses again. For instance, you might have made significant
headway into your project, but then have to remember the PCR conditions
you used to create your sequencing library months prior.</p>
<p>Other questions might arise along the way:</p>
<ul>
<li>What were your best alignment results?</li>
<li>Which folder were they in: Analysis1, AnalysisRedone, or
AnalysisRedone2?</li>
<li>Which quality cutoff did you use?</li>
<li>What version of a given program did you implement your analysis
in?</li>
</ul>
<p>In this exercise we will setup a file system for the project we will
be working on during this workshop.</p>
<p>We will start by creating a directory that we can use for the rest of
the workshop. First navigate to your home directory. Then confirm that
you are in the correct directory using the <code>pwd</code> command.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> pwd</span></code></pre>
</div>
<p>You should see the output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>/home/chewbe</code></pre>
</div>
<div id="tip" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip" class="callout-inner">
<h3 class="callout-title">Tip<a class="anchor" aria-label="anchor" href="#tip"></a>
</h3>
<div class="callout-content">
<p>If you aren’t in your home directory, the easiest way to get there is
to enter the command <code>cd</code>, which always returns you to
home.</p>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>Use the <code>mkdir</code> command to make the following
directories:</p>
<pre><code><span><span class="va">molepi</span></span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">docs</span>  </span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">data</span>  </span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">results</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir molepi</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir molepi/docs</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir molepi/data</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir molepi/results</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Use <code>ls -R</code> to verify that you have created these
directories. The <code>-R</code> option for <code>ls</code> stands for
recursive. This option causes <code>ls</code> to return the contents of
each subdirectory within the directory iteratively.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-R</span> molepi</span></code></pre>
</div>
<p>You should see the following output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>molepi/:
data  docs  results

molepi/data:

molepi/docs:

molepi/results: </code></pre>
</div>
</section><section id="selection-of-a-reference-genome"><h2 class="section-heading">Selection of a reference genome<a class="anchor" aria-label="anchor" href="#selection-of-a-reference-genome"></a>
</h2>
<hr class="half-width">
<p>Reference sequences (including many pathogen genomes) are available
at <a href="https://www.ncbi.nlm.nih.gov/refseq/" class="external-link">NCBI’s refseq
database</a></p>
<p>A reference genome is a genome that was previously sequenced and is
closely related to the isolates we would like to analyse. The selection
of a closely related reference genome is not trivial and will warrant an
analysis in itself. However, for simplicity, here we will work with the
<em>M. tuberculosis</em> reference genome H37Rv.</p>
<div class="section level3">
<h3 id="download-reference-genomes-from-ncbi">Download reference genomes from NCBI<a class="anchor" aria-label="anchor" href="#download-reference-genomes-from-ncbi"></a>
</h3>
<p>Download the <em>M.tuberculosis</em> reference genome from the NCBI
ftp site.</p>
<p>First, we switch to the data folder to store all our data</p>
<pre class="source"><code>$ cd molepi/data </code></pre>
<p>The reference genome will be downloaded programmatically from NCBI
with <code>wget</code>. <code>wget</code> is a computer program that
retrieves content from web servers. Its name derives from World Wide Web
and get.</p>
<pre class="source"><code>$ wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/195/955/GCF_000195955.2_ASM19595v2/GCF_000195955.2_ASM19595v2_genomic.fna.gz</code></pre>
<p>This file is compressed as indicated by the extension of
<code>.gz</code>. It means that this file has been compressed using the
<code>gzip</code> command.</p>
<p>Extract the file by typing</p>
<pre class="source"><code>$ gunzip GCF_000195955.2_ASM19595v2_genomic.fna.gz</code></pre>
<p>Make sure that is was extracted</p>
<pre class="source"><code>$ ls</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GCF_000195955.2_ASM19595v2_genomic.fna</span></span></code></pre>
</div>
<div id="challenge-what-is-the-size-of-the-genome" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-what-is-the-size-of-the-genome" class="callout-inner">
<h3 class="callout-title">Challenge: What is the size of the
genome?<a class="anchor" aria-label="anchor" href="#challenge-what-is-the-size-of-the-genome"></a>
</h3>
<div class="callout-content">
<p>Find out how many basepairs the genome has. Hints:</p>
<pre><code><span><span class="va">assembly</span><span class="op">-</span><span class="va">stats</span></span></code></pre>
<p>Get assembly statistics from FASTA and FASTQ files.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The genome has 4’411’532 bp.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> assembly-stats GCF_000195955.2_ASM19595v2_genomic.fna</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
stats for GCF_000195955.2_ASM19595v2_genomic.fna
sum = 4411532, n = 1, ave = 4411532.00, largest = 4411532
N50 = 4411532, n = 1
N60 = 4411532, n = 1
N70 = 4411532, n = 1
N80 = 4411532, n = 1
N90 = 4411532, n = 1
N100 = 4411532, n = 1
N_count = 0
Gaps = 0
</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="loops"><h2 class="section-heading">Loops<a class="anchor" aria-label="anchor" href="#loops"></a>
</h2>
<hr class="half-width">
<p><em>Loops</em> are key to productivity improvements through
automation as they allow us to execute commands repeatedly. Similar to
wildcards and tab completion, using loops also reduces the amount of
typing (and typing mistakes). Our next task is to download our <a href="molecular-epidemiology.html">data</a> from the short read archive
<a href="https://www.ebi.ac.uk/ena" class="external-link">(SRA) at the European Nucleotide
Archive (ENA)</a>. There are many repositories for public data. Some
model organisms or fields have specific databases, and there are ones
for particular types of data. Two of the most comprehensive are the
National Center for Biotechnology Information (NCBI) and European
Nucleotide Archive (EMBL-EBI). In this lesson we’re working with the
ENA, but the general process is the same for any database.</p>
<p>We can do this one by one but given that each download takes about
one to two hours, this could keep us up all night. Instead of
downloading one by one we can apply a loop. Let’s see what that looks
like and then we’ll discuss what we’re doing with each line of our
loop.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for filename in ERR01 ERR02 ERR03</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> echo <span class="ex">ftp://ftp.sra.ebi.ac.uk/</span><span class="st">"</span><span class="va">${filename}</span><span class="st">"</span><span class="ex">.fastq.gz</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>When the shell sees the keyword <code>for</code>, it knows to repeat
a command (or group of commands) once for each item in a list. Each time
the loop runs (called an iteration), an item in the list is assigned in
sequence to the <strong>variable</strong>, and the commands inside the
loop are executed, before moving on to the next item in the list.</p>
<p>Inside the loop, we call for the variable’s value by putting
<code>$</code> in front of it. The <code>$</code> tells the shell
interpreter to treat the <strong>variable</strong> as a variable name
and substitute its value in its place, rather than treat it as text or
an external command.</p>
<p>In this example, the list is seven filenames/ Each time the loop
iterates, it will assign a file name to the variable
<code>filename</code> and run the <code>wget</code> command. The first
time through the loop, <code>$filename</code> is
<code>ERR01.fastq.gz</code>. The interpreter runs the command
<code>wget</code> on <code>ERR01.fastq.gz</code> at the server <a href="ftp://ftp.sra.ebi.ac.uk/">ftp://ftp.sra.ebi.ac.uk/</a> For the
second iteration, <code>$filename</code> becomes
<code>ERR02.fastq.gz</code>. This time, the shell runs <code>wget</code>
on <code>ERR02.fastq.gz</code>.</p>
<p>Use {} to wrap the variable so that .fastq.gz will not be interpreted
as part of the variable name. In addition, quoting the shell variables
is a good practice AND necessary if your variables have spaces in
them.</p>
<p>For more, check <a href="https://mywiki.wooledge.org/BashPitfalls" class="external-link">Bash Pitfalls</a></p>
<div id="follow-the-prompt" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="follow-the-prompt" class="callout-inner">
<h3 class="callout-title">Follow the Prompt<a class="anchor" aria-label="anchor" href="#follow-the-prompt"></a>
</h3>
<div class="callout-content">
<p>The shell prompt changes from <code>$</code> to <code>&gt;</code> and
back again as we were typing in our loop. The second prompt,
<code>&gt;</code>, is different to remind us that we haven’t finished
typing a complete command yet. A semicolon, <code>;</code>, can be used
to separate two commands written on a single line.</p>
</div>
</div>
</div>
<div id="same-symbols-different-meanings" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="same-symbols-different-meanings" class="callout-inner">
<h3 class="callout-title">Same Symbols, Different Meanings<a class="anchor" aria-label="anchor" href="#same-symbols-different-meanings"></a>
</h3>
<div class="callout-content">
<p>Here we see <code>&gt;</code> being used a shell prompt, whereas
<code>&gt;</code> is also used to redirect output. Similarly,
<code>$</code> is used as a shell prompt, but, as we saw earlier, it is
also used to ask the shell to get the value of a variable.</p>
<p>If the <em>shell</em> prints <code>&gt;</code> or <code>$</code> then
it expects you to type something, and the symbol is a prompt.</p>
<p>If <em>you</em> type <code>&gt;</code> or <code>$</code> yourself, it
is an instruction from you that the shell to redirect output or get the
value of a variable.</p>
</div>
</div>
</div>
<p>We have called the variable in this loop <code>filename</code> in
order to make its purpose clearer to human readers. The shell itself
doesn’t care what the variable is called; if we wrote this loop as:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for x in ERR01 ERR02 ERR03</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${x}</span><span class="st">"</span>.fastq.gz</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>or:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for temperature in ERR01 ERR02 ERR03</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${temperature}</span><span class="st">"</span>.fastq.gz</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>it would work exactly the same way. <em>Don’t do this.</em> Programs
are only useful if people can understand them, so meaningless names
(like <code>x</code>) or misleading names (like
<code>temperature</code>) increase the odds that the program won’t do
what its readers think it does.</p>
<div id="multipart-commands" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="multipart-commands" class="callout-inner">
<h3 class="callout-title">Multipart commands<a class="anchor" aria-label="anchor" href="#multipart-commands"></a>
</h3>
<div class="callout-content">
<p>The <code>for</code> loop is interpreted as a multipart command. If
you press the up arrow on your keyboard to recall the command, it may be
shown like so (depends on your bash version):</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for filename in ERR01 ERR02 ERR03<span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${filename}</span><span class="st">"</span>.fastq.gz <span class="kw">;</span> <span class="cf">done</span></span></code></pre>
</div>
<p>When you check your history later, it will help your remember what
you did!</p>
</div>
</div>
</div>
</section><section id="download-the-sequenced-genomes-from-the-european-nucleotide-archive-ena"><h2 class="section-heading">Download the sequenced genomes from the European Nucleotide Archive
(ENA)<a class="anchor" aria-label="anchor" href="#download-the-sequenced-genomes-from-the-european-nucleotide-archive-ena"></a>
</h2>
<hr class="half-width">
<p>To ensure that we complete this tutorial in time, we will not perform
all operations on all genomes, only on one or two. The principles are -
however - the same for two or seven (or, to some extent, 189)
genomes.</p>
<p>Let’s download data for the <em>M. tuberculosis</em> genomes with a
<code>for</code> loop. We would use the following loop to download all
samples:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for files in ERR029/ERR029207 ERR029/ERR029206 ERR026/ERR026478 ERR026/ERR026474 ERR026/ERR026473 ERR026/ERR026481 ERR026/ERR026482</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># wget "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/"${files}"/*fastq.gz"</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>Note the <code>#</code> that prevents the <code>wget</code> command
from running: that would take too long, so we’re downloading only one
(that’s a short loop!):</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd molepi/data</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for files in ERR029/ERR029206 </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/"</span><span class="va">${files}</span><span class="st">"/*fastq.gz"</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>This will run for a little while (hopefully no more than 5-10
minutes).</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>
<code>wget</code> is a computer program to get data from the
internet</li>
<li>
<code>for</code> loops let you perform the same set of operations on
multiple files with a single command</li>
<li>Sequencing data is large</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-quality-control"><p>Content from <a href="quality-control.html">Assessing Read Quality and Trimming</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/quality-control.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 50 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>Let’s check if the downloads finished. Let’s have a look at the
files.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I describe the quality of my data?</li>
<li>How can I get rid of sequence data that doesn’t meet my quality
standards?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Clean FastQC reads for further analysis.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<pre class="source"><code>$ ls</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">ERR029206_1.fastq.gz</span></span>
<span><span class="va">ERR029206_2.fastq.gz</span> </span></code></pre>
</div>
<p>All new files start with ERR and have a <code>fastq.gz</code>
extension. It means they are in fastq format and compressed. Let’s make
a copy of one file and uncompress this file to inspect it further</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ERR029206_1.fastq.gz ERR029206_1.copy.fastq.gz</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> ERR029206_1.copy.fastq.gz</span></code></pre>
</div>
<section id="bioinformatics-workflows"><h2 class="section-heading">Bioinformatics workflows<a class="anchor" aria-label="anchor" href="#bioinformatics-workflows"></a>
</h2>
<hr class="half-width">
<p>When working with high-throughput sequencing data, the raw reads you
get off of the sequencer will need to pass through a number of different
tools in order to generate your final desired output. The execution of
this set of tools in a specified order is commonly referred to as a
<em>workflow</em> or a <em>pipeline</em>.</p>
<p>An example of the workflow we will be using is provided below with a
brief description of each step.</p>
<figure><img src="../fig/Workflow.png" alt="Schematic workflow of the bioinformatics workflow" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal">
<li>Quality control - Assessing quality using FastQC</li>
<li>Quality control - Trimming and/or filtering reads (if
necessary)</li>
<li>Align reads to reference genome - Snippy</li>
<li>Variant calling - Snippy</li>
<li>Clustering variants - iTOL</li>
<li>Assembly - SKESA</li>
<li>Annotation (extra material)</li>
<li>Pangenome analysis (extra material)</li>
<li>Clustering presence and absence of genes (extra material)</li>
<li>Comparison of clustering methods (extra material)</li>
<li>Data visualization - Microreact</li>
</ol>
<p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community. The
tools that are used to analyze data at different stages of the workflow
are therefore built under the assumption that the data will be provided
in a specific format.</p>
</section><section id="quality-control"><h2 class="section-heading">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<hr class="half-width">
<p>The first step in the variant calling workflow is to take the FASTQ
files received from the sequencing facility and assess the quality of
the sequence reads.</p>
<figure><img src="../fig/var_calling_workflow_qc.png" alt="Quality control workflow" class="figure mx-auto d-block"></figure><p>You’ve already seen that part, but repetition doesn’t hurt. If you
feel that it’s anyway unnecessary, move on to the next part.</p>
<div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a>
</h3>
<p>Although it looks complicated (and it is), it’s easy to understand
the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">fastq</a>
format with a little decoding. Some rules about the format include…</p>
<table class="table">
<colgroup>
<col width="5%">
<col width="94%">
</colgroup>
<thead><tr class="header">
<th>Line</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Always begins with ‘@’ and then information about the read</td>
</tr>
<tr class="even">
<td>2</td>
<td>The actual DNA sequence</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Always begins with a ‘+’ and sometimes the same info in line 1</td>
</tr>
<tr class="even">
<td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr>
</tbody>
</table>
<p>We can view the first complete read in one of the files our dataset
by using <code>head</code> to look at the first four lines.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n4</span> ERR029206_1.copy.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@ERR029206.1 IL31_5505:8:1:6233:1087#4/1
TCNAGTCAGCACACACATGCGAAAGAATCCACCGACTAGGGTCAGCGGGGTTTGCAGTTGGTCGCGGACGTAACCG
+
&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>One of the nucleotides in this read is unknown (<code>N</code>).</p>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g. 1 in 10)
or, equivalently, the base call accuracy (eg 90%). To make it possible
to line up each individual nucleotide with its quality score, the
numerical score is converted into a code where each individual character
represents the numerical quality score for an individual nucleotide. For
example, in the line above, the quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>The <code>&lt;</code> character and each of the <code>=</code>
characters represent the encoded quality for an individual nucleotide.
The numerical value assigned to each of these characters depends on the
sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using by Illumina version 1.8 onwards. Each character is
assigned a quality score between 0 and 40 as shown in the chart
below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI
                  |         |         |         |         |
Quality score:    0........10........20........30........40                                </code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. This quality score is logarithmically
based, so a quality score of 10 reflects a base call accuracy of 90%,
but a quality score of 20 reflects a base call accuracy of 99%. These
probability values are the results from the base calling algorithm and
dependent on how much signal was captured for the base
incorporation.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@ERR029206.1 IL31_5505:8:1:6233:1087#4/1
TCNAGTCAGCACACACATGCGAAAGAATCCACCGACTAGGGTCAGCGGGGTTTGCAGTTGGTCGCGGACGTAACCG
+
&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>We can now see that the quality of an <code>N</code> (<code>'</code>)
is 6, which corresponds to a over 25% error probability ($10^{-} $).</p>
<blockquote>
<h3 id="quality-encodings-vary">Quality Encodings Vary</h3>
<p>Although we’ve used a particular quality encoding system to
demonstrate interpretation of read quality, different sequencing
machines use different encoding systems. This means that, depending on
which sequencer you use to generate your data, a <code>#</code> may not
be an indicator of a poor quality base call.</p>
</blockquote>
<div id="callout1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout<a class="anchor" aria-label="anchor" href="#callout1"></a>
</h3>
<div class="callout-content">
<p>This mainly relates to older Solexa/Illumina data, but it’s essential
that you know which sequencing platform was used to generate your data,
so that you can tell your quality control program which encoding to use.
If you choose the wrong encoding, you run the risk of throwing away good
reads or (even worse) not throwing away bad reads!</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="assessing-quality-using-fastqc">Assessing Quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a>
</h3>
<p>In real life, you won’t be assessing the quality of your reads by
visually inspecting your FASTQ files. Rather, you’ll be using a software
program to assess read quality and filter out poor quality reads. We’ll
first use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads.Later in our workflow, we’ll use
another program to filter out poor quality reads.</p>
<p>FastQC has a number of features which can give you a quick impression
of any problems your data may have, so you can take these issues into
consideration before moving forward with your analyses. Rather than
looking at quality scores for each individual read, FastQC looks at
quality collectively across all reads within a sample. The image below
shows a FastQC-generated plot that indicates a very high quality
sample:</p>
<figure><img src="../fig/good_quality1.8.png" alt="A FastQC output showing reads with good quality" class="figure mx-auto d-block"></figure><p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. For each position, there is a box-and-whisker plot
showing the distribution of quality scores for all reads at that
position. The horizontal red line indicates the median quality score and
the yellow box shows the 2nd to 3rd quartile range. This means that 50%
of reads have a quality score that falls within the range of the yellow
box at that position. The whiskers show the range to the 1st and 4th
quartile.</p>
<p>For each position in this sample, the quality values do not drop much
lower than 32. This is a high quality score. The plot background is also
color-coded to identify good (green), acceptable (yellow), and bad (red)
quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the
spectrum.</p>
<figure><img src="../fig/bad_quality1.8.png" alt="A FastQC output showing reads with bad quality" class="figure mx-auto d-block"></figure><p>Here, we see positions within the read in which the boxes span a much
wider range. Also, quality scores drop quite low into the “bad” range,
particularly on the tail end of the reads. The FastQC tool produces
several other diagnostic plots to assess sample quality, in addition to
the one plotted above.</p>
</div>
<div class="section level3">
<h3 id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a>
</h3>
<p>Navigate to your FASTQ dataset:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/data/</span></code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>We downloaded the samples. How many FASTQ files are in this dataset?
Why is this? What is the difference between the two files that start
with the same name? Discuss with your neighbour.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>There are 2 FASTQ files. Each sample has <a href="https://thesequencingcenter.com/knowledge-base/what-are-paired-end-reads/" class="external-link">paired-end
reads</a>.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, including files that
are compressed (<code>.gz</code>) so we can use the
<code>*.fastq.gz</code> wildcard to run FastQC on all of the compressed
FASTQ files in this directory.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="pp">*</span>.fastq.gz</span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of ERR029206_1.fastq.gz
Approx 5% complete for ERR029206_1.fastq.gz
Approx 10% complete for ERR029206_1.fastq.gz
Approx 15% complete for ERR029206_1.fastq.gz
Approx 20% complete for ERR029206_1.fastq.gz
Approx 25% complete for ERR029206_1.fastq.gz
Approx 30% complete for ERR029206_1.fastq.gz</code></pre>
</div>
<p>In total, it should take about 15-20 minutes for FastQC to run on all
four of our FASTQ files. When the analysis completes, your prompt will
return. So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 85% complete for ERR029206_2.fastq.gz
Approx 90% complete for ERR029206_2.fastq.gz
Approx 95% complete for ERR029206_2.fastq.gz
Analysis complete for ERR029206_2.fastq.gz
data $</code></pre>
</div>
<p>The FastQC program has created several new files within our
<code>/data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.copy.fastq</span></span>
<span><span class="va">ERR029206_1.fastq.gz</span></span>
<span><span class="va">ERR029206_1_fastqc.html</span></span>
<span><span class="va">ERR029206_1_fastqc.zip</span></span>
<span><span class="va">ERR029206_2.fastq.gz</span></span>
<span><span class="va">ERR029206_2_fastqc.html</span></span>
<span><span class="va">ERR029206_2_fastqc.zip</span></span></code></pre>
</div>
<p>For each input gzipped FASTQ file, FastQC has created a
<code>.zip</code> file and a <code>.html</code> file. The
<code>.zip</code> file extension indicates that this is actually a
compressed set of multiple output files. We’ll be working with these
output files soon. The <code>.html</code> file is a stable webpage
displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir ../results/fastqc_untrimmed_reads</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.zip ../results/fastqc_untrimmed_reads/</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.html ../results/fastqc_untrimmed_reads/</span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ../results/fastqc_untrimmed_reads/</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="viewing-html-files">Viewing HTML files<a class="anchor" aria-label="anchor" href="#viewing-html-files"></a>
</h3>
<p>If we are working on our local computers, we may display each of
these HTML files as a webpage:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> firefox ERR029206_1_fastqc.html</span></code></pre>
</div>
<p>You may need to replace the command <code>firefox</code> with the
name of another web navigator if you have one.</p>
<div id="exercise-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-1"></a>
</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Do the sample looks good in
terms of per base sequence quality?</p>
</div>
</div>
</div>
<div id="exercise-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-2"></a>
</h3>
<div class="callout-content">
<p>What are the read lengths of the different samples?</p>
</div>
</div>
</div>
</div>
</section><section id="cleaning-reads"><h2 class="section-heading">Cleaning Reads<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a>
</h2>
<hr class="half-width">
<p>It’s very common to have some reads within a sample, or some
positions (near the beginning or end of reads) across all reads that are
low quality and should be discarded. We will use a program called <a href="https://github.com/lh3/seqtk" class="external-link">seqtk</a> to filter poor quality
reads and trim poor quality bases from our samples.</p>
<div class="section level3">
<h3 id="seqtk-options">Seqtk Options<a class="anchor" aria-label="anchor" href="#seqtk-options"></a>
</h3>
<p>Seqtk is a program written C and aims to be a Swiss army knife for
sequencing reads. You don’t need to learn C to use Seqtk, but the fact
that it’s a C program helps explain the syntax that is used to run
Seqtk. Seqtk takes as input files either FASTQ files or gzipped FASTQ
files and outputs FASTQ or FASTA files. The basic command to run Seqtk
starts like this:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> seqtk</span></code></pre>
</div>
<p>That’s just the basic command, however. Seqtk has a variety of
options and parameters. We will need to specify what options we want to
use for our analysis. Here are some of the options:</p>
<table class="table">
<thead><tr class="header">
<th>option</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>seq</code></td>
<td>common transformation of FASTA/Q</td>
</tr>
<tr class="even">
<td><code>comp</code></td>
<td>get the nucleotide composition of FASTA/Q</td>
</tr>
<tr class="odd">
<td><code>trimfq</code></td>
<td>trim FASTQ using the Phred algorithm</td>
</tr>
</tbody>
</table>
<p>In addition to these options, there are a number if trimming options
available:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> seqtk trimfq</span></code></pre>
</div>
<table class="table">
<colgroup>
<col width="8%">
<col width="91%">
</colgroup>
<thead><tr class="header">
<th>step</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>-q</code></td>
<td>error rate threshold (disabled by -b/-e) [0.05]</td>
</tr>
<tr class="even">
<td><code>-l</code></td>
<td>maximally trim down to INT bp (disabled by -b/-e) [30]</td>
</tr>
<tr class="odd">
<td><code>-b</code></td>
<td>trim INT bp from left (non-zero to disable -q/-l) [0]</td>
</tr>
<tr class="even">
<td><code>-e</code></td>
<td>trim INT bp from right (non-zero to disable -q/-l) [0].</td>
</tr>
</tbody>
</table>
<p>We will use only a few of these options in our analysis. It is
important to understand the steps you are using to clean your data.</p>
<p>A complete command for trimming with seqtk will look something like
this:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> seqtk trimfq <span class="at">-q</span> 0.01 ERR01_1.fastq.gz <span class="op">&gt;</span> ERR01_1_trim.fastq</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="trimming">Trimming<a class="anchor" aria-label="anchor" href="#trimming"></a>
</h3>
<p>Now we will run seqtk trimfq on our data. To begin, navigate to your
<code>data</code> directory:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/data/</span></code></pre>
</div>
<p>We are going to run seqtk on one sample giving it an error rate
threshold of 0.01 which indicates the base call accuracy. We request
that, after trimming, the chances that a base is called incorrectly are
only 1 in 10000.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> seqtk trimfq <span class="at">-q</span> 0.01 ERR029206_1.fastq.gz <span class="op">&gt;</span> ERR029206_1_trim.fastq</span></code></pre>
</div>
<p>Notice that we needed to redirect the output to a file. If we don’t
do that, the trimmed fastq data will be displayed in the console.</p>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise-3"></a>
</h3>
<div class="callout-content">
<p>Use seqtk fqchk to compare the untrimmed and trimmed reads of
ERR029206_1 in terms of number of sequenced bases, percentage of A,G,C,T
and N and average quality. What do you notice? Discuss with your
neighbor.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1.copy.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 76; max_len: 76; avg_len: 76.00; 36 distinct quality values
POS	#bases	%A	%C	%G	%T	%N	avgQ	errQ	%low	%high
ALL	315791552	18.0	32.1	32.2	17.7	0.0	30.6	21.4	5.4	94.6</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 69.12; 36 distinct quality values
POS	#bases	%A	%C	%G	%T	%N	avgQ	errQ	%low	%high
ALL	287211920	18.2	32.0	31.9	17.8	0.0	31.7	28.8	1.3	98.7</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We’ve just successfully trimmed one of our FASTQ files! However,
there is some bad news. <code>seqtk</code> can only operate on one
sample at a time and we have more than one sample. The good news is that
we can use a <code>for</code> loop to iterate through our sample files
quickly!</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for infile in <span class="pp">*</span>.fastq.gz</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> trimfq <span class="at">-q</span> 0.01 <span class="st">"</span><span class="va">${infile}</span><span class="st">"</span> <span class="op">&gt;</span>  <span class="st">"</span><span class="va">${outfile}</span><span class="st">"</span> </span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>The new part in our <code>for</code> loop is the line:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span></code></pre>
</div>
<p><code>infile</code> is the first variable in our loop and takes the
value of each of the FASTQ files in our directory. <code>outfile</code>
is the second variable in our loop and is defined by adding
<code>_trim.fastq</code> to the end of our input file name. Use
<code>{}</code> to wrap the variable so that <code>_trim.fastq</code>
will not be interpreted as part of the variable name. In addition,
quoting the shell variables is a good practice AND necessary if your
variables have spaces in them.</p>
<p>For more, check <a href="https://mywiki.wooledge.org/BashPitfalls" class="external-link">Bash Pitfalls</a>. There
are no spaces before or after the <code>=</code>.</p>
<p>Go ahead and run the for loop. It should take a few minutes for seqtk
to run for each of our fourteen input files. Once it’s done running,
take a look at your directory contents.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.copy.fastq</span> </span>
<span><span class="va">ERR029206_1.fastq.gz</span></span>
<span><span class="va">ERR029206_1.fastq.gz_trim.fastq</span> </span>
<span><span class="va">ERR029206_1_trim.fastq</span> </span>
<span><span class="va">ERR029206_2.fastq.gz</span></span>
<span><span class="va">ERR029206_2.fastq.gz_trim.fastq</span></span>
<span><span class="va">GCF_000195955.2_ASM19595v2_genomic.fna</span></span></code></pre>
</div>
<p>We’ve now completed the trimming and filtering steps of our quality
control process! Before we move on, let’s move our trimmed FASTQ files
to a new subdirectory within our <code>data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir trimmed_fastq</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>fastq.gz_trim<span class="pp">*</span> trimmed_fastq</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd trimmed_fastq</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.fastq.gz_trim.fastq</span> </span>
<span><span class="va">ERR029206_2.fastq.gz_trim.fastq</span> </span></code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<p>Again, use seqtk fqchk to compare the untrimmed and trimmed reads of
both samples. Note the number of bases ‘#bases’ of the trimmed and
untrimmed reads. Calculate the theoretical coverage of the genomes
before and after trimming, assuming that all our genomes do have the
same size as our reference genome (4411532 bases).</p>
<p>Hint: Sum up forward and reverse reads!</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1.fastq.gz_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 69.12; 36 distinct quality values
POS	#bases	%A	%C	%G	%T	%N	avgQ	errQ	%low	%high
ALL	287211920	18.2	32.0	31.9	17.8	0.0	31.7	28.8	1.3	98.7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_2.fastq.gz_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 63.79; 36 distinct quality values
POS	#bases	%A	%C	%G	%T	%N	avgQ	errQ	%low	%high
ALL	265048852	18.1	32.1	32.1	17.7	0.0	31.2	25.7	2.5	97.5</code></pre>
</div>
<p>Coverage = #bases (forward + reverse) / genome size In this case:
125.18 = (287211920 + 265048852) / 4411532</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>In bioinformatic workflows the output of one tool is the input of
the other.</li>
<li>FastQC is used to judge the quality of sequencing reads.</li>
<li>Data cleaning is an essential step in a genomics pipeline.</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></section><section id="aio-sequence-assembly"><p>Content from <a href="sequence-assembly.html">Sequence assembly</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/sequence-assembly.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 40 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>Sequence assembly means the alignment and merging of reads in order
to reconstruct the original sequence. The assembly of a genome from
short sequencing reads can take a while. We will therefore run this
process only for two genomes.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can the information in the sequencing reads be reduced?</li>
<li>What are the different methods for assembly?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand differences between assembly methods</li>
<li>Assemble the short reads</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="sequence-assembly"><h2 class="section-heading">Sequence assembly<a class="anchor" aria-label="anchor" href="#sequence-assembly"></a>
</h2>
<hr class="half-width">
<p>The assembler we will run is <a href="https://github.com/ncbi/SKESA" class="external-link">SKESA</a>. SKESA generates a final
assembly from multiple kmers. A list of kmers is automatically selected
by SKESA using the maximum read length of the input data, and each
individual kmer contributes to the final assembly. If desired, the
lengths of kmers can be specified with the –kmer and –steps flags which
will adjust automatic kmer selection.</p>
</section><section id="assembly"><h2 class="section-heading">Assembly<a class="anchor" aria-label="anchor" href="#assembly"></a>
</h2>
<hr class="half-width">
<p>Because assembly of each genome might take a couple of hours, we will
run all assemblies in a loop overnight. It is important to run them
within the screen session or else the process will be terminated if we
disconnect from the machine.</p>
<p>Preparation</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir ~/molepi/results/assembly</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/data/trimmed_fastq</span></code></pre>
</div>
<p>To run SKESA we will use the skesa command with a number of option
that we will explore later on. We can start the loop (again, a short
loop) with the assemblies with:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for sample in ERR029206</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">skesa</span> <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>--cores 2 <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>--memory 8 <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>--fastq <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_1.fastq.gz_trim.fastq <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_2.fastq.gz_trim.fastq <span class="dv">1</span><span class="op">&gt;</span> ../../results/assembly/<span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>.fasta</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>A non-quoted backslash, <code>\</code> is used as an escape character
in Bash. It preserves the literal value of the next character that
follows, with the exception of newline. This means that the back slash
starts a new line without starting a new command - we only add it for
better readability.</p>
<p>The assembly should take about 2-3 minutes per genome, using 2 cores
and 8 GB of memory.</p>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Downloading the missing assemblies
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>If you haven’t assembled all samples, download the finished contigs
to your computer and decompress them. They will decompress in a folder
called <code>assembly</code>, so one solution is to rename the assembly
folder you already have to something else:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/molepi/results</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> assembly partial_assembly</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/lguy/course-microbial-genomics/raw/main/episodes/files/assemblies.tar.gz .</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvzf assemblies.tar.gz</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> assembly</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ERR026473.fasta ERR026478.fasta ERR026482.fasta ERR029207.fasta
ERR026474.fasta ERR026481.fasta ERR029206.fasta</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-what-do-the-command-line-parameters-of-skesa-mean" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-what-do-the-command-line-parameters-of-skesa-mean" class="callout-inner">
<h3 class="callout-title">Challenge: What do the command line parameters
of SKESA mean??<a class="anchor" aria-label="anchor" href="#challenge-what-do-the-command-line-parameters-of-skesa-mean"></a>
</h3>
<div class="callout-content">
<p>Find out a) the version of SKESA you are using and b) what the used
command line parameters mean.</p>
<p>Hint:</p>
<pre><code>$ skesa -h</code></pre>
<p>prints the ‘help’</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ skesa -h
..
General options:
-h [ --help ]                 Produce help message
-v [ --version ]              Print version
..
--cores arg (=0)              Number of cores to use (default all) [integer]
--memory arg (=32)            Memory available (GB, only for sorted counter)
..  
--fastq arg                   Input fastq file(s) (could be used multiple times for different runs) [string]
..
..</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-how-many-contigs-were-generated-by-skesa" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-how-many-contigs-were-generated-by-skesa" class="callout-inner">
<h3 class="callout-title">Exercise: How many contigs were generated by
SKESA??<a class="anchor" aria-label="anchor" href="#exercise-how-many-contigs-were-generated-by-skesa"></a>
</h3>
<div class="callout-content">
<p>Find out how many contigs there are in the <em>M. tuberculosis</em>
isolates.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">assembly-stats</span> ERR026473.fasta</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>stats for ERR029206.fasta
sum = 4233864, n = 451, ave = 9387.73, largest = 70968
N50 = 20633, n = 68
N60 = 16723, n = 91
N70 = 13473, n = 120
N80 = 9592, n = 157
N90 = 5883, n = 213
N100 = 200, n = 451
N_count = 0
Gaps = 0</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Assembly is a process which aligns and merges fragments from a
longer DNA sequence in order to reconstruct the original sequence.</li>
<li>k-mers are short fragments of DNA of length k</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-snps-phylogeny"><p>Content from <a href="snps-phylogeny.html">SNP phylogeny</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/snps-phylogeny.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 75 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>In this episode we will try to pinpoint single nucleotide variants or
single nucleotide polymorphism (SNPs) between our samples and the
reference. The SNPs are determined by a process called read mapping in
which they are aligned to the reference sequence.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to generate a phylogenetic tree from SNP data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Map reads against a reference genome</li>
<li>Extract single nucleotide polymorphisms</li>
<li>Establish a phylogenetic tree from the SNP data</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<figure><img src="../fig/snps_reads_02.png" alt="Mapping of SNP reads" class="figure mx-auto d-block"></figure><p>The identified SNPs will be used to compare the isolates to each
other and to establish a phylogenetic tree.</p>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">
  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div>
  Getting missing data
  </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>If you don’t have all the trimmed reads, you can download them (will
take a while, <em>dont’t do it if you have the reads</em>):</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/molepi/data</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> trimmed_fastq trimmed_fastq_partial</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://uppsala.box.com/s/s1zpt0w84iq4o42qhgho5vrz9qnpxouk trimmed_fastq.tar.gz</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvzf trimmed_fastq.tar.gz</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> trimmed_fastq</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<section id="snp-calling"><h2 class="section-heading">SNP calling<a class="anchor" aria-label="anchor" href="#snp-calling"></a>
</h2>
<hr class="half-width">
<p><a href="https://github.com/tseemann/snippy" class="external-link">snippy</a> is a pipeline
that contains different tools to determine SNPs in sequencing reads
against a reference genome. It takes forward and reverse reads of
paired-end sequences and aligns them against a reference.</p>
<p>First we’ll create a folder to hold the results from snippy:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/results</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir snps</span></code></pre>
</div>
<p>We need to activate a separate environment to run snippy:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mamba activate snippy</span></code></pre>
</div>
<p>Snippy is fast but a single run still takes about 2-3 minutes. We
would therefore tell snippy to run all the samples after each other.
However this time we cannot use a wildcard to do so. We would instead
run all the samples in a loop. The “real” loop is commented out, and we
use one that runs only the sample we’ve analyzed so far.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ../data/trimmed_fastq/</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> <span class="co">#for sample in ERR026473 ERR026474 ERR026478 ERR026481 ERR026482 ERR029206 ERR029207</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> for sample in ERR029206</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">snippy</span> <span class="at">--ram</span> 1 <span class="at">--outdir</span> ../../results/snps/<span class="st">"</span><span class="va">${sample}</span><span class="st">"</span> <span class="at">--ref</span> ../GCF_000195955.2_ASM19595v2_genomic.fna <span class="at">--R1</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_1.fastq.gz_trim.fastq <span class="at">--R2</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_2.fastq.gz_trim.fastq</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>Here, we provide snippy with an output folder
(<code>--outdir</code>), the location of the reference genome
(<code>--ref</code>), and the trimmed read files for each end of the
pair (<code>--R1</code> and <code>--R2</code>). We also indicate that
snippy should use 1 Gb of memory (<code>--ram 1</code>, very specific
issue to this VM).</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n10</span> ~/molepi/results/snps/ERR029206/snps.tab </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>CHROM	POS	TYPE	REF	ALT	EVIDENCE	FTYPE	STRAND	NT_POS	AA_POS	EFFECT	LOCUS_TAG	GENE	PRODUCT
NC_000962.3	1849	snp	C	A	A:217 C:0
NC_000962.3	1977	snp	A	G	G:118 A:0
NC_000962.3	4013	snp	T	C	C:176 T:0
NC_000962.3	7362	snp	G	C	C:156 G:0
NC_000962.3	7585	snp	G	C	C:143 G:0
NC_000962.3	9304	snp	G	A	A:127 G:0
NC_000962.3	11820	snp	C	G	G:164 C:0
NC_000962.3	11879	snp	A	G	G:125 A:0
NC_000962.3	14785	snp	T	C	C:163 T:1</code></pre>
</div>
<p>This list gives us information on every SNP that was found by snippy
when compared to the reference genome. The first SNP is found at the
position 1849 of the reference genome, and is a C in the H37Rv
(reference strain) and an A in isolate ERR029206. There is a high
confidence associated with this SNP: an A has been found 217 times in
the sequencing reads and never a C at this position.</p>
<p>Now download a <a href="../files/snps.tar.gz">reduced version of the
result of snippy</a> on the other samples, save it in the
<code>~/molepi/results/</code> folder. We will also rename the existing
<code>snps</code> folder to save it.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/molepi/results/</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> snps snps_partial</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> snps</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> snps</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvzf ../snps.tar.gz</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ERR026473     ERR026482
ERR026474     ERR029206
ERR026478     ERR029207
ERR026481</code></pre>
</div>
<p>The folders don’t contain everything that is output by
<code>snippy</code>, to save space. But it is enough to run
<code>snippy-core</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ERR026473</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ref.fa          snps.filt.vcf   snps.tab
ref.fa.fai      snps.gff        snps.txt
snps.aligned.fa snps.html       snps.vcf
snps.bam.bai    snps.log        snps.vcf.gz
snps.bed        snps.raw.vcf    snps.vcf.gz.csi
snps.csv        snps.subs.vcf</code></pre>
</div>
<div id="challenge-how-many-snps-were-identified-in-each-sample" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-how-many-snps-were-identified-in-each-sample" class="callout-inner">
<h3 class="callout-title">Challenge: How many SNPs were identified in
each sample??<a class="anchor" aria-label="anchor" href="#challenge-how-many-snps-were-identified-in-each-sample"></a>
</h3>
<div class="callout-content">
<p>Find out how many SNPs were identified in the <em>M.
tuberculosis</em> isolates when compared to H37Rv. Hint: The
<code>.txt</code> file in the snippy output contains summary
information:</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat ~/molepi/results/snps/ERR029207/snps.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DateTime        2023-05-17T13:51:57
ReadFiles       [...]/data/trimmed_fastq/ERR029207_1.fastq.gz_trim.fastq [...]/data/trimmed_fastq/ERR029207_2.fastq.gz_trim.fastq
Reference       [...]/data/GCF_000195955.2_ASM19595v2_genomic.fna
ReferenceSize   4411532
Software        snippy 4.6.0
Variant-COMPLEX 32
Variant-DEL     57
Variant-INS     52
Variant-MNP     2
Variant-SNP     1290
VariantTotal    1433</code></pre>
</div>
<p>This <em>M. tuberculosis</em> isolate contains 1290 SNPs compared to
H37Rv.</p>
<p>Repeat this for the other isolates. How about using a
<code>for</code> loop?</p>
</div>
</div>
</div>
</div>
</section><section id="core-snps"><h2 class="section-heading">Core SNPs<a class="anchor" aria-label="anchor" href="#core-snps"></a>
</h2>
<hr class="half-width">
<p>In order to compare the identified SNPs with each other we need to
know if a certain position exists in all isolates. A core site can have
the same nucleotide in every sample (monomorphic) or some samples can be
different (polymorphic).</p>
<p>In this second part, after identifying them, snippy will concatenate
the core SNPs, i.e. ignoring sites that are monomorphic in all isolates
and in the reference. Concatenation of the SNP sites considerably
reduces the size of the alignment.</p>
<p>The <code>--ref</code> argument provides the reference genome. Each
folder containing the result of the previous step of <code>snippy</code>
is then added to the command line.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/results/snps/</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snippy-core <span class="at">--ref</span><span class="op">=</span>../../data/GCF_000195955.2_ASM19595v2_genomic.fna ERR026473 ERR026474 ERR026478 ERR026481 ERR026482 ERR029206 ERR029207</span></code></pre>
</div>
<p>The last few lines look like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Loaded 1 sequences totalling 4411532 bp.
Will mask 0 regions totalling 0 bp ~ 0.00%
0	ERR026473	snp=1378	del=174	ins=165	het=747	unaligned=125683
1	ERR026474	snp=1369	del=180	ins=143	het=811	unaligned=123020
2	ERR026478	snp=1381	del=221	ins=143	het=638	unaligned=112145
3	ERR026481	snp=1349	del=215	ins=138	het=754	unaligned=126217
4	ERR026482	snp=1348	del=215	ins=145	het=911	unaligned=127554
5	ERR029206	snp=1352	del=152	ins=98	het=1839	unaligned=122677
6	ERR029207	snp=1355	del=152	ins=97	het=1900	unaligned=118570
Opening: core.tab
Opening: core.vcf
Processing contig: NC_000962.3
Generating core.full.aln
Creating TSV file: core.txt
Running: snp-sites -c -o core.aln core.full.aln
This analysis is totally hard-core!
Done.</code></pre>
</div>
<p>Our output was written to ‘core.aln’. But let’s have a look at the
results.</p>
<div id="discussion-whats-in-the-output-of-snippy" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-whats-in-the-output-of-snippy" class="callout-inner">
<h3 class="callout-title">Discussion: What’s in the output of
snippy??<a class="anchor" aria-label="anchor" href="#discussion-whats-in-the-output-of-snippy"></a>
</h3>
<div class="callout-content">
<p>Have a look at the content of these three files with <code>cat</code>
or <code>head</code>.</p>
<p><code>core.aln</code></p>
<p><code>core.full.aln</code></p>
<p><code>core.tab</code></p>
<p><code>core.txt</code></p>
<p>What is the difference between these files? Why is core.aln smaller
than ? What is in core.aln?</p>
</div>
</div>
</div>
<p>Let’s deactivate the specific snippy <code>env</code> before we move
on:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mamba deactivate snippy</span></code></pre>
</div>
</section><section id="phylogenetic-tree"><h2 class="section-heading">Phylogenetic tree<a class="anchor" aria-label="anchor" href="#phylogenetic-tree"></a>
</h2>
<hr class="half-width">
<p>Phylogenetic trees have been discussed during the lectures. We will
here establish a phylogenetic tree from the file ‘core.aln’ with <a href="http://www.iqtree.org/" class="external-link">IQ-TREE</a>. IQ-TREE is a phylogeny
software based on the maximum-likelihood principle. IQ-TREE has been
widely used and cited. There is a very wide range of parameters that
need to be chosen, such as nucleotide or amino-acid substitution
models.</p>
<p>In our case, we want IQ-TREE to automatically select the best
substitution model. IQ-TREE does that by testing many (among a very
large collection) substitution models. We also have SNP data, which by
definition do not contain constant (invariable) sites. We thus input the
alignment with the <code>-s</code> option and the model with
<code>-m MFP+ASC</code>. <code>MFP</code> will tell IQ-TREE to test a
range of models, and <code>ASC</code> will correct for the fact that
there is no constant sites.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> iqtree <span class="at">-s</span> core.aln <span class="at">-m</span> MFP+ASC</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Total wall-clock time used: 0.312 sec (0h:0m:0s)

Analysis results written to:
  IQ-TREE report:                core.aln.iqtree
  Maximum-likelihood tree:       core.aln.treefile
  Likelihood distances:          core.aln.mldist
  Screen log file:               core.aln.log

Date and Time: Thu May 18 09:00:37 2023</code></pre>
</div>
<p>With this small data set, IQ-TREE finishes very quickly. Let’s put
the resulting files into a separate folder and let’s rename our
resulting tree.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/results</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir tree</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mv snps/core.aln.<span class="pp">*</span> tree</span></code></pre>
</div>
<p>Let’s inspect our tree, and give it another extension to make it
clear it is a newick file.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/molepi/results/</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd tree</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mv core.aln.treefile core.aln.newick</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> core.aln.newick</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>(ERR026473:0.0004854701,ERR026474:0.0004356437,((((ERR026478:0.0000010000,ERR029207:0.0000010000):0.0000010000,ERR029206:0.0000131789):0.0006712219,Reference:0.0075887710):0.0000152316,(ERR026481:0.0000066238,ERR026482:0.0000131899):0.0005602421):0.0002373720);</code></pre>
</div>
<p>This does not look much like a tree yet. The tree is written in a
bracket annotation, the <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick format</a>. In
order to make sense of it we can better view this in a tree viewer.</p>
<div id="challenge-can-you-identify-what-substitution-model-iq-tree-used" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-can-you-identify-what-substitution-model-iq-tree-used" class="callout-inner">
<h3 class="callout-title">Challenge: Can you identify what substitution
model IQ-TREE used?<a class="anchor" aria-label="anchor" href="#challenge-can-you-identify-what-substitution-model-iq-tree-used"></a>
</h3>
<div class="callout-content">
<p>Hint: The log file of IQ-TREE contains a lot of information.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat core.aln.iqtree</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Best-fit model according to BIC: TVMe+ASC
...</code></pre>
</div>
<p>IQ-TREE chose the TVMe+ASC model. We’ve discussed the ASC above, and
you can read more about the TVMe in the <a href="https://www.iqtree.org/doc/Substitution-Models" class="external-link">IQ-TREE
manual</a>. In short, it is a mdoel where base frequencies are deemed
equal, and all rates estimated independently, except the rates of
transversions (A &lt;-&gt; G and C &lt;-&gt; T) are equal.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Single nucleotide polymorphisms can be identified by mapping reads
to a reference genome</li>
<li>Parameters for the analysis have to be selected based on expected
outcomes for this organism</li>
<li>Concatenation of SNPs helps to reduce analysis volume</li>
<li>Phylogenetic trees can be written with a bracket syntax in Newick
format</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-data-visualization"><p>Content from <a href="data-visualization.html">Data Visualization</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/data-visualization.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 70 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<section id="comparison-of-clustering-results"><h2 class="section-heading">Comparison of clustering results<a class="anchor" aria-label="anchor" href="#comparison-of-clustering-results"></a>
</h2>
<hr class="half-width">
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How are phylogenetic trees viewed and compared?</li>
<li>How can I visualize several layers of data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explore the resulting trees in conjunction with the meta data.</li>
<li>Make an estimation on the likelihood of transmission events</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>We grouped (clustered) our isolates by the information that we
extracted from the sequencing reads. We compared our isolates to a
reference genome and clustered them on basis of the single nucleotide
variations that they exhibit when compared to this reference.</p>
<figure><img src="../fig/Workflow.png" alt="workflow" class="figure mx-auto d-block"></figure><p>Next, we will visualize this clustering in the context of the meta
data. Make sure you have the SNP tree that you generated (or can be
downloaded here <a href="../files/core.aln.newick">core.aln.newick</a>) on
your own computer. Download also the <a href="../files/Molepi_meta.csv">metadata file</a>.</p>
</section><section id="visualization-of-genetic-information-and-metadata"><h2 class="section-heading">Visualization of genetic information and metadata<a class="anchor" aria-label="anchor" href="#visualization-of-genetic-information-and-metadata"></a>
</h2>
<hr class="half-width">
<p>Visualization is frequently used to aid the interpretation of complex
datasets. Within microbial genomics, visualizing the relationships
between multiple genomes as a tree provides a framework onto which
associated data (geographical, temporal, phenotypic and epidemiological)
are added to generate hypotheses and to explore the dynamics of the
system under investigation.</p>
<ol style="list-style-type: decimal">
<li>Open Chrome. Go to <a href="https://microreact.org/" class="external-link">microreact</a>.</li>
<li>Click on ‘Upload’. Drag-and-drop the newick tree produced by snippy
and IQ-TREE (<a href="../files/core.aln.newick">core.aln.newick</a>).</li>
<li>Click on “Add more files” and select the the <a href="../files/Molepi_meta.csv">metadata file</a>, which is a table
comprising metadata file. Choose ‘Data (csv or tsv)’ as file kind.</li>
<li>Click on continue.</li>
</ol>
<p>First, the tree should be rooted with the reference strain (the one
that has the longest branch). Hover over the middle of the long branch,
until a circle appears. Right-click on it, and select “Set as root
(re-root)”.</p>
<p>Explore the location, time and further meta data. Play around with
the different visualization options with the help of the <a href="https://docs.microreact.org/" class="external-link">documentation of microreact</a>.</p>
<div id="discussion" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Which transmission events are likely based on the metadata
(location, RFLP, date) alone?</li>
<li>Which transmission events are likely based on the SNP data?</li>
<li>Draw a transmission tree if possible.</li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Genetic information can confirm or contradict the meta data</li>
</ul>
</div>
</div>
</div>
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</section></section><section id="aio-conclusions"><p>Content from <a href="conclusions.html">Transmission?</a></p>
<hr>
<p> Last updated on 2023-08-15 | 
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/conclusions.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 25 minutes </p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>The low mutation rate of <em>M. tuberculosis</em> means that even at
the highest resolution provided by whole genome sequencing it is still
difficult to confidently affirm the inferences of transmission made by
traditional epidemiological techniques. This means it is very difficult
to determine transmission inclusively. However, whole genome sequencing
does in some cases allow us to exclude direct transmission, by using the
phylogenetic context provided by other strains. Not only does whole
genome sequencing provide the inter-cluster differentiation provided by
current typing methods, but it also achieves intra-cluster resolution
which can be used to inform epidemiological investigation.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>In which cases is transmission likely?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Compare and interpret the different data sources.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="discussion" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion<a class="anchor" aria-label="anchor" href="#discussion"></a>
</h3>
<div class="callout-content">
<p>Compare the pangenome clustering with the clustering based on SNPs
and with the data from epidemiological contact tracing:</p>
<p>Cluster A</p>
<ul>
<li>Patient A1 - sample ERR029207</li>
<li>Patient A2 - sample ERR029206</li>
<li>Patient A3 - sample ERR026478</li>
</ul>
<p>Cluster B</p>
<ul>
<li>Patient B1 - sample ERR026474</li>
<li>Patient B2 - sample ERR026473</li>
</ul>
<p>Cluster C</p>
<ul>
<li>Patient C1 - sample ERR026481</li>
<li>Patient C2 - sample ERR026482</li>
</ul>
<p>Which transmission events are affirmed by the genomic data? Which
ones partially or not? Why?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>SNP phylogeny and metadata can convey different messages</li>
<li>Human interpretation is often needed to weigh the different
information sources.</li>
<li>The low mutation rate of <em>M. tuberculosis</em> does not allow to
make confident inferences of transmission but does allow to exclude
transmission</li>
</ul>
</div>
</div>
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:lionel.guy@imbim.uu.se">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.14.0" class="external-link">sandpaper (0.14.0)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.7.1" class="external-link">pegboard (0.7.1)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.3.1" class="external-link">varnish (0.3.1)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "evolution, phylogenetics, molecular epidemiology",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "dateCreated": "2023-05-08",
  "dateModified": "2023-10-10",
  "datePublished": "2023-10-10"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

