<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Microbial Genomics: evolution, phylogenetics, and molecular epidemiology: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/dc/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/dc/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/dc/favicon-16x16.png">
<link rel="manifest" href="../favicons/dc/site.webmanifest">
<link rel="mask-icon" href="../favicons/dc/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav data"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="dc-logo" alt="Data Carpentry" src="../assets/images/data-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="../assets/images/data-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr></ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="ncbi.html">1. NCBI</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="blast.html">2. BLAST</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="msa.html">3. Multiple sequence alignment</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="phylogenetics.html">4. Phylogenetics</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="quality-control.html">5. Reads, QC and trimming</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="sequence-assembly.html">6. Sequence assembly</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="snps-phylogeny.html">7. Mapping</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="data-visualization.html">8. Molecular epidemiology</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
</ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-ncbi"><p>Content from <a href="ncbi.html">NCBI</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/ncbi.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<section><h2 class="section-heading" id="exercise-1-global-cross-database-ncbi-search">Exercise 1: Global Cross-database NCBI Search<a class="anchor" aria-label="anchor" href="#exercise-1-global-cross-database-ncbi-search"></a>
</h2>
<hr class="half-width">
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<p>Exercise 1 - How do you efficiently and reliably navigate NCBI’s
website?</p>
<p>Exercise 2 - How do you search sequences? - How do you use the right
BLAST flavor?</p>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<p>Exercise 1 - Familiarize yourself with the differences NCBI databases
- Browse and search the databases, using cross-links</p>
<p>Exercise 2 - Use and understand blast alignment tools - Characterize
sequences - Understand genomic context and synteny</p>
</div>
</div>
</div>
</div>
</div>
<p>This part is to encourage you to explore NCBI resources. Questions
are examples or real-life questions that you might ask yourself later.
There are not necessarily exactly one solution to the question.</p>
<p>Start by going to NCBI’s web site, main search page: <a href="https://www.ncbi.nlm.nih.gov/search/" class="external-link uri">https://www.ncbi.nlm.nih.gov/search/</a>. There, you have
the list of most of NCBI’s databases, sorted by category. Take some time
to explore the following sections: Literature, Genomes, Genes and
Proteins.</p>
<div class="section level3">
<h3 id="task-1-1-find-publications-this-is-a-warm-up">Task 1.1: Find publications (this is a warm up!)<a class="anchor" aria-label="anchor" href="#task-1-1-find-publications-this-is-a-warm-up"></a>
</h3>
<p>You want to find the following articles:</p>
<ol style="list-style-type: decimal">
<li>an article about <em>Escherichia coli</em> O104:H4 published by
Matthew K Waldor in the New England Journal of Medicine; and</li>
<li>a paper that has <em>Escherichia coli</em> in the title, whose last
author is L. Wang, and was published in PLoS ONE in 2014.</li>
</ol>
<p>NOTE: you do not know the exact title, list of authors, PMID etc. Use
your search skills (solution provided below). There may be more than one
result.</p>
<div id="challenge-1.1.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.1.1" class="callout-inner">
<h3 class="callout-title">Challenge 1.1.1</h3>
<div class="callout-content">
<p>Find the two articles above using NCBI’s literature database.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>Use the advanced search from PubMed, by clicking on “Advanced”
(tutorial <a href="https://www.ncbi.nlm.nih.gov/books/NBK3827/#pubmedhelp.PubMed_Quick_Start" class="external-link uri">https://www.ncbi.nlm.nih.gov/books/NBK3827/#pubmedhelp.PubMed_Quick_Start</a>),
to build an exact search. Use the different fields to build a search
that returns only a single match. In particular, look at the <a href="https://pubmed.ncbi.nlm.nih.gov/help/#using-search-field-tags" class="external-link">Search
field tags</a>.</p>
</div>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<p>Use search terms with search tags (“search term”[searchTag]“)and
combine them with Boolean operators (”AND”).</p>
</div>
</div>
</div>
</div>
<div id="accordionHint3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint3" aria-expanded="false" aria-controls="collapseHint3">
  <h4 class="accordion-header" id="headingHint3"> Give me a hint </h4>
</button>
<div id="collapseHint3" class="accordion-collapse collapse" data-bs-parent="#accordionHint3" aria-labelledby="headingHint3">
<div class="accordion-body">
<p>Useful search tags: Author [au], Date of publication [dp], Journal
[ta], Last author [lastau], Title [ti].</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>Waldor[au] AND “Escherichia coli O104:H4” AND NEJM[ta]</li>
<li>“Wang L”[lastau] AND “Escherichia coli”[ti] AND 2014[dp] and “PLOS
One”[ta].</li>
</ol>
</div>
</div>
</div>
</div>
<div id="challenge-1.1.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1.1.2" class="callout-inner">
<h3 class="callout-title">Challenge 1.1.2</h3>
<div class="callout-content">
<p>Now try to find the most recent paper citing the first article by
Rasko et al above. Compare the use of</p>
<ul>
<li>NCBI (“Cited by” link) and</li>
<li>Google Scholar (<a href="https://scholar.google.com" class="external-link uri">https://scholar.google.com</a>) (the language settings of
Google Scholar can be adjusted in the menu under Settings) and</li>
<li>Web of Knowledge (<a href="https://www.webofknowledge.com" class="external-link uri">https://www.webofknowledge.com</a>; Access from UU: <a href="https://www.ub.uu.se/" class="external-link uri">https://www.ub.uu.se/</a> &gt;
Databases A-Z &gt; W &gt; Web of Science Core Collection).</li>
</ul>
<p>Then:</p>
<ul>
<li>How did the two search engines compare?</li>
<li>Which one was easier to work with dates?</li>
<li>Which one returned most results?</li>
<li>Which one has more complete information?</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-2-find-gene-records">Task 1.2: Find Gene records<a class="anchor" aria-label="anchor" href="#task-1-2-find-gene-records"></a>
</h3>
<p>You want to find sequences for the subunit A of the Shiga toxin in
<em>E. coli</em> O157:H7 Sakai. Try to search in the three following
databases: Gene, Nucleotide and Protein.</p>
<div id="challenge-1.2.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.2.1" class="callout-inner">
<h3 class="callout-title">Challenge 1.2.1</h3>
<div class="callout-content">
<p>What do the three different databases contain? What information do
you get from them?</p>
</div>
</div>
</div>
<div id="accordionHint4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint4" aria-expanded="false" aria-controls="collapseHint4">
  <h4 class="accordion-header" id="headingHint4"> Give me a hint </h4>
</button>
<div id="collapseHint4" class="accordion-collapse collapse" data-bs-parent="#accordionHint4" aria-labelledby="headingHint4">
<div class="accordion-body">
<p>NCBI is not as smart as Google, and copy/pasting might fail. Try
spelling out <em>Escherichia coli</em> and removing parts of the search
text.</p>
<p>If you get too many hits, go to the field “Search details”, and
refine your search, using the [Organism] field, for example. You don’t
need to get exactly one hit or to get the result you want on top. By
default, the search will include everything that has the species name
anywhere in the record.</p>
<p>Nucleotide database contains mostly <strong>genome</strong> records,
so the Shiga toxin gene might be anywhere on these contigs/genomes.</p>
</div>
</div>
</div>
</div>
<div id="challenge-1.2.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1.2.2" class="callout-inner">
<h3 class="callout-title">Challenge 1.2.2</h3>
<div class="callout-content">
<p>What do the five top results from the Protein database show?</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-3-understand-gene-records">Task 1.3: Understand Gene records<a class="anchor" aria-label="anchor" href="#task-1-3-understand-gene-records"></a>
</h3>
<p>Click on a Gene search result from Task 1.2 and skim through the
entire page.</p>
<div id="challenge-1.3.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3.1" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.1</h3>
<div class="callout-content">
<p>Who published the sequence? When? Is there a paper?</p>
</div>
</div>
</div>
<div id="accordionHint5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint5" aria-expanded="false" aria-controls="collapseHint5">
  <h4 class="accordion-header" id="headingHint5"> Give me a hint </h4>
</button>
<div id="collapseHint5" class="accordion-collapse collapse" data-bs-parent="#accordionHint5" aria-labelledby="headingHint5">
<div class="accordion-body">
<p>There is often a “Bibliography” section. There is also often a
“PubMed” link under “Related Information” in the right menu.</p>
</div>
</div>
</div>
</div>
<div id="challenge-1.3.2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3.2" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.2</h3>
<div class="callout-content">
<p>What is known about the gene? Does the record include taxonomy
information?</p>
</div>
</div>
</div>
<div id="accordionHint6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint6" aria-expanded="false" aria-controls="collapseHint6">
  <h4 class="accordion-header" id="headingHint6"> Give me a hint </h4>
</button>
<div id="collapseHint6" class="accordion-collapse collapse" data-bs-parent="#accordionHint6" aria-labelledby="headingHint6">
<div class="accordion-body">
<p>The taxonomic information is under the “Summary” section, under
“Organism” and “Lineage”.</p>
</div>
</div>
</div>
</div>
<p>Click on the Organism link and go to the Taxonomy Browser.</p>
<div id="challenge-1.3.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1.3.3" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.3</h3>
<div class="callout-content">
<p>What can you learn about species taxonomy there? What other useful
information is given on this page?</p>
</div>
</div>
</div>
<p>Back to the Gene page, have a look at the “Related information”
section in the right column.</p>
<div id="challenge-1.3.4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3.4" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.4</h3>
<div class="callout-content">
<p>Where can you go from there?</p>
</div>
</div>
</div>
<div id="accordionHint7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint7" aria-expanded="false" aria-controls="collapseHint7">
  <h4 class="accordion-header" id="headingHint7"> Give me a hint </h4>
</button>
<div id="collapseHint7" class="accordion-collapse collapse" data-bs-parent="#accordionHint7" aria-labelledby="headingHint7">
<div class="accordion-body">
<p>All the information in that section (“Related information”) is coming
through the Entrez system. In <code>GenBank</code> records, look for
<code>\dbxref</code> (database cross-reference) which provides the
Entrez links.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-4-understanding-protein-records-and-sequences">Task 1.4: Understanding Protein records and sequences<a class="anchor" aria-label="anchor" href="#task-1-4-understanding-protein-records-and-sequences"></a>
</h3>
<p>Click on the Protein link for this gene. Look at the content of one
of the records.</p>
<div id="challenge-1.4.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.4.1" class="callout-inner">
<h3 class="callout-title">Challenge 1.4.1</h3>
<div class="callout-content">
<p>What can you learn here? What information do you find here? What is
the name of this format? Are there any known protein domains in this
protein? Can you identify a link to these?</p>
</div>
</div>
</div>
<div id="accordionHint8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint8" aria-expanded="false" aria-controls="collapseHint8">
  <h4 class="accordion-header" id="headingHint8"> Give me a hint </h4>
</button>
<div id="collapseHint8" class="accordion-collapse collapse" data-bs-parent="#accordionHint8" aria-labelledby="headingHint8">
<div class="accordion-body">
<p>You can change the format by clicking on the top left corner.</p>
<p>Look also for “Related information” in the right menu and for the
<code>/db_xref</code> links in the main file.</p>
</div>
</div>
</div>
</div>
<div id="challenge-1.4.2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.4.2" class="callout-inner">
<h3 class="callout-title">Challenge 1.4.2</h3>
<div class="callout-content">
<p>Can you find only the sequence information in <code>FASTA</code>
format? How is the format organized? What is included in the header?
Discuss the advantages of the two formats. Can you find a definition of
both formats on the NCBI website or elsewhere?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>If you are stuck, here is the link to</p>
<ul>
<li>one Gene record: <a href="https://www.ncbi.nlm.nih.gov/gene/916678" class="external-link uri">https://www.ncbi.nlm.nih.gov/gene/916678</a>
</li>
<li>one Protein record: <a href="https://www.ncbi.nlm.nih.gov/protein/NP_311001.1" class="external-link uri">https://www.ncbi.nlm.nih.gov/protein/NP_311001.1</a>
</li>
</ul>
</div>
</div>
</div>
</div>
<div id="challenge-1.4.3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.4.3" class="callout-inner">
<h3 class="callout-title">Challenge 1.4.3</h3>
<div class="callout-content">
<p>Can you find a 3D protein structure for this protein?</p>
</div>
</div>
</div>
<div id="accordionHint9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint9" aria-expanded="false" aria-controls="collapseHint9">
  <h4 class="accordion-header" id="headingHint9"> Give me a hint </h4>
</button>
<div id="collapseHint9" class="accordion-collapse collapse" data-bs-parent="#accordionHint9" aria-labelledby="headingHint9">
<div class="accordion-body">
<p>Not all proteins have 3D structures. If your protein doesn’t have
one, try the following Shiga toxin:</p>
<p><a href="https://www.ncbi.nlm.nih.gov/protein/NP_311001.1" class="external-link uri">https://www.ncbi.nlm.nih.gov/protein/NP_311001.1</a></p>
</div>
</div>
</div>
</div>
<div id="accordionHint10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint10" aria-expanded="false" aria-controls="collapseHint10">
  <h4 class="accordion-header" id="headingHint10"> Give me a hint </h4>
</button>
<div id="collapseHint10" class="accordion-collapse collapse" data-bs-parent="#accordionHint10" aria-labelledby="headingHint10">
<div class="accordion-body">
<p>In the right menu, under “Related Information”, there may be a
“Related Structures” link.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>See here: <a href="https://www.ncbi.nlm.nih.gov/structure?LinkName=protein_structure_related_list&amp;from_uid=15832228" class="external-link uri">https://www.ncbi.nlm.nih.gov/structure?LinkName=protein_structure_related_list&amp;from_uid=15832228</a></p>
</div>
</div>
</div>
</div>
<div id="challenge-1.4.4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.4.4" class="callout-inner">
<h3 class="callout-title">Challenge 1.4.4</h3>
<div class="callout-content">
<p>Can you display the genome record on which the gene is encoded?</p>
</div>
</div>
</div>
<div id="accordionHint11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint11" aria-expanded="false" aria-controls="collapseHint11">
  <h4 class="accordion-header" id="headingHint11"> Give me a hint </h4>
</button>
<div id="collapseHint11" class="accordion-collapse collapse" data-bs-parent="#accordionHint11" aria-labelledby="headingHint11">
<div class="accordion-body">
<p>If you lost the gene page, here is another one:</p>
<p><a href="https://www.ncbi.nlm.nih.gov/gene/26516284" class="external-link uri">https://www.ncbi.nlm.nih.gov/gene/26516284</a></p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>There are several ways to display the GenBank file. - In the “Genomic
regions, transcripts, and products” section, there is a link to the
GenBank record - There is another one in the “Related sequences” section
(column “Accession and Version”).</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-2-sequence-alignments">Exercise 2: Sequence alignments<a class="anchor" aria-label="anchor" href="#exercise-2-sequence-alignments"></a>
</h2>
<hr class="half-width">
<p>In this exericse, you will use the sequence alignment tool BLAST to
align and retrieve sequences.</p>
<div id="caution1" class="callout caution">
<div class="callout-square">
<i class="callout-icon" data-feather="alert-triangle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Caution</h3>
<div class="callout-content">
<p>NOTE: the NCBI blast service can be under very heavy load (especially
during daytime in the US), and don’t be surprised if a single blast
takes &gt;10 minutes. As an alternative, you can use the BLAST service
at the EBI website (<a href="http://www.ebi.ac.uk" class="external-link uri">http://www.ebi.ac.uk</a>). Note that the EBI doesn’t have
the same level of tool integration as NCBI, and some parts of the
questions might be more challenging to answer with the EBI tools.</p>
<p>In general, try reduce the load on the NCBI server and consider
smaller databases than the standard <code>nr</code>/<code>nt</code>
ones, for example <code>refseq_select</code> or
<code>refseq_protein</code>: under the “Choose Search Set” heading on
the Blast search page, you can choose the right “Database”.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-2-1-find-nucleotide-blast-hits">Task 2.1: Find nucleotide blast hits<a class="anchor" aria-label="anchor" href="#task-2-1-find-nucleotide-blast-hits"></a>
</h3>
<p>Find the NCBI blast main page.</p>
<div id="challenge-2.1.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1.1" class="callout-inner">
<h3 class="callout-title">Challenge 2.1.1</h3>
<div class="callout-content">
<ul>
<li>What is BLAST?</li>
<li>What different kinds of BLAST searches are there?</li>
<li>What is blastn? blastx? blastp? tblastn?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint12" aria-expanded="false" aria-controls="collapseHint12">
  <h4 class="accordion-header" id="headingHint12"> Give me a hint </h4>
</button>
<div id="collapseHint12" class="accordion-collapse collapse" data-bs-parent="#accordionHint12" aria-labelledby="headingHint12">
<div class="accordion-body">
<p>See the different flavors of BLAST: <a href="https://blast.ncbi.nlm.nih.gov/doc/blast-quick-start-guide/" class="external-link uri">https://blast.ncbi.nlm.nih.gov/doc/blast-quick-start-guide/</a></p>
</div>
</div>
</div>
</div>
<p>Find sequence hits for the following sequence, but first consider the
options available on the BLAST page. One option that might come in very
handy is the “open results in another window” option, which will allow
you to modify your query and rerun it quickly without losing the
results:</p>
<pre><code>&gt;a
GATTACTTCAGCCAAAAGGAACACCTGTATATGAAGTGTATATTATTTAAATGGGTACTG
TGCCTGTTACTGGGTTTTTCTTCGGTATCCTATTCCCGGGAGTTTACGATAGACTTTTCG
ACCCAACAAAGTTATGTCTCTTCGTTAAATAGTATACGGACAGAGATATCGACCCCTCTT
GAACATATATCTCAGGGGACCACATCGGTGTCTGTTATTAACCACACCCCACCGGGCAGT
TATTTTGCTGTGGATATACGAGGGCTTGATGTCTATCAGGCGCGTTTTGACCATCTTCGT
CTGATTATTGAGCAAAATAATTTATATGTGGCCGGGTTCGTTAATACGGCAACAAATACT
TTCTACCGTTTTTCAGATTTTACACATATATCAGTGCCCGGTGTGACAACGGTTTCCATG
ACAACGGACAGCAGTTATACCACTCTGCAACGTGTCGCAGCGCTGGAACGTTCCGGAATG
CAAATCAGTCGTCACTCACTGGTTTCATCATATCTGGCGTTAATGGAGTTCAGTGGTAAT
ACAATGACCAGAGATGCATCCAGAGCAGTTCTGCGTTTTGTCACTGTCACAGCAGAAGCC
TTACGCTTCAGGCAGATACAGAGAGAATTTCGTCAGGCACTGTCTGAAACTGCTCCTGTG
TATACGATGACGCCGGGAGACGTGGACCTCACTCTGAACTGGGGGCGAATCAGCAATGTG
CTTCCGGAGTATCGGGGAGAGGATGGTGTCAGAGTGGGGAGAATATCCTTTAATAATATA
TCAGCGATACTGGGGACTGTGGCCGTTATACTGAATTGCCATCATCAGGGGGCGCGTTCT
GTTCGCGCCGTGAATGAAGAGAGTCAACCAGAATGTCAGATAACTGGCGACAGGCCCGTT
ATAAAAATAAACAATACATTATGGGAAAGTAATACAGCTGCAGCGTTTCTGAACAGAAAG
TCACAGTTTTTATATACAACGGGTAAATAAAGGAGTTAAGCATGAAGAAGATGTTTATGG</code></pre>
<div id="challenge-2.1.2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1.2" class="callout-inner">
<h3 class="callout-title">Challenge 2.1.2</h3>
<div class="callout-content">
<ul>
<li>Is it a gene?</li>
<li>If yes, what does it encode?</li>
<li>What is the closest organism? What can you infer about its taxonomic
distribution?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint13" aria-expanded="false" aria-controls="collapseHint13">
  <h4 class="accordion-header" id="headingHint13"> Give me a hint </h4>
</button>
<div id="collapseHint13" class="accordion-collapse collapse" data-bs-parent="#accordionHint13" aria-labelledby="headingHint13">
<div class="accordion-body">
<p>What does a protein-coding gene need to produce proteins?</p>
</div>
</div>
</div>
</div>
<p>Now go back to the blast main page.</p>
<div id="challenge-2.1.3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1.3" class="callout-inner">
<h3 class="callout-title">Challenge 2.1.3</h3>
<div class="callout-content">
<ul>
<li>Did you choose the right blast algorithm to answer the questions
above?</li>
<li>Can you do better?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint14" aria-expanded="false" aria-controls="collapseHint14">
  <h4 class="accordion-header" id="headingHint14"> Give me a hint </h4>
</button>
<div id="collapseHint14" class="accordion-collapse collapse" data-bs-parent="#accordionHint14" aria-labelledby="headingHint14">
<div class="accordion-body">
<p>What if you try another flavor of the blast suite?</p>
<p>What if you aligned the gene to the genome while keeping the codon
frame?</p>
</div>
</div>
</div>
</div>
<p>Repeat the procedure for this other sequence:</p>
<pre><code>&gt;b
GATTACTTCAGCCAAAAGGAACACCTGTATATGAAGTGTATATTATTTAAATGGGTACTG
TGCCTGTTACTGGGTTTTTCTTCGGTATCCTATTCCCGGGAGTTTACGATAGACTTTTCG
ACCCAACAAAGTTATGTCTCTTCGTTAAATAGTATACGGACAGAGATATCGACCCCTCTT
GAACATATATCTCAGGGGACCACATCGGTGTCTGTTATTAACCACACCCCACCGGGCAGT
TATTTTGCTGTGGATATACGAGGGCTTGATGTCTATCAGGCGCGTTTTGACCATCTTCGT
CTGATTATTGAGCAAAATAATTTATATGTGGCCGGGTTCGTTAATACGGCAACAAATACT
TTCTACCGTTTTTCAGATTTTACACATATATCAGTGCCCGGTGGACAACGGTTTCCATGA
CAACGGACAGCAGTTATACCACTCTGCAACGTGTCGCAGCGCTGGAACGTTCCGGAATGC
AAATCAGTCGTCACTCACTGGTTTCATCATATCTGGCGTTAATGGAGTTCAGTGGTAATA
CAATGACCAGAGATGCATCCAGAGCAGTTCTGCGTTTTGACTGTCACTGTCACAGCAGAA
GCCTTACGCTTCAGGCAGATACAGAGAGAATTTCGTCAGGCACTGTCTGAAACTGCTCCT
GTGTATACGATGACGCCGGGAGACGTGGACCTCTTTTTTTACTCTGAACTGGGGGCGAAT
CAGCAATGTGCTTCCGGAGTATCGGGGAGAGGATGGTGTCAGAGTGGGGAGAATATCCTT
TAATAATATATCAGCGATACTGGGGACTGTGGCCGTTATACTGAATTGCCATCATCAGGG
GGCGCGTTCTGTTCGCGCCGTGAATGAAGAGAGTCAACCAGAATGTCAGATAACTGGCGA
CAGGCCCGTTATAAAAATAAACAATACATTATGGGAAAGTAATACAGCTGCAGCGTTTCT
GAACAGAAAGTCACAGTTTTTATATACAACGGGTAAATAAAGGAGTTAAGCATGAAGAAG
ATGTTTATGG</code></pre>
<div id="challenge-2.1.4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1.4" class="callout-inner">
<h3 class="callout-title">Challenge 2.1.4</h3>
<div class="callout-content">
<p>How does sequence <code>b</code> compare to the first sequence? Is it
also a gene?</p>
</div>
</div>
</div>
<div id="accordionHint15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint15" aria-expanded="false" aria-controls="collapseHint15">
  <h4 class="accordion-header" id="headingHint15"> Give me a hint </h4>
</button>
<div id="collapseHint15" class="accordion-collapse collapse" data-bs-parent="#accordionHint15" aria-labelledby="headingHint15">
<div class="accordion-body">
<p>if you get stuck, try e.g. StarORF (<a href="http://star.mit.edu/orf/index.html" class="external-link uri">http://star.mit.edu/orf/index.html</a>).</p>
</div>
</div>
</div>
</div>
<div id="challenge-2.1.5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1.5" class="callout-inner">
<h3 class="callout-title">Challenge 2.1.5</h3>
<div class="callout-content">
<ul>
<li>Can you perform a pairwise alignment between both sequences?</li>
<li>What are the differences?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint16" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint16" aria-expanded="false" aria-controls="collapseHint16">
  <h4 class="accordion-header" id="headingHint16"> Give me a hint </h4>
</button>
<div id="collapseHint16" class="accordion-collapse collapse" data-bs-parent="#accordionHint16" aria-labelledby="headingHint16">
<div class="accordion-body">
<p>Check the “Align two or more sequences” on the BLAST search page.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-2-2-find-protein-blast-hits">Task 2.2: Find protein blast hits<a class="anchor" aria-label="anchor" href="#task-2-2-find-protein-blast-hits"></a>
</h3>
<p>Now find sequences with similarity to this protein</p>
<pre><code>&gt;c
MMNRSIYRQFVRYTIPTVAAMLVNGLYQVVDGIFIGHYVGAEGLAGINVAWPVIGTILGIGMLVGVGTGA
LASIKQGEGHPDSAKRILATGLMSLLLLAPIVAVILWFMADDFLRWQGAEGRVFELGLQYLQVLIVGCLF
TLGSIAMPFLLRNDHSPNLATLLMVIGALTNIALDYLFIAWLQWELTGAAIATTLAQVVVTLLGLAYFFS
ARANMRLTRRCLRLEWHSLPKIFAIGVSSFFMYAYGSTMVALHNTLMIEYGNAVMVGAYAVMGYIVTIYY
LTVEGIANGMQPLASYHFGARNYDHIRKLLRLAMGIAVLGGMAFVLVLNLFPEQAIAIFNSSDSELIAGA
QMGIKLHLFAMYLDGFLVVSAAYYQSTNRGGKAMFITVGNMSIMLPFLFILPQFFGLTGVWIALPISNIV
LSTVVGIMLWRDVNKMGKPTQVSYA</code></pre>
<div id="challenge-2.2.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.2.1" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.1</h3>
<div class="callout-content">
<ul>
<li>What is this gene coding for?</li>
<li>Where did it likely come from?</li>
</ul>
</div>
</div>
</div>
<div id="challenge-2.2.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.2.2" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.2</h3>
<div class="callout-content">
<ul>
<li>Can you align nucleotide sequences (sequence <code>a</code>, Task
2.1) against a protein database?</li>
<li>If so, how?</li>
</ul>
</div>
</div>
</div>
<div id="challenge-2.2.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.2.3" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.3</h3>
<div class="callout-content">
<ul>
<li>Can you align proteins to a nucleotide database?</li>
<li>If so, how? If not, why not?</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-2-3-find-gene-and-genome-hits">Task 2.3: Find gene and genome hits<a class="anchor" aria-label="anchor" href="#task-2-3-find-gene-and-genome-hits"></a>
</h3>
<p>This task should be performed at the NCBI page (not at EBI).</p>
<p>What is the following sequence?</p>
<pre><code>&gt;d
GTGTCTAAAGAAAAATTTGAACGTACAAAACCGCACGTTAACGTTGGTACTATCGGCCACGTTGACCACG
GTAAAACTACTCTGACCGCTGCAATCACCACCGTACTGGCTAAAACCTACGGCGGTGCTGCTCGTGCATT
CGACCAGATCGATAACGCGCCGGAAGAAAAAGCTCGTGGTATCACCATCAACACTTCTCACGTTGAATAT
GACACCCCGACCCGTCACTACGCACACGTAGACTGCCCGGGGCACGCCGACTATGTTAAAAACATGATCA
CCGGTGCTGCTCAGATGGACGGCGCGATCCTGGTAGTTGCTGCGACTGACGGCCCGATGCCGCAGACTCG
TGAGCACATCCTGCTGGGTCGTCAGGTAGGCGTTCCGTACATCATCGTGTTCCTGAACAAATGCGACATG
GTTGATGACGAAGAGCTGCTGGAACTGGTTGAAATGGAAGTTCGTGAACTTCTGTCTCAGTACGACTTCC
CGGGCGACGACACTCCGATCGTTCGTGGTTCTGCTCTGAAAGCGCTGGAAGGCGACGCAGAGTGGGAAGC
GAAAATCCTGGAACTGGCTGGCTTCCTGGATTCTTACATTCCGGAACCAGAGCGTGCGATTGACAAGCCG
TTCCTGCTGCCGATCGAAGACGTGTTCTCCATCTCCGGTCGTGGTACCGTTGTTACCGGTCGTGTAGAAC
GCGGTATCATCAAAGTTGGTGAAGAAGTTGAAATCGTTGGTATCAAAGAGACTCAGAAGTCTACCTGTAC
TGGCGTTGAAATGTTCCGCAAACTGCTGGACGAAGGCCGTGCTGGTGAGAACGTAGGTGTTCTGCTGCGT
GGTATCAAACGTGAAGAAATCGAACGTGGTCAGGTACTGGCTAAGCCGGGCACCATCAAACCGCACACCA
AGTTCGAATCTGAAGTGTACATTCTGTCCAAAGATGAAGGCGGCCGTCATACTCCGTTCTTCAAAGGCTA
CCGTCCGCAGTTCTACTTCCGTACTACTGACGTGACTGGTACCATCGAACTGCCGGAAGGCGTAGAGATG
GTAATGCCGGGCGACAACATCAAAATGGTTGTTACCCTGATCCACCCGATCGCGATGGACGACGGTCTGC
GTTTCGCAATCCGTGAAGGCGGCCGTACCGTTGGTGCGGGCGTTGTTGCTAAAGTTCTGGGCTAA</code></pre>
<div id="challenge-2.3.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.3.1" class="callout-inner">
<h3 class="callout-title">Challenge 2.3.1</h3>
<div class="callout-content">
<ul>
<li>What organism is it likely from?</li>
<li>Is it a gene?</li>
<li>If yes, which one?</li>
<li>What does it code for?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint17" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint17" aria-expanded="false" aria-controls="collapseHint17">
  <h4 class="accordion-header" id="headingHint17"> Give me a hint </h4>
</button>
<div id="collapseHint17" class="accordion-collapse collapse" data-bs-parent="#accordionHint17" aria-labelledby="headingHint17">
<div class="accordion-body">
<p>Use rather blastn to identify the origin of the gene. Explore the
alignments of the first hit, and try the “Graphics” button associated
with the results.</p>
</div>
</div>
</div>
</div>
<div id="challenge-2.3.2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.3.2" class="callout-inner">
<h3 class="callout-title">Challenge 2.3.2</h3>
<div class="callout-content">
<ul>
<li>Where is it located on the genome (precision at ~10kb is good
enough)?</li>
<li>Are there several copies in the genome?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint18" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint18" aria-expanded="false" aria-controls="collapseHint18">
  <h4 class="accordion-header" id="headingHint18"> Give me a hint </h4>
</button>
<div id="collapseHint18" class="accordion-collapse collapse" data-bs-parent="#accordionHint18" aria-labelledby="headingHint18">
<div class="accordion-body">
<p>Mark down the name of the gene (generally three letters in lower case
and one in uppercase). From the protein record, there is a link to the
genome record, which has a “graphics” interface.</p>
</div>
</div>
</div>
</div>
<div id="accordionHint19" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint19" aria-expanded="false" aria-controls="collapseHint19">
  <h4 class="accordion-header" id="headingHint19"> Give me a hint </h4>
</button>
<div id="collapseHint19" class="accordion-collapse collapse" data-bs-parent="#accordionHint19" aria-labelledby="headingHint19">
<div class="accordion-body">
<p>Not all genomes are equally well annotated and you might have to
search for a genome that is well annotated. One good example is
<em>Escherichia fergusonii</em> ATCC 35469</p>
</div>
</div>
</div>
</div>
<div id="challenge-2.3.3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.3.3" class="callout-inner">
<h3 class="callout-title">Challenge 2.3.3</h3>
<div class="callout-content">
<ul>
<li>How far is this gene (and possible copies) from the origin of
replication?</li>
<li>How is it oriented with respect to replication?</li>
</ul>
</div>
</div>
</div>
<div id="accordionHint20" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint20" aria-expanded="false" aria-controls="collapseHint20">
  <h4 class="accordion-header" id="headingHint20"> Give me a hint </h4>
</button>
<div id="collapseHint20" class="accordion-collapse collapse" data-bs-parent="#accordionHint20" aria-labelledby="headingHint20">
<div class="accordion-body">
<p>What genes are reliable markers for the origin of replication? Can
you find a paper about that? Now can you find that gene in the genome
above?</p>
</div>
</div>
</div>
</div>
<div id="accordionHint21" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint21" aria-expanded="false" aria-controls="collapseHint21">
  <h4 class="accordion-header" id="headingHint21"> Give me a hint </h4>
</button>
<div id="collapseHint21" class="accordion-collapse collapse" data-bs-parent="#accordionHint21" aria-labelledby="headingHint21">
<div class="accordion-body">
<p>What is the <em>dnaA</em> gene doing?</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-2-4-optional-explore-the-synteny-functions-of-the-string-database">Task 2.4 (optional): Explore the synteny functions of the STRING
database<a class="anchor" aria-label="anchor" href="#task-2-4-optional-explore-the-synteny-functions-of-the-string-database"></a>
</h3>
<p>We are interested in genes that are normally located next to each
other, and their taxonomic breadth.</p>
<p>Open the STRING database (<a href="https://string-db.org/" class="external-link uri">https://string-db.org/</a>). Search for the Shiga
toxin-encoding genes in bacteria.</p>
<div id="challenge-2.4.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.4.1" class="callout-inner">
<h3 class="callout-title">Challenge 2.4.1</h3>
<div class="callout-content">
<ul>
<li>Can you find the Shiga toxin-coding genes in <em>Escherichia
coli</em> K-12?</li>
<li>Do you find the same strain as above?</li>
</ul>
</div>
</div>
</div>
<div id="challenge-2.4.2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.4.2" class="callout-inner">
<h3 class="callout-title">Challenge 2.4.2</h3>
<div class="callout-content">
<p>Can you tell on what chromosomal feature these genes are located?</p>
</div>
</div>
</div>
<div id="accordionHint22" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint22" aria-expanded="false" aria-controls="collapseHint22">
  <h4 class="accordion-header" id="headingHint22"> Give me a hint </h4>
</button>
<div id="collapseHint22" class="accordion-collapse collapse" data-bs-parent="#accordionHint22" aria-labelledby="headingHint22">
<div class="accordion-body">
<p>What genes are located next to these genes?</p>
</div>
</div>
</div>
</div>
<div id="accordionHint23" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint23" aria-expanded="false" aria-controls="collapseHint23">
  <h4 class="accordion-header" id="headingHint23"> Give me a hint </h4>
</button>
<div id="collapseHint23" class="accordion-collapse collapse" data-bs-parent="#accordionHint23" aria-labelledby="headingHint23">
<div class="accordion-body">
<p>What is a lysogenic phage?</p>
</div>
</div>
</div>
</div>
<p>Feel free to explore the different functions of the database. What do
the different links in the network mean? What happens if you remove the
least reliable interaction sources (i.e. text mining and databases?</p>
<p>Now modify the settings to keep only Neighborhood as edges. Go back
to the Viewers tab. Search for the other gene discussed above,
<em>tufA</em>, do the same as for <em>stx1</em>, and compare the
results.</p>
<div id="challenge-2.4.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.4.3" class="callout-inner">
<h3 class="callout-title">Challenge 2.4.3</h3>
<div class="callout-content">
<ul>
<li>How many genes are co-localized with <em>stx1</em>?</li>
<li>What are the functions of these?</li>
</ul>
</div>
</div>
</div>
<div id="challenge-2.4.4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.4.4" class="callout-inner">
<h3 class="callout-title">Challenge 2.4.4</h3>
<div class="callout-content">
<ul>
<li>How many genes are co-localized with <em>tufA</em>?</li>
<li>What are the functions of the gene?</li>
<li>Does the function correlate with the taxonomic breadth of the
gene?</li>
</ul>
</div>
</div>
</div>
<p>For both genes, select the Neighborhood viewer.</p>
<div id="challenge-2.4.5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.4.5" class="callout-inner">
<h3 class="callout-title">Challenge 2.4.5</h3>
<div class="callout-content">
<ul>
<li>How taxonomically widespread are the <em>stx1</em>, respectively
<em>tufA</em> genes?</li>
<li>How conserved is the gene order for these two genes?</li>
</ul>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>NCBI website is very, very comprehensive</li>
<li>The Entrez cross-referencing system helps retrieving links bits of
data of different type.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-blast"><p>Content from <a href="blast.html">BLAST</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/blast.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to use the command-line version of <code>BLAST</code>?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Run different flavors of <code>BLAST</code>
</li>
<li>Use different databases</li>
<li>Use profile-based <code>BLAST</code> to retrieve distant
homologs</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>The goal of this exercise is to practice with the command line-based
version of <code>BLAST</code>. It is also the first step in building
phylogenetic trees, namely by gathering homologous sequences that will
then be aligned. The alignment is finally used to build a tree.</p>
<p>The whole exercise is based on RpoB, the <a href="https://en.wikipedia.org/wiki/RpoB" class="external-link">β-subunit of the bacterial RNA
polymerase</a>. This protein is essential to the cell, and present in
all bacteria and plastids. It is also very conserved, and thus suitable
for deep-scale phylogenies.</p>
<div class="section level3">
<h3 id="blast">BLAST<a class="anchor" aria-label="anchor" href="#blast"></a>
</h3>
<p><code>BLAST</code> (basic local alignment search tool, also referred
to as the Google of biological research) is one of the most widely used
tools in bioinformatics. The wide majority of its uses can be performed
online, but for larger searches, batches or advanced uses, the command
line version, performed locally on a computer or cluster, is
indispensable.</p>
</div>
<div class="section level3">
<h3 id="resources-at-uppmax">Resources at UPPMAX<a class="anchor" aria-label="anchor" href="#resources-at-uppmax"></a>
</h3>
<p><code>BLAST</code> is available at UPPMAX. To load the
<code>blast</code> module, you will need to load the
<code>bioinfo-tools</code> module first. The <code>blast</code> module
also loads the <code>blast_databases</code> module, to be able to use
the standard databases that NCBI maintains.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools blast</span></code></pre>
</div>
<p>Databases available at UPPMAX are described here: <a href="https://docs.uppmax.uu.se/databases/blast/" class="external-link uri">https://docs.uppmax.uu.se/databases/blast/</a>. The
<code>blast_databases</code> module implies that you don’t need to
specify where the databases are located on the file system. In detail,
it sets the <code>BLASTDB</code> variable to the right folder. You can
see it by typing <code>echo $BLASTDB</code> in the terminal.</p>
<p>Gene and protein records are usually associated with a
<code>taxid</code>, to describe what organisms they come from. This can
be very useful to limit the search to a certain taxon, or to exclude
another taxon. E.g. if you want to investigate whether a certain gene
has been transferred from bacteria to archaea: you would search for that
specific by excluding all bacteria and eukaryotes.</p>
<p>Taxonomy resources at UPPMAX are described here: <a href="https://docs.uppmax.uu.se/databases/ncbi/" class="external-link uri">https://docs.uppmax.uu.se/databases/ncbi/</a></p>
</div>
</section><section><h2 class="section-heading" id="exercise-0-login-to-uppmax">Exercise 0: Login to Uppmax<a class="anchor" aria-label="anchor" href="#exercise-0-login-to-uppmax"></a>
</h2>
<hr class="half-width">
<p>This is just a reminder of the introduction to Linux and how to login
to Uppmax.</p>
<div id="challenge-0.1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-0.1" class="callout-inner">
<h3 class="callout-title">Challenge 0.1</h3>
<div class="callout-content">
<p>Login to Uppmax and navigate to the course folder, create a folder
for a <code>blast</code> exercise.</p>
<p>Remember the best practices you learned for file naming.</p>
<p>All exercises should be performed inside
<code>/proj/g2020004/nobackup/3MK013/&lt;username&gt;</code> where
<code>&lt;username&gt;</code> is your own folder.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<p><code>ssh</code> is used to connect to an external server.
<code>-Y</code> forwards the graphical display to your computer. The
address of the server is <code>rackham.uppmax.uu.se</code>. You need to
add your user name in front of it, with a <code>@</code> in between.</p>
<p>The course folder is at
<code>/proj/g2020004/nobackup/3MK013/</code>.</p>
<p>There, make a folder with your username.</p>
<p>Inside it, create a <code>blast</code> folder for this exercise.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ssh</span> –Y username@rackham.uppmax.uu.se</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">mkdir</span> blast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-0.2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-0.2" class="callout-inner">
<h3 class="callout-title">Challenge 0.2</h3>
<div class="callout-content">
<p>Access an <code>interactive</code> session that is booked for us. The
session is <code>uppmax2025-3-4</code>. Use the <code>snowy</code>
cluster, for 4 hours.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> <span class="op">&lt;</span>project<span class="op">&gt;</span> -M <span class="op">&lt;</span>cluster<span class="op">&gt;</span> -t <span class="op">&lt;</span>hh:mm:ss<span class="op">&gt;</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-1-finding-and-retrieving-homologs">Exercise 1: Finding and retrieving homologs<a class="anchor" aria-label="anchor" href="#exercise-1-finding-and-retrieving-homologs"></a>
</h2>
<hr class="half-width">
<p>Here, you will first find the protein query to start your search
with. You will test different methods to create a file containing only
RpoB from <em>E. coli</em>. Then, you will use that file as a query to
find homologs of RpoB in different databases.</p>
<div class="section level3">
<h3 id="task-1-1-retrieve-the-rpob-sequence-for-e--coli">Task 1.1: Retrieve the RpoB sequence for <em>E. coli</em>
<a class="anchor" aria-label="anchor" href="#task-1-1-retrieve-the-rpob-sequence-for-e--coli"></a>
</h3>
<p>First, use your recently acquired <a href="https://en.wiktionary.org/wiki/-fu#English" class="external-link">NCBI-fu</a> to
retrieve the sequence from <em>E. coli</em> K-12. There are many
genomes, and thus many identical proteins, grouped into <a href="https://www.ncbi.nlm.nih.gov/ipg/" class="external-link">Identical Protein Groups</a>.
NCBI’s databases are (almost) non-redundant (that’s where the name of
the largest protein database, <code>nr</code> comes from), and thus only
one representative per IPG is present. Thus search in IPG, and find the
IPG’s accession number. Write down the accession number somewhere.</p>
<div id="challenge-1.1.1-finding-e.-colis-rpob" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.1.1-finding-e.-colis-rpob" class="callout-inner">
<h3 class="callout-title">Challenge 1.1.1: Finding <em>E. coli</em>’s
RpoB</h3>
<div class="callout-content">
<p>Find the accession number for the IPG containing the RpoB sequence of
<em>E. coli</em> K-12, and download the sequence in <code>FASTA</code>
format.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The accession number is <code>WP_000263098.1</code>.</p>
<p>The sequence in fasta format can be downloaded at <a href="https://www.ncbi.nlm.nih.gov/protein/" class="external-link uri">https://www.ncbi.nlm.nih.gov/protein/</a> by searching the
accession number clicking on the “Send to” link and then choosing
<code>FASTA</code> format. Save the file under
<code>rpoB_ecoli.fasta</code>.</p>
<p>The fasta file should look like:</p>
<pre><code>&gt;WP_000263098.1 MULTISPECIES: DNA-directed RNA polymerase subunit beta [Enterobacteriaceae]
MVYSYTEKKRIRKDFGKRPQVLDVPYLLSIQLDSFQKFIEQDPEGQYGLEAAFRSVFPIQSYSGNSELQY
VSYRLGEPVFDVQECQIRGVTYSAPLRVKLRLVIYEREAPEGTVKDIKEQEVYMGEIPLMTDNGTFVING</code></pre>
</div>
</div>
</div>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" aria-labelledby="headingInstructor2" data-bs-parent="#accordionInstructor2">
<div class="accordion-body">
<p>The IPG record can be found here: <a href="https://www.ncbi.nlm.nih.gov/ipg/1258700" class="external-link uri">https://www.ncbi.nlm.nih.gov/ipg/1258700</a></p>
<p>The fasta file can be found here: <a href="https://www.ncbi.nlm.nih.gov/protein/WP_000263098.1?report=fasta" class="external-link uri">https://www.ncbi.nlm.nih.gov/protein/WP_000263098.1?report=fasta</a></p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-2-push-the-sequence-to-uppmax">Task 1.2: Push the sequence to UPPMAX<a class="anchor" aria-label="anchor" href="#task-1-2-push-the-sequence-to-uppmax"></a>
</h3>
<p>The sequence is located on your computer, and you need to transfer it
to your computer. You will test several methods to push it to UPPMAX, to
be able to use it as a query for <code>BLAST</code>.</p>
<div id="challenge-1.2.1-use-scp" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.2.1-use-scp" class="callout-inner">
<h3 class="callout-title">Challenge 1.2.1: Use <code>scp</code>
</h3>
<div class="callout-content">
<p><code>scp</code>, secure file copy, is a tool to copy files via SSH,
the same protocol you use to login to UPPMAX. You can use it <em>on your
computer</em> with the following syntax:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>file to copy<span class="op">&gt;</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@<span class="op">&lt;</span>server<span class="op">&gt;</span>:<span class="op">&lt;</span>remote file location<span class="op">&gt;</span></span></code></pre>
</div>
<p>The remote file location can be a relative path from your home or an
absolute path, starting with <code>/</code>.</p>
<p>To bring up a local terminal on your Windows computer, click on the
“+” sign on the main window of MobaXTerm. If that doesn’t work, use
SFTP. Click on Session, SFTP, and fill in the Remote host as usual
<code>rackham.uppmax.uu.se</code> and your username. Navigate to
<code>/proj/g2020004/nobackup/3MK013/&lt;username&gt;/blast</code> on
the right panel, then drag and drop files between your computer and
Uppmax.</p>
</div>
</div>
</div>
<div id="accordionInstructor3" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor3" aria-expanded="false" aria-controls="collapseInstructor3">
  <h3 class="accordion-header" id="headingInstructor3">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor3" class="accordion-collapse collapse" aria-labelledby="headingInstructor3" data-bs-parent="#accordionInstructor3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">scp</span> rpoB_ecoli.fasta <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:/proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/blast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" aria-labelledby="headingHint2" data-bs-parent="#accordionHint2">
<div class="accordion-body">
<p>Use <code>man scp</code> to show the manual for scp.</p>
</div>
</div>
</div>
</div>
<div id="challenge-1.2.2-use-copypaste" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.2.2-use-copypaste" class="callout-inner">
<h3 class="callout-title">Challenge 1.2.2: Use copy/paste</h3>
<div class="callout-content">
<p>A quick and dirty method is to open a graphical text editor on the
remote server and paste the information in it. One such tool is
<code>gedit</code>. Paste the content of the file into
<code>gedit</code> and save it in the appropriate folder, under
<code>rpoB_ecoli.fasta</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>On UPPMAX</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/blast</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">gedit</span> rpoB_ecoli.fasta <span class="kw">&amp;</span></span></code></pre>
</div>
<p>The <code>&amp;</code> executes the program in the background,
leaving you control of the command line.</p>
</div>
</div>
</div>
</div>
<div id="challenge-1.2.3-extract-sequence-from-database" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.2.3-extract-sequence-from-database" class="callout-inner">
<h3 class="callout-title">Challenge 1.2.3: Extract sequence from database</h3>
<div class="callout-content">
<p>Since you know the accession number of the sequence to be used as a
query, you can use the <code>blast</code> tools to extract that specific
sequence from one of the databases available at UPPMAX. The tool in
question is <code>blastdbcmd</code>. That specific sequence is
presumably present in multiple databases. In doubt, one could search
<code>nr</code>. But since <em>E. coli</em> K-12 is present in the tiny
<a href="https://blast.ncbi.nlm.nih.gov/smartblast/smartBlast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs#searchSets" class="external-link"><code>landmark</code>
database</a>, you can use that one.</p>
<p>Put the sequence, in <code>FASTA</code> format, into a file called
<code>rpoB_ecoli.fasta</code></p>
<p>The manual for <code>blastdbcmd</code> can be obtained with:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-help</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionHint3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint3" aria-expanded="false" aria-controls="collapseHint3">
  <h4 class="accordion-header" id="headingHint3"> Give me a hint </h4>
</button>
<div id="collapseHint3" class="accordion-collapse collapse" aria-labelledby="headingHint3" data-bs-parent="#accordionHint3">
<div class="accordion-body">
<p>If you have closed your session, you may need to use
<code>module load</code> to load the appropriate module.</p>
<p>You may want to look into <code>-db</code> and <code>-entry</code>
arguments.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> <span class="op">&lt;</span>db name<span class="op">&gt;</span> -entry <span class="op">&lt;</span>accession<span class="op">&gt;</span> <span class="op">&gt;</span> <span class="op">&lt;</span>file<span class="op">&gt;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor4" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor4" aria-expanded="false" aria-controls="collapseInstructor4">
  <h3 class="accordion-header" id="headingInstructor4">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor4" class="accordion-collapse collapse" aria-labelledby="headingInstructor4" data-bs-parent="#accordionInstructor4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> landmark <span class="at">-entry</span> WP_000263098 <span class="op">&gt;</span> rpoB_ecoli.fasta</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionHint4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint4" aria-expanded="false" aria-controls="collapseHint4">
  <h4 class="accordion-header" id="headingHint4"> Give me a hint </h4>
</button>
<div id="collapseHint4" class="accordion-collapse collapse" aria-labelledby="headingHint4" data-bs-parent="#accordionHint4">
<div class="accordion-body">
<p>You should use the <code>-db</code> option and the
<code>-entry</code> one.</p>
</div>
</div>
</div>
</div>
<div id="challenge-1.2.4-optional-size-of-databases" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.2.4-optional-size-of-databases" class="callout-inner">
<h3 class="callout-title">Challenge 1.2.4 (optional): Size of databases</h3>
<div class="callout-content">
<p>Can you figure out a way to find the size of these two databases? How
does it affect the time to retrieve information from them? Is it worth
thinking about it?</p>
</div>
</div>
</div>
<div id="accordionHint5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint5" aria-expanded="false" aria-controls="collapseHint5">
  <h4 class="accordion-header" id="headingHint5"> Give me a hint </h4>
</button>
<div id="collapseHint5" class="accordion-collapse collapse" aria-labelledby="headingHint5" data-bs-parent="#accordionHint5">
<div class="accordion-body">
<p>You can look at the size of the databases by using
<code>blastdbcmd</code> and the flag <code>-info</code>. To see how much
time it takes to retrieve that single sequence from either database, use
the command <code>time</code> in front of the command:</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -info</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="bu">time</span> blastdbcmd <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -entry <span class="op">&lt;</span>accession<span class="op">&gt;</span> <span class="op">&gt;</span> /dev/null</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Database: Landmark database for SmartBLAST
	403,974 sequences; 229,101,880 total residues

Date: Oct 17, 2023  5:37 PM	Longest sequence: 35,991 residues

[...]

Database: All non-redundant GenBank CDS translations+PDB+SwissProt+PIR+PRF excluding environmental samples from WGS projects
	721,441,320 sequences; 277,730,601,621 total residues

Date: Feb 25, 2024  2:57 AM	Longest sequence: 98,182 residues

[...]
real	0m1.369s     # for landmark
[...]
real	0m10.059s.   # for nr
</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor5" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor5" aria-expanded="false" aria-controls="collapseInstructor5">
  <h3 class="accordion-header" id="headingInstructor5">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor5" class="accordion-collapse collapse" aria-labelledby="headingInstructor5" data-bs-parent="#accordionInstructor5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> landmark <span class="at">-info</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> nr <span class="at">-info</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="bu">time</span> blastdbcmd <span class="at">-db</span> landmark <span class="at">-entry</span> WP_000263098 <span class="op">&gt;</span> /dev/null</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="bu">time</span> blastdbcmd <span class="at">-db</span> nr <span class="at">-entry</span> WP_000263098 <span class="op">&gt;</span> /dev/null</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-3-finding-homologs-to-rpob-with-blast">Task 1.3: Finding homologs to RpoB with <code>BLAST</code>
<a class="anchor" aria-label="anchor" href="#task-1-3-finding-homologs-to-rpob-with-blast"></a>
</h3>
<p>Now we have a query to start our search. Check that the file is
present (<code>ls</code>) and that it looks like a <code>FASTA</code>
file (<code>less</code>).</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;WP_000263098.1 unnamed protein product [Escherichia coli] &gt;NP_312937.1 RNA polymerase beta subunit [Escherichia coli O157:H7 str. Sakai]
MVYSYTEKKRIRKDFGKRPQVLDVPYLLSIQLDSFQKFIEQDPEGQYGLEAAFRSVFPIQSYSGNSELQYVSYRLGEPVF
[...]</code></pre>
</div>
<p>The file header might look slightly different, depending on how you
obtained it. Now use <code>blastp</code> to find homologs of the RpoB
protein. Use the help for <code>blastp</code> to explore the options you
need to set.</p>
<div id="challenge-1.3.1-use-blastp-to-find-homologs-in-the-landmark-database" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3.1-use-blastp-to-find-homologs-in-the-landmark-database" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.1: Use <code>blastp</code> to
find homologs in the landmark database</h3>
<div class="callout-content">
<p>The path to the databases is already set correctly by the
<code>modules load blast</code> command. See <code>echo $BLASTDB</code>
to display it. Put the results into a file called
<code>rpoB_landmark.blast</code></p>
</div>
</div>
</div>
<div id="accordionHint6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint6" aria-expanded="false" aria-controls="collapseHint6">
  <h4 class="accordion-header" id="headingHint6"> Give me a hint </h4>
</button>
<div id="collapseHint6" class="accordion-collapse collapse" aria-labelledby="headingHint6" data-bs-parent="#accordionHint6">
<div class="accordion-body">
<p>Use the following command to find options. Explore
<code>-database</code></p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-help</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -query <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> <span class="op">&gt;</span> <span class="op">&lt;</span>output file<span class="op">&gt;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor6" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor6" aria-expanded="false" aria-controls="collapseInstructor6">
  <h3 class="accordion-header" id="headingInstructor6">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor6" class="accordion-collapse collapse" aria-labelledby="headingInstructor6" data-bs-parent="#accordionInstructor6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> landmark <span class="at">-query</span> rpoB_ecoli.fasta <span class="op">&gt;</span> rpoB_landmark.blast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Explore the results of the <code>blastp</code> command by displaying
the file in which you put the output of <code>blastp</code>. Inspect the
alignments, including those at the bottom of file, which have worse
alignments. In particular look at the E-values: do all hits look like
true homologs? How can you change that?</p>
<div id="challenge-1.3.2-use-blastp-to-find-better-homologs" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3.2-use-blastp-to-find-better-homologs" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.2: Use <code>blastp</code> to
find better homologs</h3>
<div class="callout-content">
<p>Find the right option to change the E-value threshold to include
hits. What is the default E-value? What does that mean?</p>
<p>Try to rerun the <code>blastp</code> search with a more reasonable
value than the default.</p>
</div>
</div>
</div>
<div id="accordionHint7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint7" aria-expanded="false" aria-controls="collapseHint7">
  <h4 class="accordion-header" id="headingHint7"> Give me a hint </h4>
</button>
<div id="collapseHint7" class="accordion-collapse collapse" aria-labelledby="headingHint7" data-bs-parent="#accordionHint7">
<div class="accordion-body">
<p>As a rule of thumb, hits that have a E-value &lt; 1e-6 are <em>bona
fide</em> homologs. Look at the <code>-evalue</code> option.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -query <span class="op">&lt;</span>fasta_file<span class="op">&gt;</span> -evalue <span class="op">&lt;</span>evalue<span class="op">&gt;</span> <span class="op">&gt;</span> output</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor7" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor7" aria-expanded="false" aria-controls="collapseInstructor7">
  <h3 class="accordion-header" id="headingInstructor7">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor7" class="accordion-collapse collapse" aria-labelledby="headingInstructor7" data-bs-parent="#accordionInstructor7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> landmark <span class="at">-query</span> rpoB_ecoli.fasta <span class="at">-evalue</span> 1e-6 <span class="op">&gt;</span> rpoB_landmark.blast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Explore the results again. Is it better, even the last
alignments?</p>
<p>The default format of the <code>blast</code> output is
“human-friendly”, something that resembles the output that is created on
the NCBI website. To produce an output that is even closer to NCBI’s
output, use the <code>-html</code> option.</p>
<p>The default format is fine for visual inspection, but not very
convenient for computers to read. For example, in some cases you might
want to perform data analysis on the output, e.g. to sort the results
further or to compare the output of different runs. In that case, a
tabular-like output is to prefer. The main option is
<code>-outfmt</code>, and setting it to <code>6</code> will produce such
a tabular output. The columns can be further specified, see the
manual.</p>
<div id="challenge-1.3.3-explore-the-output-format-of-blast" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3.3-explore-the-output-format-of-blast" class="callout-inner">
<h3 class="callout-title">Challenge 1.3.3: Explore the output format of
<code>BLAST</code>
</h3>
<div class="callout-content">
<p>Play with the different options for outfmt, the different formats and
how to customize the tabular formats (6, 7 and 10).</p>
<p>Finally, produce a standard tabular output for the same run as above
and put the result into a file called <code>rpoB_landmark.tab</code>.
This file can then downloaded to your own computer and opened with
<code>R</code> or with Excel.</p>
<p>Note that the query sometimes yields several hits in the same query
protein. This is caused by the protein being long and having its
different domains being separated by less conserved domains.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p><code>blastp</code> command:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -query <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> -evalue <span class="op">&lt;</span>e-value<span class="op">&gt;</span> -outfmt <span class="op">&lt;</span>number<span class="op">&gt;</span> <span class="op">&gt;</span> <span class="op">&lt;</span>output_file<span class="op">&gt;</span></span></code></pre>
</div>
<p>To import the file to your computer, to the current directory, use
the same <code>scp</code> program as above. <strong>This should be run
on your own computer, not from UPPMAX</strong>.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:<span class="op">&lt;</span>course base folde<span class="op">&gt;</span>/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/blast/<span class="op">&lt;</span>file<span class="op">&gt;</span> .</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor8" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor8" aria-expanded="false" aria-controls="collapseInstructor8">
  <h3 class="accordion-header" id="headingInstructor8">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor8" class="accordion-collapse collapse" aria-labelledby="headingInstructor8" data-bs-parent="#accordionInstructor8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> landmark <span class="at">-query</span> rpoB_ecoli.fasta <span class="at">-evalue</span> 1e-6 <span class="at">-outfmt</span> 6 <span class="op">&gt;</span> rpoB_landmark.tab</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:/proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/blast/rpoB_landmark.tab .</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-4-use-a-larger-database">Task 1.4: Use a larger database<a class="anchor" aria-label="anchor" href="#task-1-4-use-a-larger-database"></a>
</h3>
<p>So far you have used the <code>landmark</code> database, which is
tiny. Now, use a different, larger database, to gather more results from
a broader set of bacteria. Run a <code>blastp</code> again. How many
significant hits do you find in these?</p>
<div id="challenge-1.4.1-larger-database" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.4.1-larger-database" class="callout-inner">
<h3 class="callout-title">Challenge 1.4.1: Larger database</h3>
<div class="callout-content">
<p><code>refseq_select_prot</code> and <code>refseq_protein</code> are
good candidates. The former is smaller than the latter. Note that these
two runs will take time, so run only the one to
<code>refseq_select_prot</code>. We will run it in the background
(adding a <code>&amp;</code> at the end of the command) so that you can
continue with other tasks and come back to that one later.</p>
<p>You can check whether <code>blastp</code> is still running by typing
<code>ps</code>:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">ps</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  PID TTY          TIME CMD
13920 pts/64   00:00:00 bash
33117 pts/64   00:00:00 blastp
35427 pts/64   00:00:00 ps</code></pre>
</div>
<p>If you see <code>blastp</code> there, it means it is still running.
It could take up to 30 minutes. When <code>blastp</code> is done, count
the number of hits obtained.</p>
</div>
</div>
</div>
<div id="accordionHint8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint8" aria-expanded="false" aria-controls="collapseHint8">
  <h4 class="accordion-header" id="headingHint8"> Give me a hint </h4>
</button>
<div id="collapseHint8" class="accordion-collapse collapse" aria-labelledby="headingHint8" data-bs-parent="#accordionHint8">
<div class="accordion-body">
<p>The list of locally available databases is listed here: <a href="https://www.uppmax.uu.se/resources/databases/blast-databases/" class="external-link uri">https://www.uppmax.uu.se/resources/databases/blast-databases/</a></p>
</div>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -query <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> -evalue <span class="op">&lt;</span>e-value<span class="op">&gt;</span> -outfmt 6 <span class="op">&gt;</span> rpoB_<span class="op">&lt;</span>db<span class="op">&gt;</span>.tab <span class="kw">&amp;</span></span></code></pre>
</div>
<p>When it has finished (it may take up to 30 minutes), count the rows
to have the number of hits:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> rpoB_<span class="op">&lt;</span>db<span class="op">&gt;</span>.tab</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>500 rpoB_refseq_select.tab</code></pre>
</div>
<p>You have actually hit the maximum number of aligned sequences to keep
by default (500). You could change that by using the
<code>-max_target_seqs</code> option and setting it to a larger
number.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor9" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor9" aria-expanded="false" aria-controls="collapseInstructor9">
  <h3 class="accordion-header" id="headingInstructor9">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor9" class="accordion-collapse collapse" aria-labelledby="headingInstructor9" data-bs-parent="#accordionInstructor9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-db</span> refseq_select_prot <span class="at">-query</span> rpoB_ecoli.fasta <span class="at">-evalue</span> 1e-6 <span class="at">-outfmt</span> 6 <span class="op">&gt;</span> rpoB_refseq_select.tab <span class="kw">&amp;</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> rpoB_refseq_select.tab</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-1-5-extract-the-sequences-from-the-database">Task 1.5: Extract the sequences from the database<a class="anchor" aria-label="anchor" href="#task-1-5-extract-the-sequences-from-the-database"></a>
</h3>
<p>You found hits, i.e. proteins that show similarity with the query
protein. To prepare for the next steps (multiple sequence alignment and
building trees), you will need to retrieve the actual proteins and put
them in a <code>FASTA</code> file. The simplest way is to directly
extract them from the database, using the same tool
(<code>blastdbcmd</code>) as above. You will first need to extract a
list of the accession numbers from the blast results.</p>
<div id="challenge-1.5.1-extract-the-sequences-with-blastdbcmd" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.5.1-extract-the-sequences-with-blastdbcmd" class="callout-inner">
<h3 class="callout-title">Challenge 1.5.1: Extract the sequences with
<code>blastdbcmd</code>
</h3>
<div class="callout-content">
<p>Use <code>blastdbcmd</code> to retrieve the accession ids from the
<code>landmark</code> database. To prepare the list of proteins to
extract, <code>cut</code> the blast results to keep the accession number
of the hits. As you noticed above, there are multiple hits per protein
and one hit per line, resulting in each subject protein being possibly
present several times. You will need to produce a non-redundant list of
ids.</p>
</div>
</div>
</div>
<div id="accordionHint9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint9" aria-expanded="false" aria-controls="collapseHint9">
  <h4 class="accordion-header" id="headingHint9"> Give me a hint </h4>
</button>
<div id="collapseHint9" class="accordion-collapse collapse" aria-labelledby="headingHint9" data-bs-parent="#accordionHint9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-help</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="fu">man</span> uniq</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a><span class="fu">man</span> sort</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionHint10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint10" aria-expanded="false" aria-controls="collapseHint10">
  <h4 class="accordion-header" id="headingHint10"> Give me a hint </h4>
</button>
<div id="collapseHint10" class="accordion-collapse collapse" aria-labelledby="headingHint10" data-bs-parent="#accordionHint10">
<div class="accordion-body">
<p>The second column of the default tabular output provides the
accession id. <code>cut</code> the tabular output of the blast file,
pipe it to <code>sort</code> and then to <code>uniq</code>, and send the
result in a file. Then use that file as a input to <code>blastcmd</code>
to extract the proteins. Last time we used <code>-entry</code> because
we had just one, but there is another argument that takes a list of
entries instead. Put the result in the <code>rpoB_homologs.fasta</code>
file.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" aria-labelledby="headingSolution10" data-bs-parent="#accordionSolution10">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2</span> <span class="op">&lt;</span>blast_output_file<span class="op">&gt;</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> <span class="op">&lt;</span>id_file<span class="op">&gt;</span> </span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -entry_batch <span class="op">&lt;</span>id_file<span class="op">&gt;</span> <span class="op">&gt;</span> rpoB_homologs.fasta</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor10" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor10" aria-expanded="false" aria-controls="collapseInstructor10">
  <h3 class="accordion-header" id="headingInstructor10">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor10" class="accordion-collapse collapse" aria-labelledby="headingInstructor10" data-bs-parent="#accordionInstructor10">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2</span> rpoB_landmark.tab <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="op">&gt;</span> rpoB_landmark_ids </span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> landmark <span class="at">-entry_batch</span> rpoB_landmark_ids <span class="op">&gt;</span> rpoB_homologs.fasta</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-1.5.2-count-the-sequences" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.5.2-count-the-sequences" class="callout-inner">
<h3 class="callout-title">Challenge 1.5.2: Count the sequences</h3>
<div class="callout-content">
<p>Can you count how many sequences were included in the fasta file? Use
<code>grep</code> and your knowledge of the <code>FASTA</code>
format.</p>
</div>
</div>
</div>
<div id="accordionHint11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint11" aria-expanded="false" aria-controls="collapseHint11">
  <h4 class="accordion-header" id="headingHint11"> Give me a hint </h4>
</button>
<div id="collapseHint11" class="accordion-collapse collapse" aria-labelledby="headingHint11" data-bs-parent="#accordionHint11">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">man</span> grep</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="fu">man</span> wc</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution11" aria-expanded="false" aria-controls="collapseSolution11">
  <h4 class="accordion-header" id="headingSolution11"> Show me the solution </h4>
</button>
<div id="collapseSolution11" class="accordion-collapse collapse" aria-labelledby="headingSolution11" data-bs-parent="#accordionSolution11">
<div class="accordion-body">
<p>You can take a look at the sequences that were included in the
blast:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> <span class="op">&lt;</span>fasta_file<span class="op">&gt;</span> <span class="kw">|</span> <span class="fu">less</span></span></code></pre>
</div>
<p>And then count them by piping the result to <code>wc</code>
instead.</p>
<p>There should be around 70 sequences.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">70</span></span></code></pre>
</div>
<p>For the upcoming episode on multiple sequence alignment, you will use
a subset of these.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor11" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor11" aria-expanded="false" aria-controls="collapseInstructor11">
  <h3 class="accordion-header" id="headingInstructor11">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor11" class="accordion-collapse collapse" aria-labelledby="headingInstructor11" data-bs-parent="#accordionInstructor11">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> rpoB_homologs.fasta <span class="kw">|</span> <span class="fu">less</span></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> rpoB_homologs.fasta <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-2-using-psi-blast-to-retrieve-distant-homologs-of-proteins">Exercise 2: Using PSI-BLAST to retrieve distant homologs of
proteins<a class="anchor" aria-label="anchor" href="#exercise-2-using-psi-blast-to-retrieve-distant-homologs-of-proteins"></a>
</h2>
<hr class="half-width">
<p>In this exercise, you will use another flavor of <code>BLAST</code>
to retrieve distant homologs of a protein. As an example, we will use
the protein RavC, present in the genomes of <a href="https://en.wikipedia.org/wiki/Legionella" class="external-link"><em>Legionella</em>
bacteria</a>. This protein is an effector protein, injected by
<em>Legionella</em> into the cytoplasm of their host (protists). The
exact function is unknown, but it is presumably important, as it is
conserved throughout the whole order <em>Legionellales</em>. Many
effectors found in this group are derived from eukaryotic proteins, and
this is what you will test here: does RavC have a homolog in
eukarotes?</p>
<p>The strategy is to use <code>psiblast</code>, which uses - instead of
a single query - a matrix of amino-acid frequencies as a query.
<code>psiblast</code> is an iterative program: you generally start with
one sequence, gather <em>bona fide</em> homologs, use the profile of
these to query the database again, gather new homologs, recalculate the
profile, then re-query the database, etc. The search may converge after
a certain number of iterations, i.e. there no more new homologs to find
with the latest profile.</p>
<p>In this case, you will use a slightly different strategy: you will
start with one sequence, align it to <code>refseq_select_prot</code> but
only to sequences in the order <em>Legionellales</em>, using the
<code>taxids</code> options that limits the taxonomic scope of the
search. You will save the profile (PSSM) generated by the first
<code>psiblast</code> round and reuse this to then query the whole
database, with three iterations.</p>
<div class="section level3">
<h3 id="task-2-1-extract-ravc-from-a-database">Task 2.1: Extract RavC from a database<a class="anchor" aria-label="anchor" href="#task-2-1-extract-ravc-from-a-database"></a>
</h3>
<p>As above, you will extract the sequence of RavC from a database. This
time you will use <code>refseq_select_prot</code>, since there are no
<em>Legionella</em> in <code>landmark</code>. The accession number of
RavC that you will use is <a href="https://www.ncbi.nlm.nih.gov/protein/WP_010945868.1" class="external-link"><code>WP_010945868.1</code></a>.</p>
<div id="challenge-2.1.1-extract-ravc" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1.1-extract-ravc" class="callout-inner">
<h3 class="callout-title">Challenge 2.1.1: Extract RavC</h3>
<div class="callout-content">
<p>Retrieve sequence <code>WP_010945868.1</code> from the
<code>refseq_protein</code> database and put it into a file called
<code>ravC_LP.fasta</code>. Check the content of the <code>FASTA</code>
file.</p>
</div>
</div>
</div>
<div id="accordionHint12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint12" aria-expanded="false" aria-controls="collapseHint12">
  <h4 class="accordion-header" id="headingHint12"> Give me a hint </h4>
</button>
<div id="collapseHint12" class="accordion-collapse collapse" aria-labelledby="headingHint12" data-bs-parent="#accordionHint12">
<div class="accordion-body">
<p><code>blastdbcmd -help</code></p>
</div>
</div>
</div>
</div>
<div id="accordionSolution12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution12" aria-expanded="false" aria-controls="collapseSolution12">
  <h4 class="accordion-header" id="headingSolution12"> Show me the solution </h4>
</button>
<div id="collapseSolution12" class="accordion-collapse collapse" aria-labelledby="headingSolution12" data-bs-parent="#accordionSolution12">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> <span class="op">&lt;</span>db<span class="op">&gt;</span> -entry <span class="op">&lt;</span>accession<span class="op">&gt;</span> <span class="op">&gt;</span> <span class="op">&lt;</span>output_file<span class="op">&gt;</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;WP_010945868.1 RMD1 family protein [Legionella pneumophila] &gt;ERH46094.1 hypothetical protein N750_05210 [Legionella pneumophila str. Leg01/53] &gt;ERH46577.1 hypothetical protein N751_07400 [Legionella pneumophila str. Leg01/11] &gt;ERI48669.1 hypothetical protein N749_09265 [Legionella pneumophila str. Leg01/20] &gt;MFO2512789.1 RMD1 family protein [Legionella pneumophila serogroup 2] &gt;MFO2594790.1 RMD1 family protein [Legionella pneumophila serogroup 3] &gt;MFO2645706.1 RMD1 family protein [Legionella pneumophila serogroup 8] &gt;MFO2989133.1 RMD1 family protein [Legionella pneumophila serogroup 6] &gt;MFO3234997.1 RMD1 family protein [Legionella pneumophila serogroup 5] &gt;MFO3476243.1 RMD1 family protein [Legionella pneumophila serogroup 7] &gt;MFO8588967.1 RMD1 family protein [Legionella pneumophila serogroup 14] &gt;MFO8774718.1 RMD1 family protein [Legionella pneumophila serogroup 10] &gt;MFP3789783.1 RMD1 family protein [Legionella pneumophila serogroup 9]
MECLSFCVAKTIDLTRLDLHLKNVSKEFSAVKTRDVIRLNSHRNKDHTLFIFKNGTVVSWGVKRYQIHEYLDIIKLLVDK
PVALLVHDEFHYQIGDKTAIEPHGFYDVDCLTIEEDSDELKLSLSYGFSQSVKLQYFETIIDALIEKYNPLIQALSHKGE
MPISRKQIQQVIGEILGAKSELNLISNFLYHPKYFWQHPTLEEHFSMLERYLHIQRRVNAINHRLDTLNEIFDMFNGYLE
SRHGHHLEIIIIVLIAVEIIIAVMNFHF</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor12" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor12" aria-expanded="false" aria-controls="collapseInstructor12">
  <h3 class="accordion-header" id="headingInstructor12">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor12" class="accordion-collapse collapse" aria-labelledby="headingInstructor12" data-bs-parent="#accordionInstructor12">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="ex">blastdbcmd</span> <span class="at">-db</span> refseq_protein <span class="at">-entry</span> WP_010945868.1 <span class="op">&gt;</span> ravC_LP.fasta</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-2-2-align-ravc-to-sequences-belonging-to-legionellales">Task 2.2: Align RavC to sequences belonging to
<em>Legionellales</em>
<a class="anchor" aria-label="anchor" href="#task-2-2-align-ravc-to-sequences-belonging-to-legionellales"></a>
</h3>
<p>That task is a bit complex, so let’s break it down in several steps.
You will:</p>
<ol style="list-style-type: decimal">
<li>align the RavC sequence to the <code>refseq_select_prot</code>
database, using <code>psiblast</code>, and put the result into the file
<code>ravC_Leg.psiblast</code>.</li>
<li>save the <a href="https://en.wikipedia.org/wiki/Position_weight_matrix" class="external-link">PSSM (the
amino-acid profile built by psiblast)</a> after the last round, both in
its “native” form and in text format.</li>
<li>filter hits so that only hits with E-value &lt; 1e-6 are shown</li>
<li>filter hits so that only hits with E-value &lt; 1e-10 are included
in the PSSM</li>
<li>filter hits so that only hits belonging to the order
<em>Legionelalles</em> are included</li>
</ol>
<div id="challenge-2.2.1-psiblast" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.2.1-psiblast" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.1: Psiblast</h3>
<div class="callout-content">
<p>Build the command, using the <code>psiblast</code> command, the query
you extracted above and the <code>refseq_select_prot</code>
database.</p>
<div id="caution1" class="callout caution">
<div class="callout-square">
<i class="callout-icon" data-feather="alert-triangle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Caution</h3>
<div class="callout-content">
<p>Don’t run the command yet, as this will run for a while! We are just
building the command now. Wait for the end of the section.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="accordionHint13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint13" aria-expanded="false" aria-controls="collapseHint13">
  <h4 class="accordion-header" id="headingHint13"> Give me a hint </h4>
</button>
<div id="collapseHint13" class="accordion-collapse collapse" aria-labelledby="headingHint13" data-bs-parent="#accordionHint13">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-help</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution13" aria-expanded="false" aria-controls="collapseSolution13">
  <h4 class="accordion-header" id="headingSolution13"> Show me the solution </h4>
</button>
<div id="collapseSolution13" class="accordion-collapse collapse" aria-labelledby="headingSolution13" data-bs-parent="#accordionSolution13">
<div class="accordion-body">
<p>That is the base of the command:</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> -db <span class="op">&lt;</span>db<span class="op">&gt;</span> <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor13" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor13" aria-expanded="false" aria-controls="collapseInstructor13">
  <h3 class="accordion-header" id="headingInstructor13">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor13" class="accordion-collapse collapse" aria-labelledby="headingInstructor13" data-bs-parent="#accordionInstructor13">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> ravC_LP.fasta <span class="at">-db</span> refseq_select_prot <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2.2.2-saving-the-pssm" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.2.2-saving-the-pssm" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.2: Saving the PSSM</h3>
<div class="callout-content">
<p>Now add the options to save the PSSM after the last round, and save
the PSSM both as binary and ascii form. Add these options to the command
above, but you will only run it when it is finished.</p>
</div>
</div>
</div>
<div id="accordionHint14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint14" aria-expanded="false" aria-controls="collapseHint14">
  <h4 class="accordion-header" id="headingHint14"> Give me a hint </h4>
</button>
<div id="collapseHint14" class="accordion-collapse collapse" aria-labelledby="headingHint14" data-bs-parent="#accordionHint14">
<div class="accordion-body">
<p>In the PSI-blast help there is a “PSI-BLAST options” section.</p>
<p>For long help pages, it is sometimes useful to pipe the help message
to <code>grep</code> and use a combination of <code>-A</code> (after),
<code>-B</code> (before), and <code>--color</code> (color matches).</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-help</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-help</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">--color</span> <span class="at">-B</span> 5 <span class="at">-A</span> 2 pssm</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution14" aria-expanded="false" aria-controls="collapseSolution14">
  <h4 class="accordion-header" id="headingSolution14"> Show me the solution </h4>
</button>
<div id="collapseSolution14" class="accordion-collapse collapse" aria-labelledby="headingSolution14" data-bs-parent="#accordionSolution14">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> <span class="op">&lt;</span>fasta_file<span class="op">&gt;</span> -db <span class="op">&lt;</span>db<span class="op">&gt;</span> -save_pssm_after_last_round <span class="at">-out_pssm</span> <span class="op">&lt;</span>pssm_file<span class="op">&gt;</span> -out_ascii_pssm <span class="op">&lt;</span>pssm_file.txt<span class="op">&gt;</span> <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor14" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor14" aria-expanded="false" aria-controls="collapseInstructor14">
  <h3 class="accordion-header" id="headingInstructor14">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor14" class="accordion-collapse collapse" aria-labelledby="headingInstructor14" data-bs-parent="#accordionInstructor14">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> ravC_LP.fasta <span class="at">-db</span> refseq_select_prot <span class="at">-save_pssm_after_last_round</span> <span class="at">-out_pssm</span> ravC_Leg.pssm <span class="at">-out_ascii_pssm</span> ravC_Leg.pssm.txt <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2.2.3-thresholds" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.2.3-thresholds" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.3: Thresholds</h3>
<div class="callout-content">
<p>Now add the E-value thresholds. You want to display hits with E-value
&lt; 1e-6, but only include those with E-value &lt; 1e-10. Add the
options to the rest.</p>
</div>
</div>
</div>
<div id="accordionSolution15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution15" aria-expanded="false" aria-controls="collapseSolution15">
  <h4 class="accordion-header" id="headingSolution15"> Show me the solution </h4>
</button>
<div id="collapseSolution15" class="accordion-collapse collapse" aria-labelledby="headingSolution15" data-bs-parent="#accordionSolution15">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> <span class="op">&lt;</span>fasta_file<span class="op">&gt;</span> -db <span class="op">&lt;</span>db<span class="op">&gt;</span> -save_pssm_after_last_round <span class="at">-out_pssm</span> <span class="op">&lt;</span>pssm_file<span class="op">&gt;</span> -out_ascii_pssm <span class="op">&lt;</span>pssm_file.txt<span class="op">&gt;</span> -inclusion_ethresh <span class="op">&lt;</span>evalue<span class="op">&gt;</span> -evalue <span class="op">&lt;</span>evalue<span class="op">&gt;</span> <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor15" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor15" aria-expanded="false" aria-controls="collapseInstructor15">
  <h3 class="accordion-header" id="headingInstructor15">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor15" class="accordion-collapse collapse" aria-labelledby="headingInstructor15" data-bs-parent="#accordionInstructor15">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> ravC_LP.fasta <span class="at">-db</span> refseq_select_prot <span class="at">-save_pssm_after_last_round</span> <span class="at">-out_pssm</span> ravC_Leg.pssm <span class="at">-out_ascii_pssm</span> ravC_Leg.pssm.txt <span class="at">-inclusion_ethresh</span> 1e-10 <span class="at">-evalue</span> 1e-6 <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2.2.4-taxonomic-range" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.2.4-taxonomic-range" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.4: Taxonomic range</h3>
<div class="callout-content">
<p>Now limit the results to the order <em>Legionellales</em>. Find the
taxid of this group on the NCBI website.</p>
<p>Note: this feature has been upgraded in the latest version of
<code>BLAST</code> (2.15.0+). It is now possible to include all
descendants of a single taxid, using only that taxid. In previous
versions of <code>BLAST</code>, you had to include all descending taxids
using a script.</p>
</div>
</div>
</div>
<div id="accordionHint15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint15" aria-expanded="false" aria-controls="collapseHint15">
  <h4 class="accordion-header" id="headingHint15"> Give me a hint </h4>
</button>
<div id="collapseHint15" class="accordion-collapse collapse" aria-labelledby="headingHint15" data-bs-parent="#accordionHint15">
<div class="accordion-body">
<p>Check out the Taxonomy section of NCBI: <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/" class="external-link uri">https://www.ncbi.nlm.nih.gov/Taxonomy/</a></p>
</div>
</div>
</div>
</div>
<div id="accordionSolution16" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution16" aria-expanded="false" aria-controls="collapseSolution16">
  <h4 class="accordion-header" id="headingSolution16"> Show me the solution </h4>
</button>
<div id="collapseSolution16" class="accordion-collapse collapse" aria-labelledby="headingSolution16" data-bs-parent="#accordionSolution16">
<div class="accordion-body">
<p><a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=118969&amp;lvl=3&amp;lin=f&amp;keep=1&amp;srchmode=1&amp;unlock" class="external-link uri">https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=118969&amp;lvl=3&amp;lin=f&amp;keep=1&amp;srchmode=1&amp;unlock</a></p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> <span class="op">&lt;</span>fasta_file<span class="op">&gt;</span> -db <span class="op">&lt;</span>db<span class="op">&gt;</span> -save_pssm_after_last_round <span class="at">-out_pssm</span> <span class="op">&lt;</span>pssm_file<span class="op">&gt;</span> -out_ascii_pssm <span class="op">&lt;</span>pssm_file.txt<span class="op">&gt;</span> -inclusion_ethresh <span class="op">&lt;</span>evalue<span class="op">&gt;</span> -evalue <span class="op">&lt;</span>evalue<span class="op">&gt;</span> -taxids <span class="op">&lt;</span>taxid<span class="op">&gt;</span> <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor16" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor16" aria-expanded="false" aria-controls="collapseInstructor16">
  <h3 class="accordion-header" id="headingInstructor16">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor16" class="accordion-collapse collapse" aria-labelledby="headingInstructor16" data-bs-parent="#accordionInstructor16">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-query</span> ravC_LP.fasta <span class="at">-db</span> refseq_select_prot <span class="at">-save_pssm_after_last_round</span> <span class="at">-out_pssm</span> ravC_Leg.pssm <span class="at">-out_ascii_pssm</span> ravC_Leg.pssm.txt <span class="at">-inclusion_ethresh</span> 1e-10 <span class="at">-evalue</span> 1e-6 <span class="at">-taxids</span> 118969 <span class="op">&gt;</span> ravC_Leg.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2.2.5-running-the-command-and-examining-the-results" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.2.5-running-the-command-and-examining-the-results" class="callout-inner">
<h3 class="callout-title">Challenge 2.2.5: Running the command and examining the results</h3>
<div class="callout-content">
<p>Run now the full command as above. When it’s done, look at the
resulting files. There should be three of them. Take some time to
explore these three files, especially the alignments resulting from the
<code>psiblast</code>.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="task-2-3-align-the-pssm-to-the-whole-database">Task 2.3: Align the PSSM to the whole database<a class="anchor" aria-label="anchor" href="#task-2-3-align-the-pssm-to-the-whole-database"></a>
</h3>
<p>You will now take the resulting PSSM and use that as a query to
perform a <code>psiblast</code> against the whole
<code>refseq_select_prot</code> database (not only against
<em>Legionellales</em>). You want to perform max 10 iterations (the
search will stop if it converges before the tenth iteration), and
increase the max number of target sequences to gather to 1000 per
iteration. You also want to change the output to a tabular form with
comments and add more columns to get into more details.</p>
<div id="challenge-2.3.1-align-the-pssm-set-e-value-thresholds-iteration-and-max-sequences" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.3.1-align-the-pssm-set-e-value-thresholds-iteration-and-max-sequences" class="callout-inner">
<h3 class="callout-title">Challenge 2.3.1: Align the PSSM, set E-value thresholds, iteration and max sequences</h3>
<div class="callout-content">
<p>Build the command as above, but don’t set the <code>-query</code>
option, use the option that allows to input a PSSM instead. Set the
E-value thresholds as above, the number of iterations to 10 and the
maximum number of sequences to gather to 1000 (per round). Direct the
result to the file <code>ravC_all.psiblast</code>.</p>
<p>Don’t run it yet, more options to come.</p>
</div>
</div>
</div>
<div id="accordionHint16" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint16" aria-expanded="false" aria-controls="collapseHint16">
  <h4 class="accordion-header" id="headingHint16"> Give me a hint </h4>
</button>
<div id="collapseHint16" class="accordion-collapse collapse" aria-labelledby="headingHint16" data-bs-parent="#accordionHint16">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-help</span></span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-help</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="at">--color</span> <span class="at">-B</span> 5 <span class="at">-A</span> 2 pssm</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution17" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution17" aria-expanded="false" aria-controls="collapseSolution17">
  <h4 class="accordion-header" id="headingSolution17"> Show me the solution </h4>
</button>
<div id="collapseSolution17" class="accordion-collapse collapse" aria-labelledby="headingSolution17" data-bs-parent="#accordionSolution17">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="ex">psiblast</span> <span class="at">-in_pssm</span> <span class="op">&lt;</span>binary pssm file<span class="op">&gt;</span> -db <span class="op">&lt;</span>db<span class="op">&gt;</span> -inclusion_ethresh <span class="op">&lt;</span>evalue<span class="op">&gt;</span> -evalue <span class="op">&lt;</span>evalue<span class="op">&gt;</span> -max_target_seqs <span class="op">&lt;</span>number<span class="op">&gt;</span> -num_iterations 10 <span class="op">&gt;</span> ravC_all.psiblast</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor17" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor17" aria-expanded="false" aria-controls="collapseInstructor17">
  <h3 class="accordion-header" id="headingInstructor17">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor17" class="accordion-collapse collapse" aria-labelledby="headingInstructor17" data-bs-parent="#accordionInstructor17">
<div class="accordion-body">
<p>psiblast -in_pssm ravC_Leg.pssm -db refseq_select_prot
-inclusion_ethresh 1e-10 -evalue 1e-6 -max_target_seqs 1000
-num_iterations 10 &gt; ravC_all.psiblast</p>
</div>
</div>
</div>
</div>
<div id="challenge-2.3.2-set-the-output-format" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.3.2-set-the-output-format" class="callout-inner">
<h3 class="callout-title">Challenge 2.3.2: Set the output format</h3>
<div class="callout-content">
<p>You want to have a tabular result format with comments, to help you
understand the output. You also want to use the standard columns but add
the query coverage per subject, the scientific and common name of the
subject sequence, as well as which super-kingdom (or domain) it belongs
to.</p>
</div>
</div>
</div>
<div id="accordionHint17" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint17" aria-expanded="false" aria-controls="collapseHint17">
  <h4 class="accordion-header" id="headingHint17"> Give me a hint </h4>
</button>
<div id="collapseHint17" class="accordion-collapse collapse" aria-labelledby="headingHint17" data-bs-parent="#accordionHint17">
<div class="accordion-body">
<p>Inspect the “Formatting options” section of the psiblast help
page.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution18" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution18" aria-expanded="false" aria-controls="collapseSolution18">
  <h4 class="accordion-header" id="headingSolution18"> Show me the solution </h4>
</button>
<div id="collapseSolution18" class="accordion-collapse collapse" aria-labelledby="headingSolution18" data-bs-parent="#accordionSolution18">
<div class="accordion-body">
<p>The correct <code>-outfmt</code> is 7. The string with the correct
columns is composed with this number and then column names, separated by
spaces, the whole enclosed by double quotes. An example in the help file
of psiblast is <code>"10 delim=@ qacc sacc score"</code></p>
<p>psiblast -in_pssm <binary pssm file> -db <db> -inclusion_ethresh
<evalue> -evalue <evalue> -max_target_seqs <number> -num_iterations 10
-outfmt “<format_number><col1><col2> …” &gt; ravC_all.psiblast</col2></col1></format_number></number></evalue></evalue></db></binary></p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor18" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor18" aria-expanded="false" aria-controls="collapseInstructor18">
  <h3 class="accordion-header" id="headingInstructor18">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor18" class="accordion-collapse collapse" aria-labelledby="headingInstructor18" data-bs-parent="#accordionInstructor18">
<div class="accordion-body">
<p>psiblast -in_pssm ravC_Leg.pssm -db refseq_select_prot
-inclusion_ethresh 1e-10 -evalue 1e-6 -max_target_seqs 1000
-num_iterations 10 -outfmt “7 qaccver saccver pident length mismatch
gapopen qstart qend sstart send qcovs evalue bitscore ssciname scomname
sskingdom” &gt; ravC_all.psiblast</p>
</div>
</div>
</div>
</div>
<div id="challenge-2.3.3-run-the-command-and-examine-the-results" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2.3.3-run-the-command-and-examine-the-results" class="callout-inner">
<h3 class="callout-title">Challenge 2.3.3: Run the command and examine the results</h3>
<div class="callout-content">
<p>This will take a while to run, maybe 30 minutes. When done, open the
result file and examine it. Did you find hits in the eukaryotes?</p>
</div>
</div>
</div>
<!---

Sometimes in future we could develop a new exercise, to retrieve 16S sequences within a specific taxon. But so far this episode is long enough, I believe, so I let it be.


## Exercise 3: Finding full-length 16S sequences within a taxon

--->
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Using BLAST is a blast!</li>
<li>Choice of database is a crucial trade-off between efficiency and
sensitivity</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-msa"><p>Content from <a href="msa.html">Multiple sequence alignment</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/msa.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 120 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you align multiple sequences?</li>
<li>Why is it important to properly align sequences?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use <code>mafft</code> to align multiple sequences.</li>
<li>Test different algorithms.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>In this episode, we are exploring multiple sequence alignment (MSA).
In the first task, you are going to use <a href="https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html" class="external-link"><code>mafft</code></a>
to align homologs of <a href="https://en.wikipedia.org/wiki/RpoB" class="external-link">RpoB</a>, the β subunit of the
bacterial RNA polymerase. It is a long, multi-domain protein, suitable
for showing issues related to MSA.</p>
<p>In the second task, you will trim that alignment to remove poorly
aligned regions.</p>
<p>But first, go to your own folder and create a phylogenetics
subfolder. You will use the alignments for other tutorials as well.</p>
</section><section><h2 class="section-heading" id="task-1-use-different-flavors-of-mafft-and-compare-the-results">Task 1: Use different flavors of <code>MAFFT</code> and compare the
results<a class="anchor" aria-label="anchor" href="#task-1-use-different-flavors-of-mafft-and-compare-the-results"></a>
</h2>
<hr class="half-width">
<p>Start by finding the data, which is a <a href="../files/rpoB.fasta">fasta file</a> containing 19 sequences of RpoB,
gathered by aligning RpoB from <em>E. coli</em> to a very small database
present at NCBI, the <a href="https://blast.ncbi.nlm.nih.gov/smartblast/smartBlast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs" class="external-link">landmark</a>
database. These sequences are a subset of the sequences gathered in the
previous exercise on <code>BLAST</code>.</p>
<p>The file is available in
<code>/proj/g2020004/nobackup/3MK013/data/</code>.</p>
<div id="challenge-1.1-prepare-the-terrain" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.1-prepare-the-terrain" class="callout-inner">
<h3 class="callout-title">Challenge 1.1: prepare the terrain</h3>
<div class="callout-content">
<p>The course base folder is at
<code>/proj/g2020004/nobackup/3MK013</code>. Go to your own folder,
create a <code>phylogenetics</code> subfolder, and move into it. Also,
start the <code>interactive</code> session, for 4 hours. The session
name is <code>uppmax2025-3-4</code> and the cluster is
<code>snowy</code>.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>Remember those commands?</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">ssh</span> <span class="at">-Y</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="bu">cd</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">mkdir</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">ls</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">interactive</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">ssh</span> <span class="at">-Y</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se</span></code></pre>
</div>
<p>Remember to replace <code>&lt;username&gt;</code> by your name.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> <span class="op">&lt;</span>session<span class="op">&gt;</span> -M <span class="op">&lt;</span>cluster<span class="op">&gt;</span> -t <span class="op">&lt;</span>hh::mm::ss<span class="op">&gt;</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>basefolder<span class="op">&gt;</span>/<span class="op">&lt;</span>username<span class="op">&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">mkdir</span> <span class="op">&lt;</span>folder<span class="op">&gt;</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>folder<span class="op">&gt;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">ssh</span> <span class="at">-Y</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2025-3-4 <span class="at">-M</span> snowy <span class="at">-t</span> 4:00:00</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">mkdir</span> phylogenetics</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="bu">cd</span> phylogenetics</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-1.2-copy-the-file" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.2-copy-the-file" class="callout-inner">
<h3 class="callout-title">Challenge 1.2: copy the file</h3>
<div class="callout-content">
<p>Copy the file to your newly created <code>phylogenetics</code>
folder. Use a relative path.</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">cp</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="bu">pwd</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">ls</span> ../..</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Use the <code>../../</code> to see what’s two folders up, and then
<code>data/rpoB/rpoB.fasta</code></p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor2" aria-labelledby="headingInstructor2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">cp</span> ../../data/rpoB/rpoB.fasta .</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-1.1-prepare-the-terrain" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1.1-prepare-the-terrain" class="callout-inner">
<h3 class="callout-title">Challenge 1.1: prepare the terrain<em>
(continued)</em>
</h3>
<div class="callout-content">
<!---

```bash
cp ../../data/rpoB/rpoB.fasta .
```

--->
</div>
</div>
</div>
<div class="section level3">
<h3 id="renaming-sequences">Renaming sequences<a class="anchor" aria-label="anchor" href="#renaming-sequences"></a>
</h3>
<p>Look at the accession ids of the fasta sequences: they are not very
informative.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'&gt;'</span> rpoB.fasta <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;WP_000263098.1 MULTISPECIES: DNA-directed RNA polymerase subunit beta [Enterobacteriaceae]
&gt;WP_011070610.1 DNA-directed RNA polymerase subunit beta [Shewanella oneidensis]
&gt;WP_003114335.1 DNA-directed RNA polymerase subunit beta [Pseudomonas aeruginosa]
&gt;WP_002228870.1 DNA-directed RNA polymerase subunit beta [Neisseria meningitidis]
&gt;WP_003436174.1 MULTISPECIES: DNA-directed RNA polymerase subunit beta [Eubacteriales]</code></pre>
</div>
<p>Keep the taxonomic name following the accession id, separated by
<code>_</code>, using a bit of <code>sed</code> magic. Save the
resulting file for later use, and show the headers again.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">sed</span> <span class="st">"/^&gt;/ s/&gt;\([^ ]*\) \([^\[]*\)\[\(.*\)\]/&gt;</span><span class="dt">\\</span><span class="st">3_</span><span class="dt">\\</span><span class="st">1/"</span>  rpoB.fasta <span class="kw">|</span> <span class="fu">sed</span> <span class="st">"s/ /_/g"</span> <span class="op">&gt;</span> rpoB_renamed.fasta</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'&gt;'</span> rpoB_renamed.fasta <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code></pre>
</div>
<div id="nerd-alert-dissecting-the-sed-magic" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="nerd-alert-dissecting-the-sed-magic" class="callout-inner">
<h3 class="callout-title">Nerd alert: dissecting the <code>sed</code>
magic</h3>
<div class="callout-content">
<p>This is optional reading.</p>
<p>The <a href="https://www.gnu.org/software/sed/manual/sed.html" class="external-link"><code>sed</code></a>
command matches (and possibly substitutes) strings (chains of
characters). In that case, the goal is to simplify the header by putting
the taxonomic group first but keeping it informative enough by adding
the accession number. The strategy is to match what is between the
<code>&gt;</code> and the first space, then what is between square
parentheses, and put them back, separated by an underscore.</p>
<p>The <code>sed</code> command first matches only lines that start with
a <code>&gt;</code> (<code>/^&gt;/</code>). It then substitutes (general
pattern <code>s/&lt;something&gt;/&lt;something else&gt;/</code>) a text
with another one. The first part is to match the accession id, between
(escaped) square brackets, which comes after the <code>&gt;</code> at
the beginning of the line. This is expressed as
<code>&gt;\([^ ]*\)</code>: match any number of non-space characters
(<code>[^ ]*</code>) and put it in memory (what is between
<code>\(</code> and <code>\)</code>). Then, the description is matched
by <code>\([^\[]*\)</code>, any number of characters that are not an
opening bracket <code>[</code>, and put into memory. Finally, the
taxonomic description is matched: <code>\[\(.*\)\]</code>, that is, any
number of characters between square brackets is stored into memory. The
whole line is then replaced with a <code>&gt;</code>, the third match
into memory, followed by an <code>_</code> and the content of the first
match into memory <code>&gt;\\3_\\1</code>. Then, all the input is
passed through sed again, to replace any space with an underscore:
<code>s/ /_/g</code> and the output is stored in a different file.</p>
</div>
</div>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&gt;Enterobacteriaceae_WP_000263098.1
&gt;Shewanella_oneidensis_WP_011070610.1
&gt;Pseudomonas_aeruginosa_WP_003114335.1
&gt;Neisseria_meningitidis_WP_002228870.1
&gt;Eubacteriales_WP_003436174.1</code></pre>
</div>
<p>It looks better.</p>
</div>
<div class="section level3">
<h3 id="alignment-with-progressive-algorithm">Alignment with progressive algorithm<a class="anchor" aria-label="anchor" href="#alignment-with-progressive-algorithm"></a>
</h3>
<p>Use <code>mafft</code> with a progressive algorithm to align the
sequences.</p>
<div id="challenge-1.3-run-mafft-with-progressive-algorithm" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.3-run-mafft-with-progressive-algorithm" class="callout-inner">
<h3 class="callout-title">Challenge 1.3: Run <code>mafft</code> with
progressive algorithm</h3>
<div class="callout-content">
<p>Use the FFT-NS-2 algorithm from <code>mafft</code> to align the
renamed sequences. Also, record the time it takes for <code>mafft</code>
to complete the task.</p>
</div>
</div>
</div>
<div id="accordionHint3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint3" aria-expanded="false" aria-controls="collapseHint3">
  <h4 class="accordion-header" id="headingHint3"> Give me a hint </h4>
</button>
<div id="collapseHint3" class="accordion-collapse collapse" data-bs-parent="#accordionHint3" aria-labelledby="headingHint3">
<div class="accordion-body">
<p>Use the <code>module</code> command to load
<code>bioinfo-tools</code> and <code>MAFFT</code>. Use <code>time</code>
to record the time.</p>
<p>The help obtained through <code>mafft -h</code> is not very
informative about algorithms, so check the <a href="https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html" class="external-link">mafft
webpage</a>.</p>
<p><code>mafft</code> actually has one executable program for each
algorithm, all starting with <code>mafft-</code>. A way to display them
all is to type that and push the Tab key twice to see all
possibilities.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">module</span> load <span class="op">&lt;</span>general module<span class="op">&gt;</span> <span class="op">&lt;</span>mafft module<span class="op">&gt;</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="bu">time</span> mafft-<span class="op">&lt;</span>algorithm<span class="op">&gt;</span> <span class="op">&lt;</span>fasta_file<span class="op">&gt;</span> <span class="op">&gt;</span> rpoB.fftns.aln</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
[...]
Strategy:
 FFT-NS-2 (Fast but rough)
 Progressive method (guide trees were built 2 times.)

If unsure which option to use, try 'mafft --auto input &gt; output'.
For more information, see 'mafft --help', 'mafft --man' and the mafft page.

The default gap scoring scheme has been changed in version 7.110 (2013 Oct).
It tends to insert more gaps into gap-rich regions than previous versions.
To disable this change, add the --leavegappyregion option.


real	0m1.125s
user	0m0.818s
sys	0m0.181s</code></pre>
</div>
<p>The last line is the output of the <code>time</code> command. It took
1.125 seconds to complete this time.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor3" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor3" aria-expanded="false" aria-controls="collapseInstructor3">
  <h3 class="accordion-header" id="headingInstructor3">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor3" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor3" aria-labelledby="headingInstructor3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools MAFFT</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="bu">time</span> mafft-fftns rpoB_renamed.fasta <span class="op">&gt;</span> rpoB.fftns.aln</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionHint4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint4" aria-expanded="false" aria-controls="collapseHint4">
  <h4 class="accordion-header" id="headingHint4"> Give me a hint </h4>
</button>
<div id="collapseHint4" class="accordion-collapse collapse" data-bs-parent="#accordionHint4" aria-labelledby="headingHint4">
<div class="accordion-body">
<p>Type <code>mafft</code> and try tab-complete to see all versions of
<code>mafft</code>.</p>
<p>Try the command <code>time</code></p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="alignment-with-iterative-algorithm">Alignment with iterative algorithm<a class="anchor" aria-label="anchor" href="#alignment-with-iterative-algorithm"></a>
</h3>
<p>Now use one of the supposedly better iterative algorithm of
<code>mafft</code> to align the same sequences. Choose the <a href="https://mafft.cbrc.jp/alignment/software/algorithms/algorithms.html" class="external-link">E-INS-i
algorithm</a> which is suited for proteins that have highly conserved
motifs interspersed with less conserved ones.</p>
<p>Take a few minutes to read upon the different alignment strategies on
the page above.</p>
<div id="challenge-1.4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1.4" class="callout-inner">
<h3 class="callout-title">Challenge 1.4</h3>
<div class="callout-content">
<p>Use the superior E-INS-i algorithm from mafft to align the renamed
sequences. Also, record the time it takes for <code>mafft</code> to
complete the task.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">time</span> mafft-<span class="op">&lt;</span>better algo<span class="op">&gt;</span> <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> <span class="op">&gt;</span> rpoB.einsi.aln</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[...]
Strategy:
 E-INS-i (Suitable for sequences with long unalignable regions, very slow)
 Iterative refinement method (&lt;16) with LOCAL pairwise alignment with generalized affine gap costs (Altschul 1998)

If unsure which option to use, try 'mafft --auto input &gt; output'.
For more information, see 'mafft --help', 'mafft --man' and the mafft page.

The default gap scoring scheme has been changed in version 7.110 (2013 Oct).
It tends to insert more gaps into gap-rich regions than previous versions.
To disable this change, add the --leavegappyregion option.

Parameters for the E-INS-i option have been changed in version 7.243 (2015 Jun).
To switch to the old parameters, use --oldgenafpair, instead of --genafpair.


real	0m7.367s
user	0m7.022s
sys	0m0.244s</code></pre>
</div>
<p>It now took 7.36 seconds to complete this time, i.e. 6 times more
than with the progressive algorithm. It doesn’t make a big difference
now, but with hundreds of sequences it will make one.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor4" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor4" aria-expanded="false" aria-controls="collapseInstructor4">
  <h3 class="accordion-header" id="headingInstructor4">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor4" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor4" aria-labelledby="headingInstructor4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="bu">time</span> mafft-einsi rpoB_renamed.fasta <span class="op">&gt;</span> rpoB.einsi.aln</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="alignment-visualization">Alignment visualization<a class="anchor" aria-label="anchor" href="#alignment-visualization"></a>
</h3>
<p>You will now inspect the two resulting alignment methods. There are
no convenient way to do that from Uppmax, and the easiest solution is to
download the alignments <strong>on your computer</strong> and to use
either <a href="https://doua.prabi.fr/software/seaview" class="external-link"><code>seaview</code></a>
(you will need to install it on your computer) or an online alignment
viewer like <a href="https://alignmentviewer.org/" class="external-link">AlignmentViewer</a>.</p>
<p>Arrange the two windows on top of each other. Change the fontsize
(Props -&gt; Fontsize in Seaview) to 8 to see a larger portion of the
alignment.</p>
<div id="challenge5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Can you spot differences? Which alignment is longer?</p>
<p>Hint: try to scroll to position 800-900. What do you see there? How
are the blocks arranged?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>Use <code>scp</code> to copy files from Uppmax to your computer.
<code>scp</code> allows wildcards, but you probably need to escape the
<code>*</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:/<span class="op">&lt;</span>absolute path to phylogenetics folder<span class="op">&gt;</span>/rpoB.\*.aln <span class="op">&lt;</span>localfolder<span class="op">&gt;</span>/</span></code></pre>
</div>
<figure><img src="../fig/seaview.png" alt="Alignments shown in seaview" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="accordionInstructor5" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor5" aria-expanded="false" aria-controls="collapseInstructor5">
  <h3 class="accordion-header" id="headingInstructor5">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor5" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor5" aria-labelledby="headingInstructor5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">scp</span> lionel@rackham.uppmax.uu.se:/proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>user<span class="op">&gt;</span>/phylogenetics/rpoB.\*.aln .</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>If you have time over, spend it exploring the different options of
Seaview/AlignmentViewer.</p>
</div>
</section><section><h2 class="section-heading" id="task-2-trim-the-alignment">Task 2: Trim the alignment<a class="anchor" aria-label="anchor" href="#task-2-trim-the-alignment"></a>
</h2>
<hr class="half-width">
<p>Often, some regions of the alignment don’t align properly, either
because they contain low complexity segments (hinges in proteins) or
evolved through repeated insertions/deletions, which alignment program
cannot handle properly. It is thus good practice to remove (trim) these
regions, as they are likely to worsen the quality of the subsequent
phylogenetic trees. On the other hand, trimming too much of the
alignment removes also potentially valuable information. There is thus a
balance to be found.</p>
<p>In this part, you will use two different alignment trimmers, <a href="http://trimal.cgenomics.org/" class="external-link"><code>TrimAl</code></a> and <a href="https://jlsteenwyk.com/ClipKIT/" class="external-link"><code>ClipKIT</code></a>, on the
results of <code>mafft</code>’s E-INS-i algorithm.</p>
<div class="section level3">
<h3 id="trimming-with-trimal">Trimming with TrimAl<a class="anchor" aria-label="anchor" href="#trimming-with-trimal"></a>
</h3>
<p>First, load the module and look at the list of options available with
<code>trimAl</code>.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">module</span> load trimAl</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="ex">trimal</span> <span class="at">-h</span></span></code></pre>
</div>
<p><code>trimAl</code> offers a lot of different options. You are going
to explore two different: gap threshold (<code>-gt</code>) and the
heuristic method <code>-automated1</code>, which automatically decides
between three automated methods, <code>-gappyout</code>,
<code>-strict</code> and <code>-strictplus</code>, based on the
properties of the alignment. The gap threshold methods removes columns
that contain a fraction of sequences lower than the cut-off.</p>
<p>For comparison purposes, you will be adding an html output.</p>
<div id="challenge-2.1-trimal-with-gap-threshold" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.1-trimal-with-gap-threshold" class="callout-inner">
<h3 class="callout-title">Challenge 2.1: TrimAl with gap threshold</h3>
<div class="callout-content">
<p>Use <code>trimAl</code> to remove positions in the alignment that
have more than 40% gaps.</p>
</div>
</div>
</div>
<div id="accordionHint5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint5" aria-expanded="false" aria-controls="collapseHint5">
  <h4 class="accordion-header" id="headingHint5"> Give me a hint </h4>
</button>
<div id="collapseHint5" class="accordion-collapse collapse" data-bs-parent="#accordionHint5" aria-labelledby="headingHint5">
<div class="accordion-body">
<p>The “gap threshold” is actually expressed as fraction of non-gap
residues.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">trimal</span> <span class="at">-in</span> <span class="op">&lt;</span>input aln<span class="op">&gt;</span> -out rpoB.einsi.trimalgt.aln <span class="at">-gt</span> <span class="op">&lt;</span>cut-off<span class="op">&gt;</span> -htmlout <span class="op">&lt;</span>html_output<span class="op">&gt;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor6" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor6" aria-expanded="false" aria-controls="collapseInstructor6">
  <h3 class="accordion-header" id="headingInstructor6">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor6" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor6" aria-labelledby="headingInstructor6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">trimal</span> <span class="at">-in</span> rpoB.einsi.aln <span class="at">-out</span> rpoB.einsi.trimalgt.aln <span class="at">-gt</span> 0.6 <span class="at">-htmlout</span> rpoB.einsi.trimalgt.aln.html</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2.2-trimal-with-automated-trimming" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.2-trimal-with-automated-trimming" class="callout-inner">
<h3 class="callout-title">Challenge 2.2: TrimAl with automated trimming</h3>
<div class="callout-content">
<p>Use <code>trimAl</code> with the automated heuristic algorithm.</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<p>Add a <code>-automated1</code> option and rerun as above, choosing a
different output file.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor7" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor7" aria-expanded="false" aria-controls="collapseInstructor7">
  <h3 class="accordion-header" id="headingInstructor7">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor7" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor7" aria-labelledby="headingInstructor7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">trimal</span> <span class="at">-in</span> rpoB.einsi.aln <span class="at">-out</span> rpoB.einsi.trimalauto.aln <span class="at">-automated1</span> <span class="at">-htmlout</span> rpoB.einsi.trimalauto.aln.html</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2.3-compare-the-results" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.3-compare-the-results" class="callout-inner">
<h3 class="callout-title">Challenge 2.3: Compare the results</h3>
<div class="callout-content">
<p>Get the files (both alignments and html files) to your own laptop and
visualize the results, by opening the <code>html</code> files with your
browser and the alignment files with <code>seaview</code> or the viewer
you used above.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<p>On your own laptop, go inside the folder where you want to import the
files. Use <code>scp</code>. On some OS it is necessary to escape the
wildcard <code>*</code>. If the output says something about
<code>no matches found</code>, try that.</p>
<p>As before, if you do not have access to a terminal on your windows
laptop, use MobaXterm and Session &gt; SFTP to copy files to your
computer.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor8" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor8" aria-expanded="false" aria-controls="collapseInstructor8">
  <h3 class="accordion-header" id="headingInstructor8">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor8" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor8" aria-labelledby="headingInstructor8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:/proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/phylogenetics/rpoB.einsi.trimal\* .</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:/proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/phylogenetics/rpoB.einsi.trimal<span class="pp">*</span> .</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="trimming-with-clipkit">Trimming with ClipKIT<a class="anchor" aria-label="anchor" href="#trimming-with-clipkit"></a>
</h3>
<p><a href="https://jlsteenwyk.com/ClipKIT/" class="external-link"><code>ClipKIT</code></a> is
one of the more recent tools to trim multiple sequence alignments. In a
nutshell, it tries to preseve phylogenetically-informative sites, rather
than trimming gappy regions. Although it also has multiple options and
modes, you will only use the default mode, <code>smart-gap</code>.</p>
<div id="challenge-2.4-use-clipkit" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.4-use-clipkit" class="callout-inner">
<h3 class="callout-title">Challenge 2.4: Use ClipKIT</h3>
<div class="callout-content">
<p>To get an idea of the modes and options, load the
<code>ClipKIT</code> module and look at the help page:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">module</span> load ClipKIT</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="ex">clipkit</span> <span class="at">-h</span></span></code></pre>
</div>
<p>Then run ClipKIT, explicitly using the <code>smart-gap</code> mode.
Compare how much ClipKIT has trimmed the original alignment compared to
trimAl.</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" data-bs-parent="#accordionSolution9" aria-labelledby="headingSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">module</span> load ClipKIT</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="ex">clipkit</span> <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="ex">clipkit</span> <span class="op">&lt;</span>input aln<span class="op">&gt;</span> -m <span class="op">&lt;</span>mode<span class="op">&gt;</span> -l <span class="at">-o</span> <span class="op">&lt;</span>output file<span class="op">&gt;</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>---------------------
| Output Statistics |
---------------------
Original length: 2043
Number of sites kept: 1543
Number of sites trimmed: 500
Percentage of alignment trimmed: 24.474%

Execution time: 0.379s</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor9" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor9" aria-expanded="false" aria-controls="collapseInstructor9">
  <h3 class="accordion-header" id="headingInstructor9">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor9" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor9" aria-labelledby="headingInstructor9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="ex">module</span> load ClipKIT</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="ex">clipkit</span> <span class="at">-h</span></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="ex">clipkit</span> rpoB.einsi.aln <span class="at">-m</span> smart-gap <span class="at">-l</span> <span class="at">-o</span> rpoB.einsi.clipkit.aln</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="comparing-all-the-results">Comparing all the results<a class="anchor" aria-label="anchor" href="#comparing-all-the-results"></a>
</h3>
<p>Import the data and inspect the three alignments.</p>
<div id="challenge-2.5-import-data-and-compare-results" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2.5-import-data-and-compare-results" class="callout-inner">
<h3 class="callout-title">Challenge 2.5: Import data and compare results</h3>
<div class="callout-content">
<p>Copy the alignment file and visualize with <code>seaview</code> or
the web-based visualization tool.</p>
</div>
</div>
</div>
<div id="accordionHint6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint6" aria-expanded="false" aria-controls="collapseHint6">
  <h4 class="accordion-header" id="headingHint6"> Give me a hint </h4>
</button>
<div id="collapseHint6" class="accordion-collapse collapse" data-bs-parent="#accordionHint6" aria-labelledby="headingHint6">
<div class="accordion-body">
<p>Use <code>scp</code> as above.</p>
<p>Then use <code>seaview</code> or another viewer to visualize and
compare results.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" data-bs-parent="#accordionSolution10" aria-labelledby="headingSolution10">
<div class="accordion-body">
<p>The three alignments on top of each other look like this. Click <a href="../fig/seaview_trimal_vs_clipkit.png">on this link</a> to better see
the figure.</p>
<figure><img src="../fig/seaview_trimal_vs_clipkit.png" alt="Alignments shown in seaview" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="accordionInstructor10" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor10" aria-expanded="false" aria-controls="collapseInstructor10">
  <h3 class="accordion-header" id="headingInstructor10">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor10" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor10" aria-labelledby="headingInstructor10">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">scp</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@rackham.uppmax.uu.se:/proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/phylogenetics/rpoB.einsi.clipkit.aln .</span></code></pre>
</div>
<p>Then use <code>seaview</code> or another viewer to visualize and
compare results.</p>
</div>
</div>
</div>
</div>
<p>It is of course difficult to draw conclusions based on this figure,
but can you spot some trends? What alignment is more likely to generate
good results?</p>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-phylogenetics"><p>Content from <a href="phylogenetics.html">Phylogenetics</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/phylogenetics.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you build a simple, distance-based phylogenetic tree?</li>
<li>How do you build a phylogenetic tree with more advanced
methods?</li>
<li>How do you ascertain statistical support for phylogenetic
trees?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn the basic of phylogenetics tree building, taking the simplest
of the examples with a UPGMA tree.</li>
<li>Learn how to use phylogenetics algorithms, neighbor-joining and
maximum-likelihood.</li>
<li>Learn how to perform and show bootstraps.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="exercise-1-paper-and-pen-phylogenetic-tree">Exercise 1: Paper-and-pen phylogenetic tree<a class="anchor" aria-label="anchor" href="#exercise-1-paper-and-pen-phylogenetic-tree"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>The exercise is done for a large part with pen and paper, and then a
demonstration in R on your laptop, using RStudio. We’ll also use the R
package <a href="https://emmanuelparadis.github.io/" class="external-link"><code>ape</code></a>, which
you should install if it’s not present on your setup. Commands can be
typed or pasted in the “Console” part of RStudio.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">'ape'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following package(s) will be installed:
- ape [5.8-1]
These packages will be installed into "~/work/course-microbial-genomics/course-microbial-genomics/renv/profiles/lesson-requirements/renv/library/linux-ubuntu-jammy/R-4.4/x86_64-pc-linux-gnu".

# Installing packages --------------------------------------------------------
- Installing ape ...                            OK [linked from cache]
Successfully installed 1 package in 5.4 milliseconds.</code></pre>
</div>
<p>And to load it:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="upgma-based-tree">UPGMA-based tree<a class="anchor" aria-label="anchor" href="#upgma-based-tree"></a>
</h3>
<p>Load the tree in fasta format, reading from a <code>temp</code>
file</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FASTAfile</span> <span class="op">&lt;-</span> <span class="fu">tempfile</span><span class="op">(</span><span class="st">"aln"</span>, fileext <span class="op">=</span> <span class="st">".fasta"</span><span class="op">)</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"&gt;A"</span>, <span class="st">"----ATCCGCTGATCGGCTG----"</span>,</span>
<span>    <span class="st">"&gt;B"</span>, <span class="st">"GCTGATCCGTTGATCGG-------"</span>,</span>
<span>    <span class="st">"&gt;C"</span>, <span class="st">"----ATCTGCTCATCGGCT-----"</span>,</span>
<span>    <span class="st">"&gt;D"</span>, <span class="st">"----ATTCGCTGAACTGCTGGCTG"</span>,</span>
<span>    file <span class="op">=</span> <span class="va">FASTAfile</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu">read.dna</span><span class="op">(</span><span class="va">FASTAfile</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre>
</div>
<p>Now look at the alignment. Notice there are gaps, which we don’t want
in this example. We also remove the non-informative (identical)
columns.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">alview</span><span class="op">(</span><span class="va">aln</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  000000000111111111122222
  123456789012345678901234
A ----ATCCGCTGATCGGCTG----
B GCTG.....T.......---....
C .......T...C.......-....
D ......T......A.T....GCTG</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aln_filtered</span> <span class="op">&lt;-</span> <span class="va">aln</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">7</span>,<span class="fl">8</span>,<span class="fl">10</span>,<span class="fl">12</span>,<span class="fl">14</span>,<span class="fl">16</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">alview</span><span class="op">(</span><span class="va">aln_filtered</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  123456
A CCCGTG
B ..T...
C .T.C..
D T...AT</code></pre>
</div>
<p>Now we have a simple alignment as in the lecture. Dots
(<code>.</code>) mean that the sequence is identical to the top row,
which makes it easier to calculate distances.</p>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”</h3>
<div class="callout-content">
<p>Let’s use a matrix to calculate distances between sequences. We’ll
just count the number of differences between the sequences. We’ll then
group the two closest sequences. Which are they?</p>
<table class="table">
<caption>Distances between the sequences.</caption>
<thead><tr class="header">
<th></th>
<th align="right">A</th>
<th align="right">B</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>B</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>C</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Here is the solution:</p>
<table class="table">
<caption>Distances between the sequences.</caption>
<thead><tr class="header">
<th></th>
<th align="right">A</th>
<th align="right">B</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>A</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>B</td>
<td align="right">1</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>C</td>
<td align="right">2</td>
<td align="right">3</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right"></td>
</tr>
</tbody>
</table>
<p>The two closest sequences are A and B.</p>
</div>
</div>
</div>
</div>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>Let’s now cluster together A and B, and calculate the average
distance from AB to the other sequences, weighted by the size of the
clusters.</p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">AB</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AB</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>C</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>D</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The average distance for AB to C is calculated as follow:</p>
<p><span class="math inline">\(d(AB,C) = \dfrac{d(A,C) + d(B,C)}{2} =
\dfrac{2 + 3}{2} = 2.5\)</span></p>
<p>And so on for the other distances:</p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">AB</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AB</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>C</td>
<td align="right">2.5</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="odd">
<td>D</td>
<td align="right">3.5</td>
<td align="right">5</td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>Now the shortest distance is AB to C. Let’s recalculate the distance
to D again.</p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">ABC</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ABC</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right"></td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><span class="math inline">\(d(ABC,D) = \dfrac{d(AB,D) * 2 + d(C,D) *
1}{2 + 1} = \dfrac{3.5 * 2 + 5 * 1}{3} = 4\)</span></p>
<table class="table">
<caption>Recalculated distances.</caption>
<thead><tr class="header">
<th></th>
<th align="right">ABC</th>
<th align="right">D</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ABC</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td>D</td>
<td align="right">4</td>
<td align="right"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<p>The tree is reconstructed by dividing the distances equally between
the two leaves. - A-B: each 0.5. - AB-C: each side gets 2.5/2 = 1.25.
The branch to AB is 1.25 - 0.75 = 0.75 - ABC-D: each side gets 4/2 = 2.
The branch to ABC is 2 - 0.75 - 0.5 = 0.75</p>
<figure><img src="../fig/upgma_manual.png" alt="Manually built UPGMA tree" class="figure mx-auto d-block"><div class="figcaption">UPGMA tree</div>
</figure><p>Let’s know do the same using bioinformatics tools.</p>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>We’ll use <code>dist.dna</code> to calculate the distances. We’ll use
a “N” model, that just counts the differences and doesn’t correct or
normalizes. We’ll use the function <code>hclust</code> to perform the
UPGMA method calculation. The tree is then plotted, and the branch
lengths plotted with <code>edgelabels</code>:</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist_matrix</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">aln_filtered</span>, model<span class="op">=</span><span class="st">"N"</span><span class="op">)</span></span>
<span><span class="va">dist_matrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  A B C
B 1
C 2 3
D 3 4 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">as.phylo</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="va">dist_matrix</span>, <span class="st">"average"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="fu">edgelabels</span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">edge.length</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/phylogenetics-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-2-neighbor-joining-and-maximum-likelihood-tree">Exercise 2: Neighbor-joining and Maximum-likelihood tree<a class="anchor" aria-label="anchor" href="#exercise-2-neighbor-joining-and-maximum-likelihood-tree"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>In the previous episode, we inferred a simple phylogenetic tree using
UPGMA, without correcting the distance matrix for multiple
substitutions. UPGMA has many shortcomings, and gets worse as distance
increases. Here we’ll test Neighbor-Joining, which is also a
distance-based, and a maximum-likelihood based approach with <a href="http://www.iqtree.org/" class="external-link">IQ-TREE</a>.</p>
</div>
<div class="section level3">
<h3 id="neighbor-joining">Neighbor-joining<a class="anchor" aria-label="anchor" href="#neighbor-joining"></a>
</h3>
<p>The principle of <a href="https://en.wikipedia.org/wiki/Neighbor_joining" class="external-link">Neighbor-joining
method</a> (NJ) is to start from a star-like tree, find the two branches
that, if joined, minimize the total tree length, join then, and repeat
with the joined branches and the rest of the branches.</p>
<div id="challenge5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>To perform NJ on our sequences, we’ll use the function inbuilt in
Seaview. First (re-)open the alignment we obtained from
<code>mafft</code> with the E-INS-i method, trimmed with ClipKIT. If it
is not on your computer anymore, transfer it again from Uppmax using
<code>scp</code> or <code>SFTP</code>.</p>
<p>If <a href="https://doua.prabi.fr/software/seaview" class="external-link">Seaview</a> is
not installed on your computer, download it and install it on your
computer.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>On a Linux computer, you can run it on the command line:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">seaview</span> rpoB.einsi.clipkit.aln <span class="kw">&amp;</span></span></code></pre>
</div>
<p>On other computers, just open the file with the regular File &gt;
Open menu.</p>
</div>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<em> (continued)</em>
</h3>
<div class="callout-content">
<p>In the Seaview window, select Trees -&gt; Distance methods. Keep the
BioNJ option ticked. BioNJ is an updated version of NJ, more accurate
for longer distances. Tick the “Bootstrap” option and leave it at 100.
We’ll discuss these later. Then click on “Go”.</p>
</div>
</div>
</div>
<p>What do you see? Is the tree rooted? Is the pattern species coherent
with what you know of the tree of life? Any weird results?</p>
<div id="challenge6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Redo the BioNJ tree for the other alignment we inferred.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">seaview</span> rpoB.fftns.aln <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Do you see any differences between the two trees? What can you make
out of it?</p>
</div>
<div class="section level3">
<h3 id="maximum-likelihood">Maximum likelihood<a class="anchor" aria-label="anchor" href="#maximum-likelihood"></a>
</h3>
<p>We will now use IQ-TREE to infer a maximum-likelihood (ML) tree of
the RpoB dataset we aligned with <code>mafft</code> previously. While
you performed the first trees on your laptop, you will infer the ML tree
on Uppmax. As usual, use the <code>interactive</code> command to gain
access to compute time. Go to the <code>phylogenetics</code> folder
created in the MSA episode, and load the right modules.</p>
<div id="challenge7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Get to the right folder, require compute time and load the right
modules. The project is <code>uppmax2025-3-4</code>. The module
containing IQ tree is called <code>iqtree</code>.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" aria-labelledby="headingHint1" data-bs-parent="#accordionHint1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">interactive</span> ...</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">module</span> load ...</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> <span class="op">&lt;</span>project<span class="op">&gt;</span> -M <span class="op">&lt;</span>cluster<span class="op">&gt;</span> -t <span class="op">&lt;</span>time<span class="op">&gt;</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="ex">module</span> load <span class="op">&lt;</span>general bioinfo module<span class="op">&gt;</span> iqtree</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2025-3-4 <span class="at">-M</span> snowy <span class="at">-t</span> 4:00:00</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools iqtree</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/phylogenetics</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Loading the <code>iqtree</code> module resulted in:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>iqtree/2.2.2.6-omp-mpi: executables are prefixed with 'iqtree2'. Additional modules are required for the multi-node MPI version. See 'module help iqtree/2.2.2.6-omp-mpi'.</code></pre>
</div>
<p>You won’t be using the multi-node MPI version, so you don’t need to
load additional modules. But you will need to use <code>iqtree2</code>
instead of the “normal” <code>iqtree</code>.</p>
<p>Then run your first tree, on the <code>ClipKIT</code>-trimmed
alignment.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> rpoB.einsi.clipkit.aln <span class="at">-m</span> MFP <span class="at">-B</span> 1000</span></code></pre>
</div>
<p>Here, we tell IQ-TREE to use the alignment
<code>-s rpoB.einsi.clipkit.aln</code>, and to test among the standard
substitution models which one fits best (<code>-m MFP</code>). We also
tell IQ-TREE to perform 1000 ultra-fast bootstraps
(<code>-B 1000</code>). We’ll discuss these later.</p>
<p>IQ-TREE is a very complete program that can do a large variety of
phylogenetic analysis. To get a flavor of what it’s capable of, look at
its <a href="http://www.iqtree.org/doc/" class="external-link">extensive
documentation</a>.</p>
<p>Have a look at the output files:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">ls</span> rpoB.einsi.clipkit.aln.<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>rpoB.einsi.clipkit.aln.bionj    rpoB.einsi.clipkit.aln.iqtree  rpoB.einsi.clipkit.aln.model.gz
rpoB.einsi.clipkit.aln.ckp.gz   rpoB.einsi.clipkit.aln.log     rpoB.einsi.clipkit.aln.splits.nex
rpoB.einsi.clipkit.aln.contree	rpoB.einsi.clipkit.aln.mldist  rpoB.einsi.clipkit.aln.treefile</code></pre>
</div>
<p>There are two important files:<br>
* <code>*.iqtree</code> file provides a text summary of the analysis. *
<code>*.treefile</code> is the resulting tree in <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick
format</a>.</p>
<p>To visualize the tree, open the <a href="https://beta.phylo.io/viewer/#" class="external-link">Beta version of phylo.io</a>.
Click on “add a tree” and copy/paste the content of the
<code>treefile</code> (use e.g. <code>less</code> or <code>cat</code> to
display it) in the box. Make sure the “Newick format” is selected and
click on “Done”.</p>
<p>The tree appears as unrooted. It is good practice to start by
ordering the nodes and root it. There is no good way to automatically
order the branches in phylo.io as of yet, but rerooting can be done by
clicking on a branch and selecting “Reroot”. Reroot the tree between
archaea and bacteria. Now the tree makes a bit more sense.</p>
<p>Scrutinize the tree. Is it different from the BioNJ tree generated in
Seaview? How?</p>
<div id="challenge8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Redo a ML tree from the other alignment (FFT-NS) we inferred with
<code>mafft</code> and display the resulting tree in FigTree.</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" aria-labelledby="headingHint2" data-bs-parent="#accordionHint2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-h</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> <span class="op">&lt;</span>alignment file<span class="op">&gt;</span> <span class="op">&lt;</span>option and text for testing model<span class="op">&gt;</span> <span class="op">&lt;</span>option and test for 1000 fast bootstraps<span class="op">&gt;</span> </span></code></pre>
</div>
<p>Import the tree <strong>on your computer</strong> and load it with
FigTree or with <a href="https://beta.phylo.io/viewer/#" class="external-link">phylo.io</a></p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">figtree</span> rpoB.fftns.aln.treefile <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" aria-labelledby="headingInstructor2" data-bs-parent="#accordionInstructor2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> rpoB.fftns.aln <span class="at">-m</span> MFP <span class="at">-B</span> 1000</span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">figtree</span> rpoB.fftns.aln.treefile <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="bootstraps">Bootstraps<a class="anchor" aria-label="anchor" href="#bootstraps"></a>
</h3>
<p>We have inferred four trees: * Two based on the alignment generated
from the E-INS-i algorithm (trimmed by ClipKIT), two from the FFT-NS. *
Two inferred with the BioNJ algorithm and two with the ML algorithm
(IQ-Tree)</p>
<p>Along the way, we’ve generated bootstraps for all our trees. Now show
them on all four trees. * For the Seaview trees, tick the ‘Br support’
box * For the trees shown in phylo.io, click on “Settings” &gt; “Branch
&amp; Labels” and above the branch, click the drop-down menu and select
“Data”.</p>
<div id="challenge9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Compare all four trees. Do you find any significant differences?</p>
</div>
</div>
</div>
<div id="accordionHint3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint3" aria-expanded="false" aria-controls="collapseHint3">
  <h4 class="accordion-header" id="headingHint3"> Give me a hint </h4>
</button>
<div id="collapseHint3" class="accordion-collapse collapse" aria-labelledby="headingHint3" data-bs-parent="#accordionHint3">
<div class="accordion-body">
<p>What are <em>Glycine</em> and <em>Arabidopsis</em>? What about
<em>Synechocystis</em> and <em>Microcystis</em>?</p>
</div>
</div>
</div>
</div>
<div id="accordionHint4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint4" aria-expanded="false" aria-controls="collapseHint4">
  <h4 class="accordion-header" id="headingHint4"> Give me a hint </h4>
</button>
<div id="collapseHint4" class="accordion-collapse collapse" aria-labelledby="headingHint4" data-bs-parent="#accordionHint4">
<div class="accordion-body">
<p>Search for the accession number of the <em>Glycine</em> RpoB sequence
on <a href="https://www.ncbi.nlm.nih.gov/search/" class="external-link">NCBI</a>. Any hint
there?</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<p>The two plant RpoB sequences are from chloroplastic genomes, and thus
branch together with the cyanobacterial sequences, in some of the
phylogenies at least.</p>
</div>
</div>
</div>
</div>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<em> (continued)</em>
</h3>
<div class="callout-content">
<p>What about the support values for grouping these two groups? How high
are they?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The simplest of the trees are distance-based.</li>
<li>UPGMA works by clustering the two closest leaves and recalculating
the distance matrix.</li>
<li>Neighbor-joining is distance-based and fast, but not necessarily
very accurate</li>
<li>Maximum-likelihood is slower, but more accurate</li>
</ul>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-3-genetic-drift">Exercise 3: Genetic drift<a class="anchor" aria-label="anchor" href="#exercise-3-genetic-drift"></a>
</h2>
<hr class="half-width">
<p>As a practical way to understand genetic drift, let’s play with
population size, selection coefficients, number of generations, etc.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How does random genetic drift influence fixation of alleles?</li>
<li>What is the influence of population size?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn the basic of phylogenetics tree building, taking the simplest
of the examples with a UPGMA tree.</li>
<li>Learn how to use phylogenetics algorithms, neighbor-joining and
maximum-likelihood.</li>
<li>Learn how to perform and show bootstraps.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How does random genetic drift influence fixation of alleles?</li>
<li>What is the influence of population size?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how population size influences the probability of
fixation of an allele.</li>
<li>Understand how slightly deleterious mutations may get fixed in small
populations.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a>
</h3>
<p>This exercise is to illustrate the concepts of selection, population
size and genetic drift, using simulations. We will use mostly <a href="https://cartwrig.ht/apps/redlynx/" class="external-link">Red Lynx</a> by <a href="https://github.com/reedacartwright" class="external-link">Reed A. Cartwright</a>.</p>
<p>Another option is to use a <a href="https://phytools.shinyapps.io/drift-selection/" class="external-link">web interface</a>
to the <a href="https://www.r-project.org/" class="external-link">R</a> <a href="https://github.com/liamrevell/learnPopGen" class="external-link">learnPopGen</a>
package, but the last one is mostly for biallelic genes (and thus not
that relevant for bacteria).</p>
<p>Open now the <a href="https://cartwrig.ht/apps/redlynx/" class="external-link">Red Lynx</a>
website and get familiar with the different options.</p>
<p>You won’t need the right panel (but feel free to explore). The
dominance option in the left panel won’t be used either.</p>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection</h3>
<div class="callout-content">
<p>In the first part, you will only play with the number of generations,
the initial frequency and the population size.</p>
<ul>
<li>Lower the number of generations to 200.</li>
<li>Adjust the population size to 1000.</li>
<li>Make sure the intial frequency is set to 50%.</li>
<li>Run a few simulations (ca 20).</li>
</ul>
<p>Did any allele got fixed? What is the range of frequencies after 200
generations?</p>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" aria-labelledby="headingSolution10" data-bs-parent="#accordionSolution10">
<div class="accordion-body">
<p>In my simulations, no allele got fixed, the final allele frequencies
range20-80%</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>Now increase the population to 100’000, clear the graph, and repeat
the simulations. What’s the range of final frequencies now?</p>
</div>
</div>
</div>
<div id="accordionSolution11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution11" aria-expanded="false" aria-controls="collapseSolution11">
  <h4 class="accordion-header" id="headingSolution11"> Show me the solution </h4>
</button>
<div id="collapseSolution11" class="accordion-collapse collapse" aria-labelledby="headingSolution11" data-bs-parent="#accordionSolution11">
<div class="accordion-body">
<p>In my simulations, no allele got fixed, and the final allele
frequencies range 45-55%</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>Now decrease the population to 10 individuals, clear the graph and
repeat these simulations. What’s the range of final frequencies now?</p>
</div>
</div>
</div>
<div id="accordionSolution12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution12" aria-expanded="false" aria-controls="collapseSolution12">
  <h4 class="accordion-header" id="headingSolution12"> Show me the solution </h4>
</button>
<div id="collapseSolution12" class="accordion-collapse collapse" aria-labelledby="headingSolution12" data-bs-parent="#accordionSolution12">
<div class="accordion-body">
<p>In my simulations, one allele got fixed quickly, the latest one was
at generation 100.</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>What do you conclude here?</p>
</div>
</div>
</div>
<div id="accordionSolution13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution13" aria-expanded="false" aria-controls="collapseSolution13">
  <h4 class="accordion-header" id="headingSolution13"> Show me the solution </h4>
</button>
<div id="collapseSolution13" class="accordion-collapse collapse" aria-labelledby="headingSolution13" data-bs-parent="#accordionSolution13">
<div class="accordion-body">
<p>It is clear that stochastic (random) variation in allele frequencies
strongly affects the probability of fixation of alleles in small
populations, not so much in large ones.</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-with-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-with-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift with selection</h3>
<div class="callout-content">
<p>So far we’ve only looked at how allele frequencies vary in the
absence of selection, that is when the two alleles provide an equal
fitness. What’s the influence of random genetic drift when alleles are
not neutral?</p>
<p>The selection strength is equivalent to the selection coefficient,
i.e. how much higher relative fitness the new allele provides. A
selection coefficient of 0.01 means that the organism harboring the new
allele has a 1% increased fitness.</p>
<ul>
<li>Increase the number of generations to 1000.</li>
<li>Set the selection strength to 0.01 (1%).</li>
<li>Set the population size first to 100’000, then to 1000, then to
100.</li>
</ul>
<p>How long does it take for the allele to get fixed - in average - with
the three population sizes?</p>
</div>
</div>
</div>
<div id="accordionSolution14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution14" aria-expanded="false" aria-controls="collapseSolution14">
  <h4 class="accordion-header" id="headingSolution14"> Show me the solution </h4>
</button>
<div id="collapseSolution14" class="accordion-collapse collapse" aria-labelledby="headingSolution14" data-bs-parent="#accordionSolution14">
<div class="accordion-body">
<p>About the same time, but the trajectories are much smoother with
larger populations. In the small population, it happens that the
beneficial allele disappears from the population (although not
often).</p>
</div>
</div>
</div>
</div>
<div id="fixation-of-slightly-deleterious-alleles-in-very-small-populations" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="fixation-of-slightly-deleterious-alleles-in-very-small-populations" class="callout-inner">
<h3 class="callout-title">Fixation of slightly deleterious alleles in very small populations</h3>
<div class="callout-content">
<p>We’re now simulating what would happen in a very small population (or
a population that undergoes very narrow bottlenecks), when a gene
mutates. We’ll have a very small population (10 individuals), a
selection value of -0.01 (the mutated allele provides a 1% lower
fitness), and a 10% initial frequency, which corresponds to one
individual getting a mutation:</p>
<ul>
<li>Set the population to 10 individuals.</li>
<li>Set generations to 200.</li>
<li>Set initial frequency to 10%.</li>
<li>Set the selection strength to -0.01.</li>
</ul>
<p>Run many simulations. What happens?</p>
</div>
</div>
</div>
<div id="accordionSolution15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution15" aria-expanded="false" aria-controls="collapseSolution15">
  <h4 class="accordion-header" id="headingSolution15"> Show me the solution </h4>
</button>
<div id="collapseSolution15" class="accordion-collapse collapse" aria-labelledby="headingSolution15" data-bs-parent="#accordionSolution15">
<div class="accordion-body">
<p>Most new alleles go extinct quickly. In my simulations, I usually get
one of the slightly deleterious mutations fixed after 20
simulations.</p>
</div>
</div>
</div>
</div>
<p>That’s it for that part, but feel free to continue playing with the
different settings later.</p>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Random genetic drift has a large influence on the probability of
fixation of alleles in small populations, even for non-neutral
alleles.</li>
<li>Random genetic drift has very little influence of the probability of
fixation of alleles in large populations.</li>
<li>Slightly deleterious mutations can get fixed into the population
through random genetic drift, if the population is small enough and the
selective value is not too large.</li>
</ul>
</div>
</div>
</div>
<!-- References -->
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-quality-control"><p>Content from <a href="quality-control.html">Reads, QC and trimming</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/quality-control.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>This episode starts a series of labs where you will trace an outbreak
of <em>Mycobacterium tuberculosis</em>, from downloading the reads to
making assumptions about the transmissions between patients.</p>
<p>The first part, here, describes how to retrieve all necessary data
from public repositories - the raw sequenced data of our isolates and a
reference genome. It also introduces for loops which we will use for the
rest of our analysis.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Trace an outbreak of <em>Mycobacterium</em>, from reads to
phylogenetic tree.</li>
<li>How can I organize my file system for a new bioinformatics
project?</li>
<li>How and where can data be downloaded?</li>
<li>How can I describe the quality of my data?</li>
<li>How can I get rid of sequence data that doesn’t meet my quality
standards?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>How do you use genomics and phylogenetics to trace a microbial
outbreak?</li>
<li>Create a file system for a bioinformatics project.</li>
<li>Download files necessary for further analysis.</li>
<li>Use ‘for’ loops to automate operations on multiple files</li>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Clean FastQC reads for further analysis.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>This part of the course is largely inspired from <a href="https://github.com/aschuerch/MolecularEpidemiology_AnalysisWGS" class="external-link">another
tutorial</a> developed by <a href="https://aschuerch.github.io/" class="external-link">Anita
Schürch</a>. We will use (part of) the data described in <a href="https://bmcinfectdis.biomedcentral.com/articles/10.1186/1471-2334-13-110" class="external-link">Bryant
et al., 2013</a>.</p>
</section><section><h2 class="section-heading" id="background-of-the-data">Background of the data<a class="anchor" aria-label="anchor" href="#background-of-the-data"></a>
</h2>
<hr class="half-width">
<p>Epidemiological contact tracing through interviewing of patients can
identify potential chains of patients that transmitted an infectious
disease to each other. Contact tracing was performed following the <a href="https://www.ncbi.nlm.nih.gov/pubmed/1643300" class="external-link">stone-in-the-pond
principle</a>, which interviews and tests potential contacts in
concentric circles around a potential bash case.</p>
<p>Tuberculosis (TB) is an infectious disease caused by
<em>Mycobacterium tuberculosis</em>. It mostly affects the lungs. An
infection with <em>M. tuberculosis</em> is often asymptomatic (latent
infection). Only in about 10% of the cases the latent infection
progresses to an active infection during a patients lifetime, which, if
untreated, leads to death in about half of the cases. The symptoms of an
active TB infection include cough, fever, night sweats, weight loss etc.
An active TB infection can spread. Once exposed, people often have
latent TB. To identify people with latent TB, a <a href="https://www.cdc.gov/tb/hcp/testing-diagnosis/tuberculin-skin-test.html" class="external-link">skin
test</a> can be applied.</p>
<p>Here we have 7 tuberculosis patients with active TB, that form three
separate clusters of potential transmission as determined by
epidemiological interviews. Patients were asked if they have been in
direct contact with each other, or if they visited the same localities.
From all patients, a bacterial isolate was grown, DNA isolated, and
whole-genome sequenced on an Illumina sequencer.</p>
<p>The three clusters consist of:</p>
<ul>
<li>Cluster 1
<ul>
<li>Patient A1 (2011) - sample ERR029207</li>
<li>Patient A2 (2011) - sample ERR029206</li>
<li>Patient A3 (2008) - sample ERR026478</li>
</ul>
</li>
<li>Cluster 2
<ul>
<li>Patient B1 (2001) - sample ERR026474</li>
<li>Patient B2 (2012) - sample ERR026473</li>
</ul>
</li>
<li>Cluster 3
<ul>
<li>Patient C1 (2011) - sample ERR026481</li>
<li>Patient C2 (2016) - sample ERR026482</li>
</ul>
</li>
</ul>
<p>The second goal of this lab is to affirm or dispute the inferred
transmissions by comparing the bacterial genomes with each other. We
will do that by identifying SNPs between the genomes.</p>
</section><section><h2 class="section-heading" id="getting-your-project-started">Getting your project started<a class="anchor" aria-label="anchor" href="#getting-your-project-started"></a>
</h2>
<hr class="half-width">
<p>Project organization is one of the most important parts of a
sequencing project, and yet is often overlooked amidst the excitement of
getting a first look at new data. Of course, while it’s best to get
yourself organized before you even begin your analyses,it’s never too
late to start, either.</p>
<p>Genomics projects can quickly accumulate hundreds of files across
tens of folders. Every computational analysis you perform over the
course of your project is going to create many files, which can
especially become a problem when you’ll inevitably want to run some of
those analyses again. For instance, you might have made significant
headway into your project, but then have to remember the PCR conditions
you used to create your sequencing library months prior.</p>
<p>Other questions might arise along the way:</p>
<ul>
<li>What were your best alignment results?</li>
<li>Which folder were they in: Analysis1, AnalysisRedone, or
AnalysisRedone2?</li>
<li>Which quality cutoff did you use?</li>
<li>What version of a given program did you implement your analysis
in?</li>
</ul>
<p>In this exercise we will setup a file system for the project we will
be working on during this workshop.</p>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>We will start by creating a sub directory that we can use for the
rest of the labs. First navigate to your own directory on Uppmax and
start an <code>interactive</code> session. Then confirm that you are in
the correct directory using the <code>pwd</code> command.</p>
<p>You should see the output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>/proj/g2020004/nobackup/3MK013/&lt;username&gt;</code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">interactive</span> ...</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2025-3-4 <span class="at">-M</span> snowy <span class="at">-t</span> 04:00:00</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Use the <code>mkdir</code> command to make the following
directories:</p>
<pre><code><span><span class="va">molepi</span></span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">docs</span></span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">data</span></span>
<span><span class="va">molepi</span><span class="op">/</span><span class="va">results</span></span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">mkdir</span> molepi molepi/docs molepi/data molepi/results</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Use <code>ls -R</code> to verify that you have created these
directories. The <code>-R</code> option for <code>ls</code> stands for
recursive. This option causes <code>ls</code> to return the contents of
each subdirectory within the directory iteratively.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-R</span> molepi</span></code></pre>
</div>
<p>You should see the following output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>molepi/:
data  docs  results

molepi/data:

molepi/docs:

molepi/results: </code></pre>
</div>
</section><section><h2 class="section-heading" id="selection-of-a-reference-genome">Selection of a reference genome<a class="anchor" aria-label="anchor" href="#selection-of-a-reference-genome"></a>
</h2>
<hr class="half-width">
<p>Reference sequences (including many pathogen genomes) are available
at <a href="https://www.ncbi.nlm.nih.gov/refseq/" class="external-link">NCBI’s refseq
database</a></p>
<p>A reference genome is a genome that was previously sequenced and is
closely related to the isolates we would like to analyse. The selection
of a closely related reference genome is not trivial and will warrant an
analysis in itself. However, for simplicity, here we will work with the
<em>M. tuberculosis</em> reference genome H37Rv.</p>
<div id="download-reference-genomes-from-ncbi" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="download-reference-genomes-from-ncbi" class="callout-inner">
<h3 class="callout-title">Download reference genomes from NCBI</h3>
<div class="callout-content">
<p>Download the <em>M.tuberculosis</em> reference genome from the NCBI
ftp site.</p>
<p>First, we switch to the <code>data</code> subfolder to store all our
data</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="bu">cd</span> molepi/data </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="download-reference-genomes-from-ncbi" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="download-reference-genomes-from-ncbi" class="callout-inner">
<h3 class="callout-title">Download reference genomes from NCBI<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>The reference genome will be downloaded programmatically from NCBI
with <code>wget</code>. <code>wget</code> is a computer program that
retrieves content from web servers. Its name derives from World Wide Web
and get.</p>
<p>Find the location of the gzipped <code>fna</code> file for that
reference genome via NCBI’s genome, by finding the FTP site and copying
the link to the <code>*_genomic.fna.gz</code> file and using that as an
argument to <code>wget</code>.</p>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-h</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>Search for <em>Mycobacterium tuberculosis</em> on the <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/" class="external-link">Datasets/Genome</a>
website. Identify the “reference genome” (at the top), click on the
accession number, and find the <code>FTP</code> tab. There is many
files, among which
<code>GCF_000195955.2_ASM19595v2_genomic.fna.gz</code>. Right-click and
copy the link.</p>
</div>
</div>
</div>
</div>
<div id="accordionInstructor2" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor2" aria-expanded="false" aria-controls="collapseInstructor2">
  <h3 class="accordion-header" id="headingInstructor2">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor2" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor2" aria-labelledby="headingInstructor2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">wget</span> ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/195/955/GCF_000195955.2_ASM19595v2/GCF_000195955.2_ASM19595v2_genomic.fna.gz</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="download-reference-genomes-from-ncbi" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="download-reference-genomes-from-ncbi" class="callout-inner">
<h3 class="callout-title">Download reference genomes from NCBI<em>
(continued)</em>
</h3>
<div class="callout-content">
<p>This file is compressed as indicated by the extension of
<code>.gz</code>. It means that this file has been compressed using the
<code>gzip</code> command.</p>
<p>Extract the file by typing</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">gunzip</span> GCF_000195955.2_ASM19595v2_genomic.fna.gz</span></code></pre>
</div>
<p>Make sure that is was extracted</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">GCF_000195955.2_ASM19595v2_genomic.fna</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="challenge-what-is-the-size-of-the-genome">Challenge: What is the size of the genome?<a class="anchor" aria-label="anchor" href="#challenge-what-is-the-size-of-the-genome"></a>
</h2>
<p>Find out how many nucleotides the genome has. There is a software
called SeqKit (and a corresponding module at Uppmax) that can help.</p>
<div class="hint:">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools SeqKit</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">seqkit</span> <span class="at">-h</span></span></code></pre>
</div>
</div>
<p>Get assembly statistics from FASTA and FASTQ files.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>The genome has 4’411’532 bp. The <code>stat</code> command of
<code>seqkit</code> is helpful.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>file                                    format  type  num_seqs    sum_len    min_len    avg_len    max_len
GCF_000195955.2_ASM19595v2_genomic.fna  FASTA   DNA          1  4,411,532  4,411,532  4,411,532  4,411,532
</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor3" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor3" aria-expanded="false" aria-controls="collapseInstructor3">
  <h3 class="accordion-header" id="headingInstructor3">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor3" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor3" aria-labelledby="headingInstructor3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">seqkit</span> stats GCF_000195955.2_ASM19595v2_genomic.fna</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="loops">Loops<a class="anchor" aria-label="anchor" href="#loops"></a>
</h2>
<hr class="half-width">
<p><em>Loops</em> are key to productivity improvements through
automation as they allow us to execute commands repeatedly. Similar to
wildcards and tab completion, using loops also reduces the amount of
typing (and typing mistakes). Our next task is to download our data (see
the introduction to this episode) from the short read archive <a href="https://www.ebi.ac.uk/ena" class="external-link">(SRA) at the European Nucleotide
Archive (ENA)</a>. There are many repositories for public data. Some
model organisms or fields have specific databases, and there are ones
for particular types of data. Two of the most comprehensive are the
National Center for Biotechnology Information (NCBI) and European
Nucleotide Archive (EMBL-EBI). In this lesson we’re working with the
ENA, but the general process is the same for any database.</p>
<p>We can do this one by one but given that each download takes about
one to two hours, this could keep us up all night. Instead of
downloading one by one we can apply a loop. Let’s see what that looks
like and then we’ll discuss what we’re doing with each line of our
loop.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> ERR029207 ERR029206 ERR026478 ERR026474 ERR026473 ERR026481 ERR026482</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>  <span class="bu">echo</span> <span class="at">-e</span> <span class="st">"</span><span class="va">$filename</span><span class="st">"</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>  <span class="fu">wget</span> <span class="at">--spider</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/<span class="st">"</span><span class="va">${filename</span><span class="op">:</span><span class="dv">0</span><span class="op">:</span><span class="dv">6</span><span class="va">}</span><span class="st">"</span>/<span class="st">"</span><span class="va">${filename}</span><span class="st">"</span>/<span class="st">"</span><span class="va">${filename}</span><span class="st">"</span>_<span class="pp">*</span>.fastq.gz</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>When the shell sees the keyword <code>for</code>, it knows to repeat
a command (or group of commands) once for each item in a list. Each time
the loop runs (called an iteration), an item in the list is assigned in
sequence to the <strong>variable</strong>, and the commands inside the
loop are executed, before moving on to the next item in the list.</p>
<p>Inside the loop, we call for the variable’s value by putting
<code>$</code> in front of it. The <code>$</code> tells the shell
interpreter to treat the <strong>variable</strong> as a variable name
and substitute its value in its place, rather than treat it as text or
an external command.</p>
<p>In this example, the list is seven accession numbers of reads
belonging to the genomes we are interested in. Each time the loop
iterates, it will assign a file name to the variable
<code>filename</code> and <code>echo</code> (print) the filename. In the
second line, it will run the <code>wget</code> command.</p>
<p>The first time through the loop, <code>$filename</code> is
<code>ERR029207</code>. The FTP site at EBI is constructed so that fastq
files are stored in a <code>vol1/fastq</code> subfolder. Read files are
not directly in that subfolder, but grouped in subfolders starting with
the six first characters of the accession number (<code>ERR029</code> in
the first case), then in the accession number subfolder. To get to the
right place, the file path is constructed with the root path
(<code>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/</code>), to which the first 6
characters of the accession number is added through that special
variable operation (<code>"${filename:0:6}"</code>) to get to the right
subfolder. Finally, the correct accession subfolder
(<code>"${filename}"</code>), the file name prefix
(<code>"${filename}"_*.fastq.gz"</code>) and the suffix
(<code>_*.fastq.gz</code>) are added. The <code>*</code> ensures that we
get reads from both ends.</p>
<p>Use <code>{}</code> to wrap the variable so that
<code>.fastq.gz</code> will not be interpreted as part of the variable
name. In addition, quoting the shell variables is a good practice AND
necessary if your variables have spaces in them.</p>
<p>For the second iteration, <code>$filename</code> becomes
<code>ERR029206</code>.</p>
<p>We added a <code>--spider</code> option to <code>wget</code> just for
this exercise, to retrieve only the names of the files and not the
actual file, to spare some time downloading files. The data is present
in the <code>data/fastq</code> subfolder of our group folder:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">ls</span> /proj/g2020004/nobackup/3MK013/data/fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ERR026473_1.fastq.gz  ERR026474_2.fastq.gz  ERR026481_1.fastq.gz  ERR026482_2.fastq.gz	ERR029207_1.fastq.gz
ERR026473_2.fastq.gz  ERR026478_1.fastq.gz  ERR026481_2.fastq.gz  ERR029206_1.fastq.gz	ERR029207_2.fastq.gz
ERR026474_1.fastq.gz  ERR026478_2.fastq.gz  ERR026482_1.fastq.gz  ERR029206_2.fastq.gz</code></pre>
</div>
<p>For more, check <a href="https://mywiki.wooledge.org/BashPitfalls" class="external-link">Bash Pitfalls</a></p>
<div id="follow-the-prompt" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="follow-the-prompt" class="callout-inner">
<h3 class="callout-title">Follow the Prompt</h3>
<div class="callout-content">
<p>The shell prompt changes from <code>$</code> to <code>&gt;</code> and
back again as we were typing in our loop. The second prompt,
<code>&gt;</code>, is different to remind us that we haven’t finished
typing a complete command yet. A semicolon, <code>;</code>, can be used
to separate two commands written on a single line.</p>
</div>
</div>
</div>
<div id="same-symbols-different-meanings" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="same-symbols-different-meanings" class="callout-inner">
<h3 class="callout-title">Same Symbols, Different Meanings</h3>
<div class="callout-content">
<p>Here we see <code>&gt;</code> being used a shell prompt, whereas
<code>&gt;</code> is also used to redirect output. Similarly,
<code>$</code> is used as a shell prompt, but, as we saw earlier, it is
also used to ask the shell to get the value of a variable.</p>
<p>If the <em>shell</em> prints <code>&gt;</code> or <code>$</code> then
it expects you to type something, and the symbol is a prompt.</p>
<p>If <em>you</em> type <code>&gt;</code> or <code>$</code> yourself, it
is an instruction from you that the shell to redirect output or get the
value of a variable.</p>
</div>
</div>
</div>
<p>We have called the variable in this loop <code>filename</code> in
order to make its purpose clearer to human readers. The shell itself
doesn’t care what the variable is called; if we wrote this loop as:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> ERR01 ERR02 ERR03</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>  <span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${x}</span><span class="st">"</span>.fastq.gz</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>or:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="cf">for</span> pizza <span class="kw">in</span> ERR01 ERR02 ERR03</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>  <span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${pizza}</span><span class="st">"</span>.fastq.gz</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>it would work exactly the same way. <em>Don’t do this.</em> Programs
are only useful if people can understand them, so meaningless names
(like <code>x</code>) or misleading names (like <code>pizza</code>)
increase the odds that the program won’t do what its readers think it
does.</p>
<div id="multipart-commands" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="multipart-commands" class="callout-inner">
<h3 class="callout-title">Multipart commands</h3>
<div class="callout-content">
<p>The <code>for</code> loop is interpreted as a multipart command. If
you press the up arrow on your keyboard to recall the command, it may be
shown like so (depends on your bash version):</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> ERR01 ERR02 ERR03<span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> ftp://ftp.sra.ebi.ac.uk/<span class="st">"</span><span class="va">${filename}</span><span class="st">"</span>.fastq.gz <span class="kw">;</span> <span class="cf">done</span></span></code></pre>
</div>
<p>When you check your history later, it will help your remember what
you did!</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="link-the-read-files">Link the read files<a class="anchor" aria-label="anchor" href="#link-the-read-files"></a>
</h3>
<p>Now link the files from the <code>data</code> folder, to perform
quality control.</p>
<div id="challenge6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>All the read files start with ERR and have a <code>fastq.gz</code>
extension. It means they are in fastq format and compressed. Softlink
the files in your <code>molepi/data</code> folder, using a nested loop.
The <code>data/fastq</code> folder is three levels up. For each
accession number there are two <code>fastq.gz</code> files, one for each
end (1 or 2).</p>
</div>
</div>
</div>
<div id="accordionHint2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint2" aria-expanded="false" aria-controls="collapseHint2">
  <h4 class="accordion-header" id="headingHint2"> Give me a hint </h4>
</button>
<div id="collapseHint2" class="accordion-collapse collapse" data-bs-parent="#accordionHint2" aria-labelledby="headingHint2">
<div class="accordion-body">
<p>Softlinks are done with <code>ln -s</code>. Use the same loop
backbone as above, and nest a <code>for</code> loop over the two ends of
the reads (1 and 2).</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/data</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> <span class="op">&lt;</span>...<span class="op">&gt;</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>  <span class="cf">for</span> <span class="op">&lt;</span>loop <span class="ex">over</span> 1 <span class="dv">2</span><span class="op">&gt;</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>    <span class="op">&lt;</span>softlink<span class="op">&gt;</span> ../../../data/fastq/<span class="op">&lt;</span>build <span class="ex">the</span> filename with $ and double quotes<span class="op">&gt;</span>.fastq.gz .</span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor4" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor4" aria-expanded="false" aria-controls="collapseInstructor4">
  <h3 class="accordion-header" id="headingInstructor4">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor4" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor4" aria-labelledby="headingInstructor4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/data</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> ERR029207 ERR029206 ERR026478 ERR026474 ERR026473 ERR026481 ERR026482</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>  <span class="cf">for</span> end <span class="kw">in</span> 1 2</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>  <span class="cf">do</span></span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>    <span class="fu">ln</span> <span class="at">-s</span> ../../../data/fastq/<span class="st">"</span><span class="va">${filename}</span><span class="st">"</span>_<span class="st">"</span><span class="va">${end}</span><span class="st">"</span>.fastq.gz .</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb24-8"><a href="#cb24-8" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="bioinformatics-workflows">Bioinformatics workflows<a class="anchor" aria-label="anchor" href="#bioinformatics-workflows"></a>
</h2>
<hr class="half-width">
<p>When working with high-throughput sequencing data, the raw reads you
get off of the sequencer will need to pass through a number of different
tools in order to generate your final desired output. The execution of
this set of tools in a specified order is commonly referred to as a
<em>workflow</em> or a <em>pipeline</em>.</p>
<p>An example of the workflow we will be using is provided below with a
brief description of each step.</p>
<figure><img src="../fig/Workflow.png" alt="Schematic workflow of the bioinformatics workflow" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal">
<li>Quality control - Assessing quality using FastQC</li>
<li>Quality control - Trimming and/or filtering reads (if
necessary)</li>
<li>Align reads to reference genome - Snippy</li>
<li>Variant calling - Snippy</li>
<li>Clustering variants - iTOL</li>
<li>Assembly - SKESA</li>
<li>Annotation (extra material)</li>
<li>Pangenome analysis (extra material)</li>
<li>Clustering presence and absence of genes (extra material)</li>
<li>Comparison of clustering methods (extra material)</li>
<li>Data visualization - Microreact</li>
</ol>
<p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community. The
tools that are used to analyze data at different stages of the workflow
are therefore built under the assumption that the data will be provided
in a specific format.</p>
</section><section><h2 class="section-heading" id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<hr class="half-width">
<p>The first step in the variant calling workflow is to take the FASTQ
files received from the sequencing facility and assess the quality of
the sequence reads.</p>
<figure><img src="../fig/var_calling_workflow_qc.png" alt="Quality control workflow" class="figure mx-auto d-block"></figure><p>You’ve already seen that part, but repetition doesn’t hurt. If you
feel that it’s anyway unnecessary, move on to the next part.</p>
<div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a>
</h3>
<p>Although it looks complicated (and it is), it’s easy to understand
the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">fastq</a>
format with a little decoding. Some rules about the format include…</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="83%">
</colgroup>
<thead><tr class="header">
<th>Line</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Always begins with ‘@’ and then information about the read</td>
</tr>
<tr class="even">
<td>2</td>
<td>The actual DNA sequence</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Always begins with a ‘+’ and sometimes the same info in line 1</td>
</tr>
<tr class="even">
<td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr>
</tbody>
</table>
<p>We can view the first complete read in one of the files our dataset
by using <code>head</code> to look at the first four lines.</p>
<div id="challenge7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>You will use zcat to extract text from a gzipped file.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@ERR029206.1 IL31_5505:8:1:6233:1087#4/1
TCNAGTCAGCACACACATGCGAAAGAATCCACCGACTAGGGTCAGCGGGGTTTGCAGTTGGTCGCGGACGTAACCG
+
&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
</div>
</div>
</div>
<div id="accordionHint3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint3" aria-expanded="false" aria-controls="collapseHint3">
  <h4 class="accordion-header" id="headingHint3"> Give me a hint </h4>
</button>
<div id="collapseHint3" class="accordion-collapse collapse" data-bs-parent="#accordionHint3" aria-labelledby="headingHint3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">zcat</span> <span class="at">-h</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="fu">head</span> <span class="at">-h</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">zcat</span> ERR029206_1.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n4</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>One of the nucleotides in this read is unknown (<code>N</code>).</p>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g. 1 in 10)
or, equivalently, the base call accuracy (eg 90%). To make it possible
to line up each individual nucleotide with its quality score, the
numerical score is converted into a code where each individual character
represents the numerical quality score for an individual nucleotide. For
example, in the line above, the quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>The <code>&lt;</code> character and each of the <code>=</code>
characters represent the encoded quality for an individual nucleotide.
The numerical value assigned to each of these characters depends on the
sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using by Illumina version 1.8 onwards. Each character is
assigned a quality score between 0 and 40 as shown in the chart
below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI
                  |         |         |         |         |
Quality score:    0........10........20........30........40                                </code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. This quality score is logarithmically
based, so a quality score of 10 reflects a base call accuracy of 90%,
but a quality score of 20 reflects a base call accuracy of 99%. These
probability values are the results from the base calling algorithm and
dependent on how much signal was captured for the base
incorporation.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@ERR029206.1 IL31_5505:8:1:6233:1087#4/1
TCNAGTCAGCACACACATGCGAAAGAATCCACCGACTAGGGTCAGCGGGGTTTGCAGTTGGTCGCGGACGTAACCG
+
&lt;='=ABBBBCBBBBBBBBC&gt;BBCBBBBBB@BBBBBB=BCBB@@B@@BB@@,@@@@@@@@@@@@@@@@@@@@@@@@@</code></pre>
</div>
<p>We can now see that the quality of an <code>N</code> (<code>'</code>)
is 6, which corresponds to a over 25% error probability:</p>
<p><span class="math inline">\(10^{-\frac{6}{10}} \cong
0.25\)</span></p>
</div>
<div class="section level3">
<h3 id="quality-encodings-vary">Quality Encodings Vary<a class="anchor" aria-label="anchor" href="#quality-encodings-vary"></a>
</h3>
<p>Although we’ve used a particular quality encoding system to
demonstrate interpretation of read quality, different sequencing
machines use different encoding systems. This means that, depending on
which sequencer you use to generate your data, a <code>#</code> may not
be an indicator of a poor quality base call.</p>
<div id="callout4" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Callout</h3>
<div class="callout-content">
<p>This mainly relates to older Solexa/Illumina data, but it’s essential
that you know which sequencing platform was used to generate your data,
so that you can tell your quality control program which encoding to use.
If you choose the wrong encoding, you run the risk of throwing away good
reads or (even worse) not throwing away bad reads!</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="assessing-quality-using-fastqc">Assessing Quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a>
</h3>
<p>In real life, you won’t be assessing the quality of your reads by
visually inspecting your FASTQ files. Rather, you’ll be using a software
program to assess read quality and filter out poor quality reads. We’ll
first use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads.Later in our workflow, we’ll use
another program to filter out poor quality reads.</p>
<p>FastQC has a number of features which can give you a quick impression
of any problems your data may have, so you can take these issues into
consideration before moving forward with your analyses. Rather than
looking at quality scores for each individual read, FastQC looks at
quality collectively across all reads within a sample. The image below
shows a FastQC-generated plot that indicates a very high quality
sample:</p>
<figure><img src="../fig/good_quality1.8.png" alt="A FastQC output showing reads with good quality" class="figure mx-auto d-block"></figure><p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. For each position, there is a box-and-whisker plot
showing the distribution of quality scores for all reads at that
position. The horizontal red line indicates the median quality score and
the yellow box shows the 2nd to 3rd quartile range. This means that 50%
of reads have a quality score that falls within the range of the yellow
box at that position. The whiskers show the range to the 1st and 4th
quartile.</p>
<p>For each position in this sample, the quality values do not drop much
lower than 32. This is a high quality score. The plot background is also
color-coded to identify good (green), acceptable (yellow), and bad (red)
quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the
spectrum.</p>
<figure><img src="../fig/bad_quality1.8.png" alt="A FastQC output showing reads with bad quality" class="figure mx-auto d-block"></figure><p>Here, we see positions within the read in which the boxes span a much
wider range. Also, quality scores drop quite low into the “bad” range,
particularly on the tail end of the reads. The FastQC tool produces
several other diagnostic plots to assess sample quality, in addition to
the one plotted above.</p>
</div>
<div class="section level3">
<h3 id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a>
</h3>
<p>Navigate to your FASTQ dataset:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/data</span></code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>We softlinked the samples. How many FASTQ files are in this dataset?
Why is this? What is the difference between the two files that start
with the same name? Discuss with your neighbour.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<p>There are 14 FASTQ files. Each sample has <a href="https://thesequencingcenter.com/knowledge-base/what-are-paired-end-reads/" class="external-link">paired-end
reads</a>.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, including files that
are compressed (<code>.gz</code>) so we can use the
<code>*.fastq.gz</code> wildcard to run FastQC on both the compressed
FASTQ files for sample ERR029206 in this directory.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="ex">module</span> load FastQC</span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a><span class="ex">fastqc</span> ERR029206<span class="pp">*</span>.fastq.gz</span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of ERR029206_1.fastq.gz
Approx 5% complete for ERR029206_1.fastq.gz
Approx 10% complete for ERR029206_1.fastq.gz
Approx 15% complete for ERR029206_1.fastq.gz
Approx 20% complete for ERR029206_1.fastq.gz
Approx 25% complete for ERR029206_1.fastq.gz
Approx 30% complete for ERR029206_1.fastq.gz</code></pre>
</div>
<p>In total, it should take a couple minutes minutes for FastQC to run
on our two FASTQ files. When the analysis completes, your prompt will
return. So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 85% complete for ERR029206_2.fastq.gz
Approx 90% complete for ERR029206_2.fastq.gz
Approx 95% complete for ERR029206_2.fastq.gz
Analysis complete for ERR029206_2.fastq.gz
data $</code></pre>
</div>
<p>The FastQC program has created several new files within our
<code>/data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="fu">ls</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
[...]
ERR029206_1.fastq.gz
ERR029206_1_fastqc.html
ERR029206_1_fastqc.zip
ERR029206_2.fastq.gz
ERR029206_2_fastqc.html
ERR029206_2_fastqc.zip
[...]</code></pre>
</div>
<p>For each input gzipped FASTQ file, FastQC has created a
<code>.zip</code> file and a <code>.html</code> file. The
<code>.zip</code> file extension indicates that this is actually a
compressed set of multiple output files. We’ll be working with these
output files soon. The <code>.html</code> file is a stable webpage
displaying the summary report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">mkdir</span> ../results/fastqc_untrimmed_reads</span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.zip ../results/fastqc_untrimmed_reads/</span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>.html ../results/fastqc_untrimmed_reads/</span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="bu">cd</span> ../results/fastqc_untrimmed_reads/</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="viewing-html-files">Viewing HTML files<a class="anchor" aria-label="anchor" href="#viewing-html-files"></a>
</h3>
<p>If we are working on our local computers, we may display each of
these HTML files as a webpage. Otherwise, copy the files to your local
computer using <code>scp</code> and use your browser:</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="ex">firefox</span> ERR029206_1_fastqc.html</span></code></pre>
</div>
<p>You may need to replace the command <code>firefox</code> with the
name of another web navigator if you have one.</p>
<div id="exercise-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Do the sample looks good in
terms of per base sequence quality?</p>
</div>
</div>
</div>
<div id="exercise-3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>What are the read lengths of the different samples?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="cleaning-reads">Cleaning Reads<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a>
</h2>
<hr class="half-width">
<p>It’s very common to have some reads within a sample, or some
positions (near the beginning or end of reads) across all reads that are
low quality and should be discarded. We will use a program called <a href="https://github.com/lh3/seqtk" class="external-link">seqtk</a> to filter poor quality
reads and trim poor quality bases from our samples.</p>
<div class="section level3">
<h3 id="seqtk-options">Seqtk Options<a class="anchor" aria-label="anchor" href="#seqtk-options"></a>
</h3>
<p>Seqtk is a program written in C and aims to be a Swiss Army Knife for
sequencing reads. You don’t need to learn C to use Seqtk, but the fact
that it’s a C program helps explain the syntax that is used to run
Seqtk. Seqtk takes as input files either FASTQ files or gzipped FASTQ
files and outputs FASTQ or FASTA files. The basic command to run Seqtk
starts like this:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="ex">module</span> load seqtk</span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a><span class="ex">seqtk</span></span></code></pre>
</div>
<p>That’s just the basic command, however. Seqtk has a variety of
options and parameters. We will need to specify what options we want to
use for our analysis. Here are some of the options:</p>
<table class="table">
<thead><tr class="header">
<th>option</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>seq</code></td>
<td>common transformation of FASTA/Q</td>
</tr>
<tr class="even">
<td><code>comp</code></td>
<td>get the nucleotide composition of FASTA/Q</td>
</tr>
<tr class="odd">
<td><code>trimfq</code></td>
<td>trim FASTQ using the Phred algorithm</td>
</tr>
</tbody>
</table>
<p>In addition to these options, there are a number if trimming options
available:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="ex">seqtk</span> trimfq</span></code></pre>
</div>
<table class="table">
<thead><tr class="header">
<th>step</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>-q</code></td>
<td>error rate threshold (disabled by -b/-e) [0.05]</td>
</tr>
<tr class="even">
<td><code>-l</code></td>
<td>maximally trim down to INT bp (disabled by -b/-e) [30]</td>
</tr>
<tr class="odd">
<td><code>-b</code></td>
<td>trim INT bp from left (non-zero to disable -q/-l) [0]</td>
</tr>
<tr class="even">
<td><code>-e</code></td>
<td>trim INT bp from right (non-zero to disable -q/-l) [0].</td>
</tr>
</tbody>
</table>
<p>We will use only a few of these options in our analysis. It is
important to understand the steps you are using to clean your data.</p>
<p>A complete command for trimming with seqtk will look something like
this:</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="ex">seqtk</span> trimfq <span class="at">-q</span> 0.01 ERR01_1.fastq.gz <span class="op">&gt;</span> ERR01_1_trim.fastq</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="trimming">Trimming<a class="anchor" aria-label="anchor" href="#trimming"></a>
</h3>
<p>Now we will run seqtk trimfq on our data. To begin, make sure you are
still in your <code>data</code> directory:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre>
</div>
<p>We are going to run seqtk on one sample giving it an error rate
threshold of 0.01 which indicates the base call accuracy. We request
that, after trimming, the chances that a base is called incorrectly are
only 1 in 10000.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="ex">seqtk</span> trimfq <span class="at">-q</span> 0.01 ERR029206_1.fastq.gz <span class="op">&gt;</span> ERR029206_1_trim.fastq</span></code></pre>
</div>
<p>Notice that we needed to redirect the output to a file. If we don’t
do that, the trimmed fastq data will be displayed in the console.</p>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Use seqtk fqchk to compare the untrimmed and trimmed reads of
ERR029206_1 in terms of number of sequenced bases, percentage of A,G,C,T
and N and average quality. What do you notice? Discuss with your
neighbor.</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" data-bs-parent="#accordionSolution9" aria-labelledby="headingSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 76; max_len: 76; avg_len: 76.00; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 315791552   18.0    32.1    32.2    17.7    0.0 30.6    21.4    5.4 94.6</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 69.12; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 287211920   18.2    32.0    31.9    17.8    0.0 31.7    28.8    1.3 98.7</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We’ve just successfully trimmed one of our FASTQ files! However,
there is some bad news. <code>seqtk</code> can only operate on one
sample at a time and we have more than one sample. The good news is that
we can use a <code>for</code> loop to iterate through our sample files
quickly! This will take a few minutes.</p>
<div id="challenge10" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Write a loop which trims all <code>.fastq.gz</code> files and
redirect the output into <code>_trim.fastq</code> files.</p>
</div>
</div>
</div>
<div id="accordionHint4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint4" aria-expanded="false" aria-controls="collapseHint4">
  <h4 class="accordion-header" id="headingHint4"> Give me a hint </h4>
</button>
<div id="collapseHint4" class="accordion-collapse collapse" data-bs-parent="#accordionHint4" aria-labelledby="headingHint4">
<div class="accordion-body">
<p>Create a new variable, built from the input file name (ending in
<code>.fastq.gz</code>).</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span></code></pre>
</div>
<p><code>infile</code> is the first variable in our loop and takes the
value of each of the FASTQ files in our directory. <code>outfile</code>
is the second variable in our loop and is defined by adding
<code>_trim.fastq</code> to the end of our input file name. Use
<code>{}</code> to wrap the variable so that <code>_trim.fastq</code>
will not be interpreted as part of the variable name. In addition,
quoting the shell variables is a good practice AND necessary if your
variables have spaces in them.</p>
<p>For more, check <a href="https://mywiki.wooledge.org/BashPitfalls" class="external-link">Bash Pitfalls</a>. There
are no spaces before or after the <code>=</code>.</p>
</div>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" data-bs-parent="#accordionSolution10" aria-labelledby="headingSolution10">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="cf">for</span> <span class="op">&lt;</span>loop <span class="ex">over</span> infiles ending with .fastq.gz<span class="op">&gt;</span></span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a>  <span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a>  <span class="ex">seqtk</span> trimfq <span class="op">&lt;</span>error prob at most 0.01<span class="op">&gt;</span> <span class="op">&lt;</span>infile<span class="op">&gt;</span> <span class="op">&gt;</span> <span class="op">&lt;</span>outfile<span class="op">&gt;</span></span>
<span id="cb50-5"><a href="#cb50-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionInstructor5" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor5" aria-expanded="false" aria-controls="collapseInstructor5">
  <h3 class="accordion-header" id="headingInstructor5">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor5" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor5" aria-labelledby="headingInstructor5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="cf">for</span> infile <span class="kw">in</span> <span class="pp">*</span>.fastq.gz</span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb51-3"><a href="#cb51-3" tabindex="-1"></a>  <span class="va">outfile</span><span class="op">=</span><span class="st">"</span><span class="va">${infile}</span><span class="st">"</span>_trim.fastq</span>
<span id="cb51-4"><a href="#cb51-4" tabindex="-1"></a>  <span class="ex">seqtk</span> trimfq <span class="at">-q</span> 0.01 <span class="st">"</span><span class="va">${infile}</span><span class="st">"</span> <span class="op">&gt;</span>  <span class="st">"</span><span class="va">${outfile}</span><span class="st">"</span> </span>
<span id="cb51-5"><a href="#cb51-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Go ahead and run the for loop. It should take a few minutes for seqtk
to run for each of our fourteen input files. Once it’s done running,
take a look at your directory contents.</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">ls</span> ERR029206<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
ERR029206_1.fastq.gz_trim.fastq  ERR029206_2.fastq.gz_trim.fastq</code></pre>
</div>
<p>We’ve now completed the trimming and filtering steps of our quality
control process! Before we move on, let’s move our trimmed FASTQ files
to a new subdirectory within our <code>data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="fu">mkdir</span> trimmed_fastq</span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="fu">mv</span> <span class="pp">*</span>fastq.gz_trim<span class="pp">*</span> trimmed_fastq</span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a><span class="bu">cd</span> trimmed_fastq</span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a><span class="fu">ls</span> ERR029206<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">...</span></span>
<span><span class="va">ERR029206_1.fastq.gz_trim.fastq</span></span>
<span><span class="va">ERR029206_2.fastq.gz_trim.fastq</span></span></code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Again, use <code>seqtk fqchk</code> to compare the untrimmed and
trimmed reads of both samples. Note the number of bases ‘#bases’ of the
trimmed and untrimmed reads. Calculate the theoretical coverage of the
genomes before and after trimming, assuming that all our genomes do have
the same size as our reference genome (4411532 bases).</p>
<p>Hint: Sum up forward and reverse reads!</p>
</div>
</div>
</div>
<div id="accordionSolution11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution11" aria-expanded="false" aria-controls="collapseSolution11">
  <h4 class="accordion-header" id="headingSolution11"> Show me the solution </h4>
</button>
<div id="collapseSolution11" class="accordion-collapse collapse" data-bs-parent="#accordionSolution11" aria-labelledby="headingSolution11">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_1.fastq.gz_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 69.12; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 287211920   18.2    32.0    31.9    17.8    0.0 31.7    28.8    1.3 98.7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="ex">seqtk</span> fqchk ERR029206_2.fastq.gz_trim.fastq <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 3</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>min_len: 30; max_len: 76; avg_len: 63.79; 36 distinct quality values
POS #bases  %A  %C  %G  %T  %N  avgQ    errQ    %low    %high
ALL 265048852   18.1    32.1    32.1    17.7    0.0 31.2    25.7    2.5 97.5</code></pre>
</div>
<p>Coverage = #bases (forward + reverse) / genome size In this case:</p>
<p><span class="math inline">\(125.18 = \dfrac{(287211920 +
265048852)}{4411532}\)</span></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>
<code>wget</code> is a computer program to get data from the
internet</li>
<li>
<code>for</code> loops let you perform the same set of operations on
multiple files with a single command</li>
<li>Sequencing data is large</li>
<li>In bioinformatic workflows the output of one tool is the input of
the other.</li>
<li>FastQC is used to judge the quality of sequencing reads.</li>
<li>Data cleaning is an essential step in a genomics pipeline.</li>
<li>We will work towards confirming or disputing transmission in TB
cases</li>
<li>After this practical training you will have some familiarity with
working on the command line</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-sequence-assembly"><p>Content from <a href="sequence-assembly.html">Sequence assembly</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/sequence-assembly.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>Sequence assembly means the alignment and merging of reads in order
to reconstruct the original sequence. The assembly of a genome from
short sequencing reads can take a while. We will therefore run this
process only for two genomes.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can the information in the sequencing reads be reduced?</li>
<li>What are the different methods for assembly?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand differences between assembly methods</li>
<li>Assemble the short reads</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="exercise-1-sequence-assembly-with-skesa">Exercise 1: Sequence assembly with SKESA<a class="anchor" aria-label="anchor" href="#exercise-1-sequence-assembly-with-skesa"></a>
</h2>
<hr class="half-width">
<p>The assembler we will run is <a href="https://github.com/ncbi/SKESA" class="external-link">SKESA</a>. SKESA generates a final
assembly from multiple kmers. A list of kmers is automatically selected
by SKESA using the maximum read length of the input data, and each
individual kmer contributes to the final assembly. If desired, the
lengths of kmers can be specified with the <code>--kmer</code> and
<code>--steps</code> flags which will adjust automatic kmer
selection.</p>
<p>Because assembly of each genome would take a long time, you will only
run one assembly here.</p>
<div class="section level3">
<h3 id="preparation">Preparation<a class="anchor" aria-label="anchor" href="#preparation"></a>
</h3>
<p>Start the interactive session, go to the <code>molepi</code>
subfolder, create a <code>results/assembly</code> subfolder. Then go to
the folder containing the fastq files.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2025-3-4 <span class="at">-M</span> snowy <span class="at">-t</span> 04:00:00</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">mkdir</span> results/assembly</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="bu">cd</span> data/trimmed_fastq</span></code></pre>
</div>
<p>To run SKESA we will use the skesa command with a number of option
that we will explore later on. We can start the loop (again, a short
loop with only one sample in this actual case to spare computing
resources) with the assemblies with:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools SKESA</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> ERR029206</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">skesa</span> <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>--cores 2 <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>--memory 8 <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>--fastq <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_1.fastq.gz_trim.fastq <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_2.fastq.gz_trim.fastq <span class="dv">1</span><span class="op">&gt;</span> ../../results/assembly/<span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>.fasta</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>A non-quoted backslash, <code>\</code> is used as an escape character
in Bash. It preserves the literal value of the next character that
follows, with the exception of newline. This means that the back slash
starts a new line without starting a new command - we only add it for
better readability.</p>
<p>The assembly should take about 5 minutes per genome, using 2 cores
and 8 GB of memory.</p>
</div>
<div class="section level3">
<h3 id="downloading-the-missing-assemblies">Downloading the missing assemblies<a class="anchor" aria-label="anchor" href="#downloading-the-missing-assemblies"></a>
</h3>
<p>If you haven’t assembled all samples, download the finished contigs
to your computer and decompress them. They will decompress in a folder
called <code>assembly</code>, so one solution is to rename the assembly
folder you already have to something else:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="bu">cd</span> ../../results/</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">mv</span> assembly partial_assembly</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">tar</span> xvzf /proj/g2020004/nobackup/3MK013/data/assemblies.tar.gz</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">ls</span> assembly</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ERR026473.fasta ERR026478.fasta ERR026482.fasta ERR029207.fasta
ERR026474.fasta ERR026481.fasta ERR029206.fasta</code></pre>
</div>
<div id="challenge-what-do-the-command-line-parameters-of-skesa-mean" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-what-do-the-command-line-parameters-of-skesa-mean" class="callout-inner">
<h3 class="callout-title">Challenge: What do the command line parameters of SKESA mean?</h3>
<div class="callout-content">
<p>Find out a) the version of SKESA you are using and b) what the used
command line parameters mean.</p>
<p>Hint:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">skesa</span> <span class="at">-h</span></span></code></pre>
</div>
<p>prints the ‘help’</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>skesa -h
..
General options:
-h [ --help ]                 Produce help message
-v [ --version ]              Print version
..
--cores arg (=0)              Number of cores to use (default all) [integer]
--memory arg (=32)            Memory available (GB, only for sorted counter)
..
--fastq arg                   Input fastq file(s) (could be used multiple times for different runs) [string]
..
..</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="examine-the-results-of-the-assembly">Examine the results of the assembly<a class="anchor" aria-label="anchor" href="#examine-the-results-of-the-assembly"></a>
</h3>
<p>You will now explore the files output by SKESA. Find out the number
of contigs, and the various assembly statistics.</p>
<p>What do the they mean? What are N50, L50, N80 and L80?</p>
<div id="challenge-how-many-contigs-were-generated-by-skesa" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-how-many-contigs-were-generated-by-skesa" class="callout-inner">
<h3 class="callout-title">Challenge: How many contigs were generated by SKESA?</h3>
<div class="callout-content">
<p>Find out how many contigs there are in the <em>M. tuberculosis</em>
isolates.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">module</span> load GenomeTools </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">gt</span> seqstat ERR026473.fasta</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="co"># number of contigs:     390</span></span>
<span><span class="co"># total contigs length:  4239956</span></span>
<span><span class="co"># mean contig size:      10871.68</span></span>
<span><span class="co"># contig size first quartile: 989</span></span>
<span><span class="co"># median contig size:         5080</span></span>
<span><span class="co"># contig size third quartile: 15546</span></span>
<span><span class="co"># longest contig:             74938</span></span>
<span><span class="co"># shortest contig:            200</span></span>
<span><span class="co"># contigs &gt; 500 nt:           323 (82.82 %)</span></span>
<span><span class="co"># contigs &gt; 1K nt:            291 (74.62 %)</span></span>
<span><span class="co"># contigs &gt; 10K nt:           148 (37.95 %)</span></span>
<span><span class="co"># contigs &gt; 100K nt:          0 (0.00 %)</span></span>
<span><span class="co"># contigs &gt; 1M nt:            0 (0.00 %)</span></span>
<span><span class="co"># N50                25167</span></span>
<span><span class="co"># L50                54</span></span>
<span><span class="co"># N80                11379</span></span>
<span><span class="co"># L80                129</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-compare-the-results-of-the-different-samples" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-compare-the-results-of-the-different-samples" class="callout-inner">
<h3 class="callout-title">Challenge: Compare the results of the different samples</h3>
<div class="callout-content">
<p>How could you get N50 stats for all the assemblies?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="cf">for</span> accession <span class="kw">in</span> ERR029207 ERR029206 ERR026478 ERR026474 ERR026473 ERR026481 ERR026482</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"</span><span class="va">${accession}</span><span class="st">"</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="ex">gt</span> seqstat <span class="st">"</span><span class="va">${accession}</span><span class="st">"</span>.fasta <span class="kw">|</span> <span class="fu">grep</span> N50</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionHint1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button hint-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseHint1" aria-expanded="false" aria-controls="collapseHint1">
  <h4 class="accordion-header" id="headingHint1"> Give me a hint </h4>
</button>
<div id="collapseHint1" class="accordion-collapse collapse" data-bs-parent="#accordionHint1" aria-labelledby="headingHint1">
<div class="accordion-body">
<p>Hint: <a href="https://en.wikipedia.org/wiki/N50,_L50,_and_related_statistics" class="external-link uri">https://en.wikipedia.org/wiki/N50,_L50,_and_related_statistics</a>
¨¨</p>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-2-assembly-with-spades">Exercise 2: assembly with SPAdes<a class="anchor" aria-label="anchor" href="#exercise-2-assembly-with-spades"></a>
</h2>
<hr class="half-width">
<p>To compare how different assemblers perform, you will also use <a href="https://ablab.github.io/spades/" class="external-link">SPAdes</a>. SPAdes is a very
versatile and can assemble many different data types, including
combining long and short reads.</p>
<div id="challenge-assemble-the-genomes-with-spades" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-assemble-the-genomes-with-spades" class="callout-inner">
<h3 class="callout-title">Challenge: Assemble the genomes with SPAdes</h3>
<div class="callout-content">
<p>You will only perform the assembly on one genome, but use the loop
form anyway. You will use the <code>--isolate</code> mode, but the
<code>--careful</code> mode would also be interesting. This might take
10-15 minutes per genome.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">module</span> load spades</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="fu">mkdir</span> spades_assembly</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="bu">cd</span> spades_assembly</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="cf">for</span> accession <span class="kw">in</span> ERR029206</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"</span><span class="va">${accession}</span><span class="st">"</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>  <span class="ex">spades.py</span> <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>    <span class="at">--isolate</span> <span class="dt">\</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>    <span class="at">-1</span> ../../data/trimmed_fastq/<span class="st">"</span><span class="va">${accession}</span><span class="st">"</span>_1.fastq.gz_trim.fastq <span class="dt">\</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>    <span class="at">-2</span> ../../data/trimmed_fastq/<span class="st">"</span><span class="va">${accession}</span><span class="st">"</span>_2.fastq.gz_trim.fastq <span class="dt">\</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>    <span class="at">-o</span> <span class="st">"</span><span class="va">${accession}</span><span class="st">"</span>_spades <span class="dt">\</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>    <span class="at">-t</span> 2 <span class="dt">\</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>    <span class="at">-m</span> 8</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-compare-the-results-of-the-assembly-with-spades-and-skesa" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-compare-the-results-of-the-assembly-with-spades-and-skesa" class="callout-inner">
<h3 class="callout-title">Challenge: Compare the results of the assembly with SPAdes and SKESA</h3>
<div class="callout-content">
<p>Using Genome Tools, compare the result of the two assemblies</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="ex">gt</span> seqstat assembly/ERR029206.fasta</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="ex">gt</span> seqstat spades_assembly/ERR029206_spades/contigs.fasta</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exercise-3-preliminary-annotation">Exercise 3: Preliminary annotation<a class="anchor" aria-label="anchor" href="#exercise-3-preliminary-annotation"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="physical-annotation">Physical annotation<a class="anchor" aria-label="anchor" href="#physical-annotation"></a>
</h3>
<p>The goal here is to find ORFs and start codons, to find proteins in
the genome.</p>
<div id="challenge-use-prodigal" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-use-prodigal" class="callout-inner">
<h3 class="callout-title">Challenge: Use prodigal</h3>
<div class="callout-content">
<p>Find how prodigal works and use it to annotate the proteome of
ERR029206.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">prodigal</span> <span class="at">-h</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">mkdir</span> annotation</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="bu">cd</span> annotation</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="ex">module</span> load prodigal</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="ex">prodigal</span> <span class="at">-i</span> ../assembly/ERR029206.fasta <span class="at">-a</span> ERR029206.faa <span class="op">&gt;</span> ERR029206.gbk</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Examine the resulting files, both the <code>.faa</code> and the
<code>.gbk</code>. Can you make sense of the information there?</p>
</div>
<div class="section level3">
<h3 id="functional-annotation">Functional annotation<a class="anchor" aria-label="anchor" href="#functional-annotation"></a>
</h3>
<p>The physical annotation provided by prodigal identifies genes (ORFs
and start codons), but it does not give any information about the
function of the genes identified. The next step is thus to infer a
function for each of the identified genes. This process usually involves
aligning the proteins to different databases, e.g. using BLAST and other
programs. You will not perform a full functional annotation in this
exercise, but you will try to use blastp to find a RpoB (RNA polymerase
B) homolog in the genome.</p>
<div id="create-a-blast-database" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="create-a-blast-database" class="callout-inner">
<h3 class="callout-title">Create a BLAST database</h3>
<div class="callout-content">
<p>With the proteome of sample ERR029206, create a BLAST database. Use
the <code>makeblastdb</code> function that comes with the BLAST
package.</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">module</span> load</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">makeblastdb</span> <span class="at">-dbtype</span> prot <span class="at">-in</span> ERR029206.faa <span class="at">-out</span> ERR029206</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="find-rpob-homologs" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="find-rpob-homologs" class="callout-inner">
<h3 class="callout-title">Find rpoB homologs</h3>
<div class="callout-content">
<p>Now that you made the database, use the RpoB sequence from <em>E.
coli</em> that you retrieved in the BLAST exercise to find whether there
are one or several RpoB homologs in the proteome you just annotated.
Examine the BLAST results: how many homologs to RpoB in that genome?</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">blastp</span> <span class="at">-query</span> ../../../blast/rpoB_ecoli.fasta <span class="at">-db</span> ERR029206</span></code></pre>
</div>
<p>Just one has E-values at the right level.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Assembly is a process which aligns and merges fragments from a
longer DNA sequence in order to reconstruct the original sequence.</li>
<li>k-mers are short fragments of DNA of length k</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-snps-phylogeny"><p>Content from <a href="snps-phylogeny.html">Mapping</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/snps-phylogeny.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>In this episode we will try to pinpoint single nucleotide variants or
single nucleotide polymorphism (SNPs) between our samples and the
reference. The SNPs are determined by a process called read mapping in
which they are aligned to the reference sequence.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to generate a phylogenetic tree from SNP data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Map reads against a reference genome</li>
<li>Extract single nucleotide polymorphisms</li>
<li>Establish a phylogenetic tree from the SNP data</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<figure><img src="../fig/snps_reads_02.png" alt="Mapping of SNP reads" class="figure mx-auto d-block"></figure><p>The identified SNPs will be used to compare the isolates to each
other and to establish a phylogenetic tree.</p>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Getting missing data </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" aria-labelledby="headingInstructor1" data-bs-parent="#accordionInstructor1">
<div class="accordion-body">
<p>If you don’t have all the trimmed reads, you can download them (will
take a while, <em>dont’t do it if you have the reads</em>):</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2025-3-4 <span class="at">-M</span> snowy <span class="at">-t</span> 04:00:00</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/data</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">mv</span> trimmed_fastq trimmed_fastq_partial</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">tar</span> xvzf ../../../data/trimmed_fastq.tar.gz</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">ls</span> trimmed_fastq</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="snp-calling">SNP calling<a class="anchor" aria-label="anchor" href="#snp-calling"></a>
</h2>
<hr class="half-width">
<p><a href="https://github.com/tseemann/snippy" class="external-link">snippy</a> is a pipeline
that contains different tools to determine SNPs in sequencing reads
against a reference genome. It takes forward and reverse reads of
paired-end sequences and aligns them against a reference.</p>
<p>First we’ll create a folder to hold the results from snippy:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2025-3-4 <span class="at">-M</span> snowy <span class="at">-t</span> 04:00:00</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/data</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">mkdir</span> snps</span></code></pre>
</div>
<p>We need to load snippy:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools snippy</span></code></pre>
</div>
<p>Snippy is fast but a single run still takes about 15 minutes. We
would therefore tell snippy to run all the samples after each other.
However this time we cannot use a wildcard to do so. We would instead
run all the samples in a loop. The “real” loop is commented out, and we
use one that runs only the sample we’ve analyzed so far.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="bu">cd</span> ../data/trimmed_fastq/</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#for sample in ERR026473 ERR026474 ERR026478 ERR026481 ERR026482 ERR029206 ERR029207</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> ERR029206</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="ex">snippy</span> <span class="at">--ram</span> 8 <span class="at">--outdir</span> ../../results/snps/<span class="st">"</span><span class="va">${sample}</span><span class="st">"</span> <span class="at">--ref</span> ../GCF_000195955.2_ASM19595v2_genomic.fna <span class="at">--R1</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_1.fastq.gz_trim.fastq <span class="at">--R2</span> <span class="st">"</span><span class="va">${sample}</span><span class="st">"</span>_2.fastq.gz_trim.fastq</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="cf">done</span></span></code></pre>
</div>
<p>Here, we provide snippy with an output folder
(<code>--outdir</code>), the location of the reference genome
(<code>--ref</code>), and the trimmed read files for each end of the
pair (<code>--R1</code> and <code>--R2</code>). We also indicate that
snippy should use 8 Gb of memory (<code>--ram 8</code>).</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n10</span> ../../results/snps/ERR029206/snps.tab </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>CHROM	POS	TYPE	REF	ALT	EVIDENCE	FTYPE	STRAND	NT_POS	AA_POS	EFFECT	LOCUS_TAG	GENE	PRODUCT
NC_000962.3	1849	snp	C	A	A:217 C:0
NC_000962.3	1977	snp	A	G	G:118 A:0
NC_000962.3	4013	snp	T	C	C:176 T:0
NC_000962.3	7362	snp	G	C	C:156 G:0
NC_000962.3	7585	snp	G	C	C:143 G:0
NC_000962.3	9304	snp	G	A	A:127 G:0
NC_000962.3	11820	snp	C	G	G:164 C:0
NC_000962.3	11879	snp	A	G	G:125 A:0
NC_000962.3	14785	snp	T	C	C:163 T:1</code></pre>
</div>
<p>This list gives us information on every SNP that was found by snippy
when compared to the reference genome. The first SNP is found at the
position 1849 of the reference genome, and is a C in the H37Rv
(reference strain) and an A in isolate ERR029206. There is a high
confidence associated with this SNP: an A has been found 217 times in
the sequencing reads and never a C at this position.</p>
<p>Now copy a reduced version of the results on the other samples, save
it in the <code>results</code> folder. We will also rename the existing
<code>snps</code> folder to save it.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">mv</span> snps snps_partial</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">mkdir</span> snps</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="bu">cd</span> snps</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="fu">tar</span> xvzf ../../../../data/snps.tar.gz</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>ERR026473:
ref.fa.          snps.bam.bai  snps.filt.vcf  snps.log	     snps.tab  snps.vcf.gz
ref.fa.fai	     snps.bed      snps.gff       snps.raw.vcf   snps.txt  snps.vcf.gz.csi
snps.aligned.fa  snps.csv      snps.html      snps.subs.vcf  snps.vcf

ERR026474:
ref.fa		       snps.bam.bai  snps.filt.vcf  snps.log	     snps.tab  snps.vcf.gz
ref.fa.fai	     snps.bed      snps.gff       snps.raw.vcf   snps.txt  snps.vcf.gz.csi
snps.aligned.fa  snps.csv      snps.html      snps.subs.vcf  snps.vcf</code></pre>
</div>
<p>The folders don’t contain everything that is output by
<code>snippy</code>, to save space. But it is enough to run
<code>snippy-core</code>.</p>
<div id="challenge-how-many-snps-were-identified-in-each-sample" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-how-many-snps-were-identified-in-each-sample" class="callout-inner">
<h3 class="callout-title">Challenge: How many SNPs were identified in each sample??</h3>
<div class="callout-content">
<p>Find out how many SNPs were identified in the <em>M.
tuberculosis</em> isolates when compared to H37Rv. Hint: The
<code>.txt</code> file in the snippy output contains summary
information:</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">cat</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results/snps/ERR029207/snps.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DateTime        2023-05-17T13:51:57
ReadFiles       [...]/data/trimmed_fastq/ERR029207_1.fastq.gz_trim.fastq [...]/data/trimmed_fastq/ERR029207_2.fastq.gz_trim.fastq
Reference       [...]/data/GCF_000195955.2_ASM19595v2_genomic.fna
ReferenceSize   4411532
Software        snippy 4.6.0
Variant-COMPLEX 32
Variant-DEL     57
Variant-INS     52
Variant-MNP     2
Variant-SNP     1290
VariantTotal    1433</code></pre>
</div>
<p>This <em>M. tuberculosis</em> isolate contains 1290 SNPs compared to
H37Rv.</p>
<p>Repeat this for the other isolates. How about using a
<code>for</code> loop?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="core-snps">Core SNPs<a class="anchor" aria-label="anchor" href="#core-snps"></a>
</h2>
<hr class="half-width">
<p>In order to compare the identified SNPs with each other we need to
know if a certain position exists in all isolates. A core site can have
the same nucleotide in every sample (monomorphic) or some samples can be
different (polymorphic).</p>
<p>In this second part, after identifying them, snippy will concatenate
the core SNPs, i.e. ignoring sites that are monomorphic in all isolates
and in the reference. Concatenation of the SNP sites considerably
reduces the size of the alignment.</p>
<p>The <code>--ref</code> argument provides the reference genome. Each
folder containing the result of the previous step of <code>snippy</code>
is then added to the command line.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results/snps</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="ex">snippy-core</span> <span class="at">--ref</span><span class="op">=</span>../../data/GCF_000195955.2_ASM19595v2_genomic.fna ERR026473 ERR026474 ERR026478 ERR026481 ERR026482 ERR029206 ERR029207</span></code></pre>
</div>
<p>The last few lines look like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Loaded 1 sequences totalling 4411532 bp.
Will mask 0 regions totalling 0 bp ~ 0.00%
0	ERR026473	snp=1378	del=174	ins=165	het=747	unaligned=125683
1	ERR026474	snp=1369	del=180	ins=143	het=811	unaligned=123020
2	ERR026478	snp=1381	del=221	ins=143	het=638	unaligned=112145
3	ERR026481	snp=1349	del=215	ins=138	het=754	unaligned=126217
4	ERR026482	snp=1348	del=215	ins=145	het=911	unaligned=127554
5	ERR029206	snp=1352	del=152	ins=98	het=1839	unaligned=122677
6	ERR029207	snp=1355	del=152	ins=97	het=1900	unaligned=118570
Opening: core.tab
Opening: core.vcf
Processing contig: NC_000962.3
Generating core.full.aln
Creating TSV file: core.txt
Running: snp-sites -c -o core.aln core.full.aln
This analysis is totally hard-core!
Done.</code></pre>
</div>
<p>Our output was written to ‘core.aln’. But let’s have a look at the
results.</p>
<div id="discussion-whats-in-the-output-of-snippy" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-whats-in-the-output-of-snippy" class="callout-inner">
<h3 class="callout-title">Discussion: What’s in the output of snippy??</h3>
<div class="callout-content">
<p>Have a look at the content of these three files with <code>cat</code>
or <code>head</code>.</p>
<p><code>core.aln</code></p>
<p><code>core.full.aln</code></p>
<p><code>core.tab</code></p>
<p><code>core.txt</code></p>
<p>What is the difference between these files? Why is core.aln smaller
than ? What is in core.aln?</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="phylogenetic-tree">Phylogenetic tree<a class="anchor" aria-label="anchor" href="#phylogenetic-tree"></a>
</h2>
<hr class="half-width">
<p>Phylogenetic trees have been discussed during the lectures. We will
here establish a phylogenetic tree from the file ‘core.aln’ with <a href="http://www.iqtree.org/" class="external-link">IQ-TREE</a>. IQ-TREE is a phylogeny
software based on the maximum-likelihood principle. IQ-TREE has been
widely used and cited. There is a very wide range of parameters that
need to be chosen, such as nucleotide or amino-acid substitution
models.</p>
<p>In our case, we want IQ-TREE to automatically select the best
substitution model. IQ-TREE does that by testing many (among a very
large collection) substitution models. We also have SNP data, which by
definition do not contain constant (invariable) sites. We thus input the
alignment with the <code>-s</code> option and the model with
<code>-m MFP+ASC</code>. <code>MFP</code> will tell IQ-TREE to test a
range of models, and <code>ASC</code> will correct for the fact that
there is no constant sites.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">module</span> load iqtree</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> core.aln <span class="at">-m</span> MFP+ASC</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Total wall-clock time used: 0.312 sec (0h:0m:0s)

Analysis results written to:
  IQ-TREE report:                core.aln.iqtree
  Maximum-likelihood tree:       core.aln.treefile
  Likelihood distances:          core.aln.mldist
  Screen log file:               core.aln.log

Date and Time: Thu May 18 09:00:37 2023</code></pre>
</div>
<p>With this small data set, IQ-TREE finishes very quickly. Let’s put
the resulting files into a separate folder and let’s rename our
resulting tree.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">mkdir</span> tree</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="fu">mv</span> snps/core.aln.<span class="pp">*</span> tree</span></code></pre>
</div>
<p>Let’s inspect our tree, and give it another extension to make it
clear it is a newick file.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/molepi/results</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="bu">cd</span> tree</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="fu">mv</span> core.aln.treefile core.aln.newick </span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="fu">cat</span> core.aln.newick</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>(ERR026473:0.0004854701,ERR026474:0.0004356437,((((ERR026478:0.0000010000,ERR029207:0.0000010000):0.0000010000,ERR029206:0.0000131789):0.0006712219,Reference:0.0075887710):0.0000152316,(ERR026481:0.0000066238,ERR026482:0.0000131899):0.0005602421):0.0002373720);</code></pre>
</div>
<p>This does not look much like a tree yet. The tree is written in a
bracket annotation, the <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick format</a>. In
order to make sense of it we can better view this in a tree viewer.</p>
<div id="challenge-can-you-identify-what-substitution-model-iq-tree-used" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-can-you-identify-what-substitution-model-iq-tree-used" class="callout-inner">
<h3 class="callout-title">Challenge: Can you identify what substitution model IQ-TREE used?</h3>
<div class="callout-content">
<p>Hint: The log file of IQ-TREE contains a lot of information.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">cat</span> core.aln.iqtree</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>...
Best-fit model according to BIC: TVMe+ASC
...</code></pre>
</div>
<p>IQ-TREE chose the TVMe+ASC model. We’ve discussed the ASC above, and
you can read more about the TVMe in the <a href="https://www.iqtree.org/doc/Substitution-Models" class="external-link">IQ-TREE
manual</a>. In short, it is a mdoel where base frequencies are deemed
equal, and all rates estimated independently, except the rates of
transversions (A &lt;-&gt; G and C &lt;-&gt; T) are equal.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Single nucleotide polymorphisms can be identified by mapping reads
to a reference genome</li>
<li>Parameters for the analysis have to be selected based on expected
outcomes for this organism</li>
<li>Concatenation of SNPs helps to reduce analysis volume</li>
<li>Phylogenetic trees can be written with a bracket syntax in Newick
format</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-data-visualization"><p>Content from <a href="data-visualization.html">Molecular epidemiology</a></p>
<hr>
<p>Last updated on 2025-04-08 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/data-visualization.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p>The low mutation rate of <em>M. tuberculosis</em> means that even at
the highest resolution provided by whole genome sequencing it is still
difficult to confidently affirm the inferences of transmission made by
traditional epidemiological techniques. This means it is very difficult
to determine transmission inclusively. However, whole genome sequencing
does in some cases allow us to exclude direct transmission, by using the
phylogenetic context provided by other strains. Not only does whole
genome sequencing provide the inter-cluster differentiation provided by
current typing methods, but it also achieves intra-cluster resolution
which can be used to inform epidemiological investigation.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How are phylogenetic trees viewed and compared?</li>
<li>How can I visualize several layers of data?</li>
<li>In which cases is transmission likely?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explore the resulting trees in conjunction with the meta data.</li>
<li>Make an estimation on the likelihood of transmission events</li>
<li>Compare and interpret the different data sources.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="comparison-of-clustering-results">Comparison of clustering results<a class="anchor" aria-label="anchor" href="#comparison-of-clustering-results"></a>
</h2>
<hr class="half-width">
<p>We grouped (clustered) our isolates by the information that we
extracted from the sequencing reads. We compared our isolates to a
reference genome and clustered them on basis of the single nucleotide
variations that they exhibit when compared to this reference.</p>
<figure><img src="../fig/Workflow.png" alt="workflow" class="figure mx-auto d-block"></figure><p>Next, we will visualize this clustering in the context of the meta
data. Make sure you have the SNP tree that you generated (or can be
downloaded here <a href="../files/core.aln.newick">core.aln.newick</a>) on
your own computer. Download also the <a href="../files/Molepi_meta.csv">metadata file</a>.</p>
</section><section><h2 class="section-heading" id="visualization-of-genetic-information-and-metadata">Visualization of genetic information and metadata<a class="anchor" aria-label="anchor" href="#visualization-of-genetic-information-and-metadata"></a>
</h2>
<hr class="half-width">
<p>Visualization is frequently used to aid the interpretation of complex
datasets. Within microbial genomics, visualizing the relationships
between multiple genomes as a tree provides a framework onto which
associated data (geographical, temporal, phenotypic and epidemiological)
are added to generate hypotheses and to explore the dynamics of the
system under investigation.</p>
<ol style="list-style-type: decimal">
<li>Open Chrome. Go to <a href="https://microreact.org/" class="external-link">microreact</a>.</li>
<li>Click on ‘Upload’. Drag-and-drop the newick tree produced by snippy
and IQ-TREE (<a href="../files/core.aln.newick">core.aln.newick</a>).</li>
<li>Click on “Add more files” and select the the <a href="../files/Molepi_meta.csv">metadata file</a>, which is a table
comprising metadata file. Choose ‘Data (csv or tsv)’ as file kind.</li>
<li>Click on continue.</li>
</ol>
<p>First, the tree should be rooted with the reference strain (the one
that has the longest branch). Hover over the middle of the long branch,
until a circle appears. Right-click on it, and select “Set as root
(re-root)”.</p>
<p>Explore the location, time and further meta data. Play around with
the different visualization options with the help of the <a href="https://docs.microreact.org/" class="external-link">documentation of microreact</a>.</p>
<div id="discussion" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Which transmission events are likely based on the metadata
(location, RFLP, date) alone?</li>
<li>Which transmission events are likely based on the SNP data?</li>
<li>Draw a transmission tree if possible.</li>
</ol>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="discussion-1">Discussion<a class="anchor" aria-label="anchor" href="#discussion-1"></a>
</h2>
<hr class="half-width">
<p>Compare the pangenome clustering with the clustering based on SNPs
and with the data from epidemiological contact tracing:</p>
<p>Cluster A</p>
<ul>
<li>Patient A1 - sample ERR029207</li>
<li>Patient A2 - sample ERR029206</li>
<li>Patient A3 - sample ERR026478</li>
</ul>
<p>Cluster B</p>
<ul>
<li>Patient B1 - sample ERR026474</li>
<li>Patient B2 - sample ERR026473</li>
</ul>
<p>Cluster C</p>
<ul>
<li>Patient C1 - sample ERR026481</li>
<li>Patient C2 - sample ERR026482</li>
</ul>
<p>Which transmission events are affirmed by the genomic data? Which
ones partially or not? Why?</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>SNP phylogeny and metadata can convey different messages</li>
<li>Human interpretation is often needed to weigh the different
information sources.</li>
<li>The low mutation rate of <em>M. tuberculosis</em> does not allow to
make confident inferences of transmission but does allow to exclude
transmission</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:lionel.guy@imbim.uu.se">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "evolution, phylogenetics, molecular epidemiology",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/instructor/aio.html",
  "dateCreated": "2023-05-08",
  "dateModified": "2025-04-08",
  "datePublished": "2025-04-08"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

