<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Microbial Genomics: evolution, phylogenetics, and molecular epidemiology: Phylogenetics</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav data"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Data Carpentry" src="../assets/images/data-logo.svg"></div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../phylogenetics.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="../assets/images/data-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr></ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="../aio.html" role="button" aria-label="search button">search button</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Microbial Genomics: evolution, phylogenetics, and molecular epidemiology
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 33%" class="percentage">
    33%
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: 33%" aria-valuenow="33" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../phylogenetics.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="ncbi.html">1. NCBI</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="blast.html">2. BLAST</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="msa.html">3. Multiple sequence alignment</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Phylogenetics
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#exercise-1-paper-and-pen-phylogenetic-tree">Exercise 1: Paper-and-pen phylogenetic tree</a></li>
<li><a href="#exercise-2-neighbor-joining-and-maximum-likelihood-tree">Exercise 2: Neighbor-joining and Maximum-likelihood tree</a></li>
<li><a href="#exercise-3-genetic-drift">Exercise 3: Genetic drift</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="quality-control.html">5. Reads, QC and trimming</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="sequence-assembly.html">6. Sequence assembly</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="annotation.html">7. Annotation</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="snps-phylogeny.html">8. Mapping</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="data-visualization.html">9. Molecular epidemiology</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr></ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/msa.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/quality-control.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/msa.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Multiple sequence
        </a>
        <a class="chapter-link float-end" href="../instructor/quality-control.html" rel="next">
          Next: Reads, QC and... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Phylogenetics</h1>
        <p>Last updated on 2024-04-08 |
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/phylogenetics.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 240 minutes</p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do you build a simple, distance-based phylogenetic tree?</li>
<li>How do you build a phylogenetic tree with more advanced
methods?</li>
<li>How do you ascertain statistical support for phylogenetic
trees?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Learn the basic of phylogenetics tree building, taking the simplest
of the examples with a UPGMA tree.</li>
<li>Learn how to use phylogenetics algorithms, neighbor-joining and
maximum-likelihood.</li>
<li>Learn how to perform and show bootstraps.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="exercise-1-paper-and-pen-phylogenetic-tree"><h2 class="section-heading">Exercise 1: Paper-and-pen phylogenetic tree<a class="anchor" aria-label="anchor" href="#exercise-1-paper-and-pen-phylogenetic-tree"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a></h3>
<p>The exercise is done for a large part with pen and paper, and then a
demonstration in R on your laptop, using RStudio. We’ll also use the R
package <a href="https://emmanuelparadis.github.io/" class="external-link"><code>ape</code></a>, which
you should install if it’s not present on your setup. Commands can be
typed or pasted in the “Console” part of RStudio.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">'ape'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following package(s) will be installed:
- ape [5.7-1]
These packages will be installed into "~/work/course-microbial-genomics/course-microbial-genomics/renv/profiles/lesson-requirements/renv/library/R-4.3/x86_64-pc-linux-gnu".

# Installing packages --------------------------------------------------------
- Installing ape ...                            OK [linked from cache]
Successfully installed 1 package in 5.4 milliseconds.</code></pre>
</div>
<p>And to load it:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ape</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="upgma-based-tree">UPGMA-based tree<a class="anchor" aria-label="anchor" href="#upgma-based-tree"></a></h3>
<p>Load the tree in fasta format, reading from a <code>temp</code>
file</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FASTAfile</span> <span class="op">&lt;-</span> <span class="fu">tempfile</span><span class="op">(</span><span class="st">"aln"</span>, fileext <span class="op">=</span> <span class="st">".fasta"</span><span class="op">)</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"&gt;A"</span>, <span class="st">"----ATCCGCTGATCGGCTG----"</span>,</span>
<span>    <span class="st">"&gt;B"</span>, <span class="st">"GCTGATCCGTTGATCGG-------"</span>,</span>
<span>    <span class="st">"&gt;C"</span>, <span class="st">"----ATCTGCTCATCGGCT-----"</span>,</span>
<span>    <span class="st">"&gt;D"</span>, <span class="st">"----ATTCGCTGAACTGCTGGCTG"</span>,</span>
<span>    file <span class="op">=</span> <span class="va">FASTAfile</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu">read.dna</span><span class="op">(</span><span class="va">FASTAfile</span>, format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span></code></pre>
</div>
<p>Now look at the alignment. Notice there are gaps, which we don’t want
in this example. We also remove the non-informative (identical)
columns.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">alview</span><span class="op">(</span><span class="va">aln</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  000000000111111111122222
  123456789012345678901234
A ----ATCCGCTGATCGGCTG----
B GCTG.....T.......---....
C .......T...C.......-....
D ......T......A.T....GCTG</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aln_filtered</span> <span class="op">&lt;-</span> <span class="va">aln</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">7</span>,<span class="fl">8</span>,<span class="fl">10</span>,<span class="fl">12</span>,<span class="fl">14</span>,<span class="fl">16</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">alview</span><span class="op">(</span><span class="va">aln_filtered</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  123456
A CCCGTG
B ..T...
C .T.C..
D T...AT</code></pre>
</div>
<p>Now we have a simple alignment as in the lecture. Dots
(<code>.</code>) mean that the sequence is identical to the top row,
which makes it easier to calculate distances.</p>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<a class="anchor" aria-label="anchor" href="#calculating-distance-by-hand"></a>
</h3>
<div class="callout-content">
<p>Let’s use a matrix to calculate distances between sequences. We’ll
just count the number of differences between the sequences. We’ll then
group the two closest sequences. Which are they?</p>
<table class="table"><caption>Distances between the sequences.</caption>
<thead><tr class="header"><th></th>
<th align="right">A</th>
<th align="right">B</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead><tbody><tr class="odd"><td>A</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>B</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="odd"><td>C</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>D</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr></tbody></table></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>Here is the solution:</p>
<table class="table"><caption>Distances between the sequences.</caption>
<thead><tr class="header"><th></th>
<th align="right">A</th>
<th align="right">B</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead><tbody><tr class="odd"><td>A</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>B</td>
<td align="right">1</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="odd"><td>C</td>
<td align="right">2</td>
<td align="right">3</td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>D</td>
<td align="right">3</td>
<td align="right">4</td>
<td align="right">5</td>
<td align="right"></td>
</tr></tbody></table><p>The two closest sequences are A and B.</p>
</div>
</div>
</div>
</div>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<em>(continued)</em><a class="anchor" aria-label="anchor" href="#calculating-distance-by-hand"></a>
</h3>
<div class="callout-content">
<p>Let’s now cluster together A and B, and calculate the average
distance from AB to the other sequences, weighted by the size of the
clusters.</p>
<table class="table"><caption>Recalculated distances.</caption>
<thead><tr class="header"><th></th>
<th align="right">AB</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead><tbody><tr class="odd"><td>AB</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>C</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="odd"><td>D</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr></tbody></table></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The average distance for AB to C is calculated as follow:</p>
<p><span class="math inline">\(d(AB,C) = \dfrac{d(A,C) + d(B,C)}{2} =
\dfrac{2 + 3}{2} = 2.5\)</span></p>
<p>And so on for the other distances:</p>
<table class="table"><caption>Recalculated distances.</caption>
<thead><tr class="header"><th></th>
<th align="right">AB</th>
<th align="right">C</th>
<th align="right">D</th>
</tr></thead><tbody><tr class="odd"><td>AB</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>C</td>
<td align="right">2.5</td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="odd"><td>D</td>
<td align="right">3.5</td>
<td align="right">5</td>
<td align="right"></td>
</tr></tbody></table></div>
</div>
</div>
</div>
<div id="calculating-distance-by-hand" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="calculating-distance-by-hand" class="callout-inner">
<h3 class="callout-title">Calculating distance “by hand”<em>(continued)</em><a class="anchor" aria-label="anchor" href="#calculating-distance-by-hand"></a>
</h3>
<div class="callout-content">
<p>Now the shortest distance is AB to C. Let’s recalculate the distance
to D again.</p>
<table class="table"><caption>Recalculated distances.</caption>
<thead><tr class="header"><th></th>
<th align="right">ABC</th>
<th align="right">D</th>
</tr></thead><tbody><tr class="odd"><td>ABC</td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>D</td>
<td align="right"></td>
<td align="right"></td>
</tr></tbody></table></div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><span class="math inline">\(d(ABC,D) = \dfrac{d(AB,D) * 2 + d(C,D) *
1}{2 + 1} = \dfrac{3.5 * 2 + 5 * 1}{3} = 4\)</span></p>
<table class="table"><caption>Recalculated distances.</caption>
<thead><tr class="header"><th></th>
<th align="right">ABC</th>
<th align="right">D</th>
</tr></thead><tbody><tr class="odd"><td>ABC</td>
<td align="right"></td>
<td align="right"></td>
</tr><tr class="even"><td>D</td>
<td align="right">4</td>
<td align="right"></td>
</tr></tbody></table></div>
</div>
</div>
</div>
<p>The tree is reconstructed by dividing the distances equally between
the two leaves. - A-B: each 0.5. - AB-C: each side gets 2.5/2 = 1.25.
The branch to AB is 1.25 - 0.75 = 0.75 - ABC-D: each side gets 4/2 = 2.
The branch to ABC is 2 - 0.75 - 0.5 = 0.75</p>
<figure><img src="../fig/upgma_manual.png" alt="Manually built UPGMA tree" class="figure mx-auto d-block"><div class="figcaption">UPGMA tree</div>
</figure><p>Let’s know do the same using bioinformatics tools.</p>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge4"></a>
</h3>
<div class="callout-content">
<p>We’ll use <code>dist.dna</code> to calculate the distances. We’ll use
a “N” model, that just counts the differences and doesn’t correct or
normalizes. We’ll use the function <code>hclust</code> to perform the
UPGMA method calculation. The tree is then plotted, and the branch
lengths plotted with <code>edgelabels</code>:</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Show me the solution</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist_matrix</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">aln_filtered</span>, model<span class="op">=</span><span class="st">"N"</span><span class="op">)</span></span>
<span><span class="va">dist_matrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  A B C
B 1    
C 2 3  
D 3 4 5</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">as.phylo</span><span class="op">(</span><span class="fu">hclust</span><span class="op">(</span><span class="va">dist_matrix</span>, <span class="st">"average"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="fu">edgelabels</span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">edge.length</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/phylogenetics-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
</div>
</section><section id="exercise-2-neighbor-joining-and-maximum-likelihood-tree"><h2 class="section-heading">Exercise 2: Neighbor-joining and Maximum-likelihood tree<a class="anchor" aria-label="anchor" href="#exercise-2-neighbor-joining-and-maximum-likelihood-tree"></a>
</h2>
<hr class="half-width"><div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h3>
<p>In the previous episode, we inferred a simple phylogenetic tree using
UPGMA, without correcting the distance matrix for multiple
substitutions. UPGMA has many shortcomings, and gets worse as distance
increases. Here we’ll test Neighbor-Joining, which is also a
distance-based, and a maximum-likelihood based approach with <a href="http://www.iqtree.org/" class="external-link">IQ-TREE</a>.</p>
</div>
<div class="section level3">
<h3 id="neighbor-joining">Neighbor-joining<a class="anchor" aria-label="anchor" href="#neighbor-joining"></a></h3>
<p>The principle of <a href="https://en.wikipedia.org/wiki/Neighbor_joining" class="external-link">Neighbor-joining
method</a> (NJ) is to start from a star-like tree, find the two branches
that, if joined, minimize the total tree length, join then, and repeat
with the joined branches and the rest of the branches.</p>
<div id="challenge5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge5"></a>
</h3>
<div class="callout-content">
<p>To perform NJ on our sequences, we’ll use the function inbuilt in
Seaview. First (re-)open the alignment we obtained from
<code>mafft</code> with the E-INS-i method, trimmed with ClipKIT. If it
is not on your computer anymore, transfer it again from Uppmax using
<code>scp</code> or <code>SFTP</code>.</p>
<p>If <a href="https://doua.prabi.fr/software/seaview" class="external-link">Seaview</a> is
not installed on your computer, download it and install it on your
computer.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">Show me the solution</h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p>On a Linux computer, you can run it on the command line:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">seaview</span> rpoB.einsi.clipkit.aln <span class="kw">&amp;</span></span></code></pre>
</div>
<p>On other computer, just open the file with the regular File &gt; Open
menu.</p>
</div>
</div>
</div>
</div>
<div id="discussion1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<em>(continued)</em><a class="anchor" aria-label="anchor" href="#discussion1"></a>
</h3>
<div class="callout-content">
<p>In the Seaview window, select Trees -&gt; Distance methods. Keep the
BioNJ option ticked. BioNJ is an updated version of NJ, more accurate
for longer distances. Tick the “Bootstrap” option and leave it at 100.
We’ll discuss these later. Then click on “Go”.</p>
</div>
</div>
</div>
<p>What do you see? Is the tree rooted? Is the pattern species coherent
with what you know of the tree of life? Any weird results?</p>
<div id="challenge6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge6"></a>
</h3>
<div class="callout-content">
<p>Redo the BioNJ tree for the other alignment we inferred.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">Show me the solution</h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">seaview</span> rpoB.fftns.aln <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Do you see any differences between the two trees? What can you make
out of it?</p>
</div>
<div class="section level3">
<h3 id="maximum-likelihood">Maximum likelihood<a class="anchor" aria-label="anchor" href="#maximum-likelihood"></a></h3>
<p>We will now use IQ-TREE to infer a maximum-likelihood (ML) tree of
the RpoB dataset we aligned with <code>mafft</code> previously. While
you performed the first trees on your laptop, you will infer the ML tree
on Uppmax. As usual, use the <code>interactive</code> command to gain
access to compute time. Go to the <code>phylogenetics</code> folder
created in the MSA episode, and load the right modules.</p>
<div id="challenge7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge7"></a>
</h3>
<div class="callout-content">
<p>Get to the right folder, require compute time and load the right
modules.</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7">Show me the solution</h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">interactive</span> <span class="at">-A</span> uppmax2024-2-10 <span class="at">-M</span> snowy <span class="at">-t</span> 4:00:00</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="ex">module</span> load bioinfo-tools iqtree</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="bu">cd</span> /proj/g2020004/nobackup/3MK013/<span class="op">&lt;</span>username<span class="op">&gt;</span>/phylogenetics</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Loading the <code>iqtree</code> module resulted in:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>iqtree/2.2.2.6-omp-mpi: executables are prefixed with 'iqtree2'. Additional modules are required for the multi-node MPI version. See 'module help iqtree/2.2.2.6-omp-mpi'.</code></pre>
</div>
<p>You won’t be using the multi-node MPI version, so you don’t need to
load additional modules. But you will need to use <code>iqtree2</code>
instead of the “normal” <code>iqtree</code>.</p>
<p>Then run your first tree, on the <code>ClipKIT</code>-trimmed
alignment.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> rpoB.einsi.clipkit.aln <span class="at">-m</span> TEST <span class="at">-B</span> 1000</span></code></pre>
</div>
<p>Here, we tell IQ-TREE to use the alignment
<code>rpoB.einsi.clipkit.aln</code>, and to test among the standard
substitution models which one fits best (<code>-m TEST</code>). We also
tell IQ-TREE to perform 1000 ultra-fast bootstraps
(<code>-B 1000</code>). We’ll discuss these later.</p>
<p>IQ-TREE is a very complete program that can do a large variety of
phylogenetic analysis. To get a flavor of what it’s capable of, look at
its <a href="http://www.iqtree.org/doc/" class="external-link">extensive
documentation</a>.</p>
<p>Have a look at the output files:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">ls</span> rpoB.einsi.clipkit.aln.<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>rpoB.einsi.clipkit.aln.bionj    rpoB.einsi.clipkit.aln.iqtree  rpoB.einsi.clipkit.aln.model.gz
rpoB.einsi.clipkit.aln.ckp.gz   rpoB.einsi.clipkit.aln.log     rpoB.einsi.clipkit.aln.splits.nex
rpoB.einsi.clipkit.aln.contree	rpoB.einsi.clipkit.aln.mldist  rpoB.einsi.clipkit.aln.treefile</code></pre>
</div>
<p>There are two important files:<br>
* <code>*.iqtree</code> file provides a text summary of the analysis. *
<code>*.treefile</code> is the resulting tree in <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick
format</a>.</p>
<p>To visualize the tree, open the <a href="https://beta.phylo.io/viewer/#" class="external-link">Beta version of phylo.io</a>.
Click on “add a tree” and copy/paste the content of the
<code>treefile</code> (use e.g. <code>less</code> or <code>cat</code> to
display it) in the box. Make sure the “Newick format” is selected and
click on “Done”.</p>
<p>The tree appears as unrooted. It is good practice to start by
ordering the nodes and root it. There is no good way to automatically
order the branches in phylo.io as of yet, but rerooting can be done by
clicking on a branch and selecting “Reroot”. Reroot the tree between
archaea and bacteria. Now the tree makes a bit more sense.</p>
<p>Scrutinize the tree. Is it different from the BioNJ tree generated in
Seaview? How?</p>
<div id="challenge8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge8"></a>
</h3>
<div class="callout-content">
<p>Redo a ML tree from the other alignment (FFT-NS) we inferred with
<code>mafft</code> and display the resulting tree in FigTree</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8">Show me the solution</h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> rpoB.fftns.aln <span class="at">-m</span> TEST <span class="at">-B</span> 1000</span></code></pre>
</div>
<p>Import the tree on your computer and load it with FigTree or with <a href="https://beta.phylo.io/viewer/#" class="external-link">phylo.io</a></p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">figtree</span> rpoB.fftns.aln.treefile <span class="kw">&amp;</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="bootstraps">Bootstraps<a class="anchor" aria-label="anchor" href="#bootstraps"></a></h3>
<p>We have inferred four trees: * Two based on the alignment generated
from the E-INS-i algorithm (trimmed by ClipKIT), two from the FFT-NS. *
Two inferred with the BioNJ algorithm and two with the ML algorithm
(IQ-Tree)</p>
<p>Along the way, we’ve generated bootstraps for all our trees. Now show
them on all four trees. * For the Seaview trees, tick the ‘Br support’
box * For the trees shown in phylo.io, click on “Settings” &gt; “Branch
&amp; Labels” and above the branch, click the drop-down menu and select
“Data”.</p>
<div id="challenge9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge9"></a>
</h3>
<div class="callout-content">
<p>Compare all four trees. Do you find any significant differences?</p>
<p>Hint: what are <em>Glycine</em> and <em>Arabidopsis</em>? What about
<em>Synechocystis</em> and <em>Microcystis</em>?</p>
<p>Hint: search for the accession number of the <em>Glycine</em> RpoB
sequence on <a href="https://www.ncbi.nlm.nih.gov/search/" class="external-link">NCBI</a>. Any
hint there?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9">Show me the solution</h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<p>The two plant RpoB sequences are from chloroplastic genomes, and thus
branch together with the cyanobacterial sequences, in some of the
phylogenies at least.</p>
</div>
</div>
</div>
</div>
<div id="discussion2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<em>(continued)</em><a class="anchor" aria-label="anchor" href="#discussion2"></a>
</h3>
<div class="callout-content">
<p>What about the support values for grouping these two groups? How high
are they?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>The simplest of the trees are distance-based.</li>
<li>UPGMA works by clustering the two closest leaves and recalculating
the distance matrix.</li>
<li>Neighbor-joining is distance-based and fast, but not necessarily
very accurate</li>
<li>Maximum-likelihood is slower, but more accurate</li>
</ul></div>
</div>
</div>
</div>
</section><section id="exercise-3-genetic-drift"><h2 class="section-heading">Exercise 3: Genetic drift<a class="anchor" aria-label="anchor" href="#exercise-3-genetic-drift"></a>
</h2>
<hr class="half-width"><p>As a practical way to understand genetic drift, let’s play with
population size, selection coefficients, number of generations, etc.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How does random genetic drift influence fixation of alleles?</li>
<li>What is the influence of population size?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Learn the basic of phylogenetics tree building, taking the simplest
of the examples with a UPGMA tree.</li>
<li>Learn how to use phylogenetics algorithms, neighbor-joining and
maximum-likelihood.</li>
<li>Learn how to perform and show bootstraps.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How does random genetic drift influence fixation of alleles?</li>
<li>What is the influence of population size?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Understand how population size influences the probability of
fixation of an allele.</li>
<li>Understand how slightly deleterious mutations may get fixed in small
populations.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a></h3>
<p>This exercise is to illustrate the concepts of selection, population
size and genetic drift, using simulations. We will use mostly <a href="https://cartwrig.ht/apps/redlynx/" class="external-link">Red Lynx</a> by <a href="https://github.com/reedacartwright" class="external-link">Reed A. Cartwright</a>.</p>
<p>Another option is to use a <a href="https://phytools.shinyapps.io/drift-selection/" class="external-link">web interface</a>
to the <a href="https://www.r-project.org/" class="external-link">R</a> <a href="https://github.com/liamrevell/learnPopGen" class="external-link">learnPopGen</a>
package, but the last one is mostly for biallelic genes (and thus not
that relevant for bacteria).</p>
<p>Open now the <a href="https://cartwrig.ht/apps/redlynx/" class="external-link">Red Lynx</a>
website and get familiar with the different options.</p>
<p>You won’t need the right panel (but feel free to explore). The
dominance option in the left panel won’t be used either.</p>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>In the first part, you will only play with the number of generations,
the initial frequency and the population size.</p>
<ul><li>Lower the number of generations to 200.</li>
<li>Adjust the population size to 1000.</li>
<li>Make sure the intial frequency is set to 50%.</li>
<li>Run a few simulations (ca 20).</li>
</ul><p>Did any allele got fixed? What is the range of frequencies after 200
generations?</p>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10">Show me the solution</h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" aria-labelledby="headingSolution10" data-bs-parent="#accordionSolution10">
<div class="accordion-body">
<p>In my simulations, no allele got fixed, the final allele frequencies
range 20-80%</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>(continued)</em><a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>Now increase the population to 100’000, clear the graph, and repeat
the simulations. What’s the range of final frequencies now?</p>
</div>
</div>
</div>
<div id="accordionSolution11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution11" aria-expanded="false" aria-controls="collapseSolution11">
  <h4 class="accordion-header" id="headingSolution11">Show me the solution</h4>
</button>
<div id="collapseSolution11" class="accordion-collapse collapse" aria-labelledby="headingSolution11" data-bs-parent="#accordionSolution11">
<div class="accordion-body">
<p>In my simulations, no allele got fixed, and the final allele
frequencies range 45-55%</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>(continued)</em><a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>Now decrease the population to 10 individuals, clear the graph and
repeat these simulations. What’s the range of final frequencies now?</p>
</div>
</div>
</div>
<div id="accordionSolution12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution12" aria-expanded="false" aria-controls="collapseSolution12">
  <h4 class="accordion-header" id="headingSolution12">Show me the solution</h4>
</button>
<div id="collapseSolution12" class="accordion-collapse collapse" aria-labelledby="headingSolution12" data-bs-parent="#accordionSolution12">
<div class="accordion-body">
<p>In my simulations, one allele got fixed quickly, the latest one was
at generation 100.</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-without-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-without-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift without selection<em>(continued)</em><a class="anchor" aria-label="anchor" href="#genetic-drift-without-selection"></a>
</h3>
<div class="callout-content">
<p>What do you conclude here?</p>
</div>
</div>
</div>
<div id="accordionSolution13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution13" aria-expanded="false" aria-controls="collapseSolution13">
  <h4 class="accordion-header" id="headingSolution13">Show me the solution</h4>
</button>
<div id="collapseSolution13" class="accordion-collapse collapse" aria-labelledby="headingSolution13" data-bs-parent="#accordionSolution13">
<div class="accordion-body">
<p>It is clear that stochastic (random) variation in allele frequencies
strongly affects the probability of fixation of alleles in small
populations, not so much in large ones.</p>
</div>
</div>
</div>
</div>
<div id="genetic-drift-with-selection" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="genetic-drift-with-selection" class="callout-inner">
<h3 class="callout-title">Genetic Drift with selection<a class="anchor" aria-label="anchor" href="#genetic-drift-with-selection"></a>
</h3>
<div class="callout-content">
<p>So far we’ve only looked at how allele frequencies vary in the
absence of selection, that is when the two alleles provide an equal
fitness. What’s the influence of random genetic drift when alleles are
not neutral?</p>
<p>The selection strength is equivalent to the selection coefficient,
i.e. how much higher relative fitness the new allele provides. A
selection coefficient of 0.01 means that the organism harboring the new
allele has a 1% increased fitness.</p>
<ul><li>Increase the number of generations to 1000.</li>
<li>Set the selection strength to 0.01 (1%).</li>
<li>Set the population size first to 100’000, then to 1000, then to
100.</li>
</ul><p>How long does it take for the allele to get fixed - in average - with
the three population sizes?</p>
</div>
</div>
</div>
<div id="accordionSolution14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution14" aria-expanded="false" aria-controls="collapseSolution14">
  <h4 class="accordion-header" id="headingSolution14">Show me the solution</h4>
</button>
<div id="collapseSolution14" class="accordion-collapse collapse" aria-labelledby="headingSolution14" data-bs-parent="#accordionSolution14">
<div class="accordion-body">
<p>About the same time, but the trajectories are much smoother with
larger populations. In the small population, it happens that the
beneficial allele disappears from the population (although not
often).</p>
</div>
</div>
</div>
</div>
<div id="fixation-of-slightly-deleterious-alleles-in-very-small-populations" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="fixation-of-slightly-deleterious-alleles-in-very-small-populations" class="callout-inner">
<h3 class="callout-title">Fixation of slightly deleterious alleles in
very small populations<a class="anchor" aria-label="anchor" href="#fixation-of-slightly-deleterious-alleles-in-very-small-populations"></a>
</h3>
<div class="callout-content">
<p>We’re now simulating what would happen in a very small population (or
a population that undergoes very narrow bottlenecks), when a gene
mutates. We’ll have a very small population (10 individuals), a
selection value of -0.01 (the mutated allele provides a 1% lower
fitness), and a 10% initial frequency, which corresponds to one
individual getting a mutation:</p>
<ul><li>Set the population to 10 individuals.</li>
<li>Set generations to 200.</li>
<li>Set initial frequency to 10%.</li>
<li>Set the selection strength to -0.01.</li>
</ul><p>Run many simulations. What happens?</p>
</div>
</div>
</div>
<div id="accordionSolution15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution15" aria-expanded="false" aria-controls="collapseSolution15">
  <h4 class="accordion-header" id="headingSolution15">Show me the solution</h4>
</button>
<div id="collapseSolution15" class="accordion-collapse collapse" aria-labelledby="headingSolution15" data-bs-parent="#accordionSolution15">
<div class="accordion-body">
<p>Most new alleles go extinct quickly. In my simulations, I usually get
one of the slightly deleterious mutations fixed after 20
simulations.</p>
</div>
</div>
</div>
</div>
<p>That’s it for that part, but feel free to continue playing with the
different settings later.</p>
<div id="keypoints2" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints2"></a>
</h3>
<div class="callout-content">
<ul><li>Random genetic drift has a large influence on the probability of
fixation of alleles in small populations, even for non-neutral
alleles.</li>
<li>Random genetic drift has very little influence of the probability of
fixation of alleles in large populations.</li>
<li>Slightly deleterious mutations can get fixed into the population
through random genetic drift, if the population is small enough and the
selective value is not too large.</li>
</ul></div>
</div>
</div>
<!-- References -->
<!-- 
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use. 
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/msa.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/quality-control.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/msa.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Multiple sequence
        </a>
        <a class="chapter-link float-end" href="../instructor/quality-control.html" rel="next">
          Next: Reads, QC and... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/phylogenetics.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:lionel.guy@imbim.uu.se">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper" class="external-link">sandpaper (0.16.4.9000)</a>, <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.7.5.9000)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries.github.io/workbench-template-rmd/instructor/phylogenetics.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "evolution, phylogenetics, molecular epidemiology",
  "name": "Phylogenetics",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/instructor/phylogenetics.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/instructor/phylogenetics.html",
  "dateCreated": "2023-05-08",
  "dateModified": "2024-04-08",
  "datePublished": "2024-05-14"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

