---
title: 'Phylogenetics - UPGMA'
teaching: 10
exercises: 20
---

:::::::::::::::::::::::::::::::::::::: questions 

- How do you build a phylogenetic tree?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Learn the basic of phylogenetics tree building, taking the simplest of
the examples with a UPGMA tree.

::::::::::::::::::::::::::::::::::::::::::::::::

## Setup

The exercise is done for a large part with pen and paper, and then a 
demonstration in R. We'll also use the R package 
[`ape`](https://emmanuelparadis.github.io/) , which you should
install if it's not present on your setup:

```r
install.packages('ape')
```

And to load it:

```{r}
library(ape)
```

## UPGMA-based tree

Load the tree in fasta format, reading from a `temp` file
```{r}
FASTAfile <- tempfile("aln", fileext = ".fasta")
cat(">A", "----ATCCGCTGATCGGCTG----",
    ">B", "GCTGATCCGTTGATCGG-------",
    ">C", "----ATCTGCTCATCGGCT-----",
    ">D", "----ATTCGCTGAACTGCTGGCTG",
    file = FASTAfile, sep = "\n")
aln <- read.dna(FASTAfile, format = "fasta")
```
Now look at the alignment. Notice there are gaps, which we don't want in 
this example. We also remove the non-informative (identical) columns. 

```{r}
alview(aln)
aln_filtered <- aln[,c(7,8,10,12,14,16)]
alview(aln_filtered)
```
Now we have a simple alignment as in the lecture. Dots (`.`) mean that the
sequence is identical to the top row, which makes it easier to calculate 
distances.

::::::::::::::::::::::::::::::::::::: challenge 

## Calculating distance "by hand"

Let's use a matrix to calculate distances between sequences. We'll just count
the number of differences between the sequences. We'll then group the two 
closest sequences. Which are they?

Table: Distances between the sequences.

|   | A  | B  | C  | D  |
| - | -: | -: | -: | -: |
| A |    |    |    |    |
| B |    |    |    |    |
| C |    |    |    |    |
| D |    |    |    |    |

:::::::::::::::::::::::: solution 

Here is the solution:

Table: Distances between the sequences.

|   | A  | B  | C  | D  |
| - | -: | -: | -: | -: |
| A |    |    |    |    |
| B |  1 |    |    |    |
| C |  2 | 3  |    |    |
| D |  3 | 4  | 5  |    |

The two closest sequences are A and B.

:::::::::::::::::::::::::::::::::

Let's now cluster together A and B, and calculate the average distance 
from AB to the other sequences, weighted by the size of the clusters. 

Table: Recalculated distances.

|    | AB  | C  | D  |
| -  | -:  | -: | -: |
| AB |     |    |    |
| C  |     |    |    |
| D  |     |    |    |

:::::::::::::::::::::::: solution 

The average distance for AB to C is calculated as follow:

$d(AB,C) = \dfrac{d(A,C) + d(B,C)}{2} = \dfrac{2 + 3}{2} = 2.5$

And so on for the other distances:

Table: Recalculated distances.

|    | AB  | C  | D  |
| -  | -:  | -: | -: |
| AB |     |    |    |
| C  | 2.5 |    |    |
| D  | 3.5 |  5 |    |

:::::::::::::::::::::::::::::::::

Now the shortest distance is AB to C. Let's recalculate the distance to D again.

Table: Recalculated distances.

|    | ABC | D  |
| -  | -:  | -: |
| ABC|     |    |
| D  |     |    |

:::::::::::::::::::::::: solution 

$d(ABC,D) = \dfrac{d(AB,D) * 2 + d(C,D) * 1}{2 + 1} = \dfrac{3.5 * 2 + 5 * 1}{3} = 4$

Table: Recalculated distances.

|    | ABC | D  |
| -  | -:  | -: |
| ABC|     |    |
| D  | 4   |    |

:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::

The tree is reconstructed by dividing the distances equally between the two
leaves. 
- A-B: each 0.5.
- AB-C: each side gets 2.5/2 = 1.25. The branch to AB is 1.25 - 0.75 = 0.75
- ABC-D: each side gets 4/2 = 2. The branch to ABC is 2 - 0.75 - 0.5 = 0.75

![UPGMA tree](fig/upgma_manual.png){alt='Manually built UPGMA tree'}


Let's know do the same using bioinformatics tools. 

::::::::::::::::::::::::::::::::::::: challenge 

We'll use `dist.dna` to 
calculate the distances. We'll use a "N" model, that just counts the differences
and doesn't correct or normalizes. We'll use the function `hclust` to perform
the UPGMA method calculation. The tree is then plotted, and the branch 
lengths plotted with `edgelabels`:

:::::::::::::::::::::::: solution 


```{r}
dist_matrix <- dist.dna(aln_filtered, model="N")
dist_matrix
tree <- as.phylo(hclust(dist_matrix, "average"))
plot(tree)
edgelabels(tree$edge.length)

```

:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::


::::::::::::::::::::::::::::::::::::: keypoints 

- The simplest of the trees are distance-based.
- UPGMA works by clustering the two closest leaves and recalculating the 
distance matrix.

::::::::::::::::::::::::::::::::::::::::::::::::

